// koishi-plugin-gocqhttp/dist/index.js
import { defineComponent as Q, inject as k, ref as S, computed as D, resolveComponent as f, openBlock as s, createElementBlock as l, Fragment as _, unref as u, createBlock as q, withCtx as p, createElementVNode as o, createVNode as a, createTextVNode as c, toDisplayString as b, createCommentVNode as h, isRef as I, pushScopeId as j, popScopeId as N } from "../vue.js";
import { store as U, Schema as g, send as m, defineExtension as K } from "../client.js";
const A = (i) => (j("data-v-d9529b45"), i = i(), N(), i), O = { key: 0 }, R = { key: 1 }, T = { key: 2 }, z = /* @__PURE__ */ A(() => /* @__PURE__ */ o("p", null, "\u8BF7\u4F7F\u7528\u624B\u673A\u767B\u5F55 QQ \u626B\u63CF\u4E8C\u7EF4\u7801\uFF1A", -1)), G = ["src"], H = /* @__PURE__ */ A(() => /* @__PURE__ */ o("p", null, "\u8BF7\u586B\u5199\u56FE\u4E2D\u7684\u5185\u5BB9\uFF1A", -1)), J = ["src"], L = { class: "submit" }, M = /* @__PURE__ */ A(() => /* @__PURE__ */ o("p", null, "\u8BF7\u8F93\u5165\u77ED\u4FE1\u9A8C\u8BC1\u7801\uFF1A", -1)), P = { class: "submit" }, W = { class: "submit" }, X = /* @__PURE__ */ A(() => /* @__PURE__ */ o("p", null, "\u8D26\u53F7\u5DF2\u5F00\u542F\u8BBE\u5907\u9501\u3002\u8BF7\u9009\u62E9\u9A8C\u8BC1\u65B9\u5F0F (\u5C06\u5728 10 \u79D2\u540E\u81EA\u52A8\u9009\u62E9 2)\uFF1A", -1)), Y = { class: "submit" }, Z = /* @__PURE__ */ A(() => /* @__PURE__ */ o("p", null, "\u767B\u5F55\u9700\u8981\u6ED1\u6761\u9A8C\u8BC1\u7801\u3002\u8BF7\u9009\u62E9\u9A8C\u8BC1\u65B9\u5F0F (\u5C06\u5728 10 \u79D2\u540E\u81EA\u52A8\u9009\u62E9 1)\uFF1A", -1)), uu = { class: "submit" }, tu = { key: 9 }, eu = ["href"], ou = /* @__PURE__ */ Q({
  __name: "settings",
  setup(i) {
    const C = k("manager.settings.local"), d = k("manager.settings.config"), v = k("manager.settings.current"), F = S(""), r = D(() => {
      var E, t;
      if (C.value.name === "@koishijs/plugin-adapter-onebot")
        return `${((E = d.value) == null ? void 0 : E.platform) || "onebot"}:${(t = d.value) == null ? void 0 : t.selfId}`;
    }), e = D(() => {
      var E;
      return (E = U.gocqhttp) == null ? void 0 : E[r.value];
    }), V = D(() => {
      if (!!e.value && e.value.status !== "init")
        return e.value.status === "offline" ? "error" : e.value.status === "success" ? "success" : "warning";
    }), w = g.object({
      gocqhttp: g.object({
        enabled: g.boolean().default(!1).description("\u662F\u5426\u81EA\u52A8\u521B\u5EFA go-cqhttp \u5B50\u8FDB\u7A0B\u3002"),
        password: g.string().role("secret").description("\u673A\u5668\u4EBA\u7684\u5BC6\u7801\u3002")
      }).description("go-cqhttp \u57FA\u7840\u8BBE\u7F6E")
    });
    return (E, t) => {
      const y = f("el-input"), B = f("k-button"), x = f("k-comment"), $ = f("k-form");
      return s(), l(_, null, [
        u(e) ? (s(), q(x, {
          key: 0,
          class: "gocqhttp",
          type: u(V)
        }, {
          default: p(() => [
            u(e).status === "offline" ? (s(), l("p", O, " \u672A\u8FDE\u63A5\u5230 go-cqhttp \u5B50\u8FDB\u7A0B\u3002 ")) : u(e).status === "init" ? (s(), l("p", R, " \u6B63\u5728\u521B\u5EFA go-cqhttp \u5B50\u8FDB\u7A0B\u2026\u2026 ")) : u(e).status === "success" ? (s(), l("p", T, " \u5DF2\u6210\u529F\u8FDE\u63A5 go-cqhttp \u5B50\u8FDB\u7A0B\u3002 ")) : u(e).status === "qrcode" ? (s(), l(_, { key: 3 }, [
              z,
              o("img", {
                class: "qrcode",
                src: u(e).image
              }, null, 8, G)
            ], 64)) : u(e).status === "captcha" ? (s(), l(_, { key: 4 }, [
              H,
              o("img", {
                src: u(e).image
              }, null, 8, J),
              o("div", L, [
                a(y, {
                  modelValue: F.value,
                  "onUpdate:modelValue": t[0] || (t[0] = (n) => F.value = n)
                }, null, 8, ["modelValue"]),
                a(B, {
                  onClick: t[1] || (t[1] = (n) => u(m)("gocqhttp/write", u(r), F.value))
                }, {
                  default: p(() => [
                    c("\u63D0\u4EA4")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(e).status === "sms" ? (s(), l(_, { key: 5 }, [
              M,
              o("div", P, [
                a(y, {
                  modelValue: F.value,
                  "onUpdate:modelValue": t[2] || (t[2] = (n) => F.value = n)
                }, null, 8, ["modelValue"]),
                a(B, {
                  onClick: t[3] || (t[3] = (n) => u(m)("gocqhttp/write", u(r), F.value))
                }, {
                  default: p(() => [
                    c("\u63D0\u4EA4")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(e).status === "sms-confirm" ? (s(), l(_, { key: 6 }, [
              o("p", null, "\u8D26\u53F7\u5DF2\u5F00\u542F\u8BBE\u5907\u9501\u3002\u70B9\u51FB\u786E\u8BA4\u5C06\u5411\u624B\u673A " + b(u(e).phone) + " \u53D1\u9001\u77ED\u4FE1\u9A8C\u8BC1\u7801\u3002", 1),
              o("div", W, [
                a(B, {
                  onClick: t[4] || (t[4] = (n) => u(m)("gocqhttp/write", u(r), ""))
                }, {
                  default: p(() => [
                    c("\u786E\u8BA4")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(e).status === "sms-or-qrcode" ? (s(), l(_, { key: 7 }, [
              X,
              o("div", Y, [
                a(B, {
                  onClick: t[5] || (t[5] = (n) => u(m)("gocqhttp/write", u(r), "1"))
                }, {
                  default: p(() => [
                    c("1. \u5411\u624B\u673A " + b(u(e).phone) + " \u53D1\u9001\u77ED\u4FE1\u9A8C\u8BC1\u7801", 1)
                  ]),
                  _: 1
                }),
                a(B, {
                  onClick: t[6] || (t[6] = (n) => u(m)("gocqhttp/write", u(r), "2"))
                }, {
                  default: p(() => [
                    c("2. \u4F7F\u7528\u624B\u673A\u767B\u5F55 QQ \u5E76\u626B\u7801\u9A8C\u8BC1")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(e).status === "slider-or-qrcode" ? (s(), l(_, { key: 8 }, [
              Z,
              o("div", uu, [
                a(B, {
                  disabled: "",
                  onClick: t[7] || (t[7] = (n) => u(m)("gocqhttp/write", u(r), "1"))
                }, {
                  default: p(() => [
                    c("1. \u4F7F\u7528\u6D4F\u89C8\u5668\u6293\u53D6\u6ED1\u6761\u5E76\u767B\u5F55 (\u6682\u4E0D\u652F\u6301)")
                  ]),
                  _: 1
                }),
                a(B, {
                  onClick: t[8] || (t[8] = (n) => u(m)("gocqhttp/write", u(r), "2"))
                }, {
                  default: p(() => [
                    c("2. \u4F7F\u7528\u624B\u673A\u767B\u5F55 QQ \u5E76\u626B\u7801\u9A8C\u8BC1 (\u9700\u8981\u624B\u673A\u548C Koishi \u5728\u540C\u4E00\u7F51\u7EDC\u4E0B)")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(e).status === "slider" ? (s(), l("p", tu, [
              c("\u8D26\u53F7\u5DF2\u5F00\u542F\u8BBE\u5907\u9501\u3002\u8BF7\u70B9\u51FB"),
              o("a", {
                href: u(e).link,
                target: "_blank"
              }, "\u6B64\u94FE\u63A5", 8, eu),
              c("\u9A8C\u8BC1\u540E\u91CD\u542F Bot\u3002")
            ])) : h("", !0)
          ]),
          _: 1
        }, 8, ["type"])) : h("", !0),
        u(r) ? (s(), q($, {
          key: 1,
          modelValue: u(d),
          "onUpdate:modelValue": t[9] || (t[9] = (n) => I(d) ? d.value = n : null),
          initial: u(v).config,
          schema: u(w)
        }, null, 8, ["modelValue", "initial", "schema"])) : h("", !0)
      ], 64);
    };
  }
});
const su = (i, C) => {
  const d = i.__vccOpts || i;
  for (const [v, F] of C)
    d[v] = F;
  return d;
}, nu = /* @__PURE__ */ su(ou, [["__scopeId", "data-v-d9529b45"]]), ru = K((i) => {
  i.slot({
    type: "market-settings",
    component: nu
  });
});
export {
  ru as default
};
