// koishi-plugin-gocqhttp/dist/index.js
import { defineComponent as Q, inject as k, ref as S, computed as D, resolveComponent as f, openBlock as s, createElementBlock as l, Fragment as _, unref as u, createBlock as q, withCtx as p, createElementVNode as o, createVNode as r, createTextVNode as a, toDisplayString as b, createCommentVNode as h, isRef as I, pushScopeId as j, popScopeId as N } from "../vue.js";
import { store as U, Schema as A, send as m, defineExtension as K } from "../client.js";
const E = (i) => (j("data-v-c331b3ef"), i = i(), N(), i), O = { key: 0 }, R = { key: 1 }, T = { key: 2 }, z = /* @__PURE__ */ E(() => /* @__PURE__ */ o("p", null, "\u8BF7\u4F7F\u7528\u624B\u673A\u767B\u5F55 QQ \u626B\u63CF\u4E8C\u7EF4\u7801\uFF1A", -1)), G = ["src"], H = /* @__PURE__ */ E(() => /* @__PURE__ */ o("p", null, "\u8BF7\u586B\u5199\u56FE\u4E2D\u7684\u5185\u5BB9\uFF1A", -1)), J = ["src"], L = { class: "submit" }, M = /* @__PURE__ */ E(() => /* @__PURE__ */ o("p", null, "\u8BF7\u8F93\u5165\u77ED\u4FE1\u9A8C\u8BC1\u7801\uFF1A", -1)), P = { class: "submit" }, W = { class: "submit" }, X = /* @__PURE__ */ E(() => /* @__PURE__ */ o("p", null, "\u8D26\u53F7\u5DF2\u5F00\u542F\u8BBE\u5907\u9501\u3002\u8BF7\u9009\u62E9\u9A8C\u8BC1\u65B9\u5F0F (\u5C06\u5728 10 \u79D2\u540E\u81EA\u52A8\u9009\u62E9 2)\uFF1A", -1)), Y = { class: "submit" }, Z = /* @__PURE__ */ E(() => /* @__PURE__ */ o("p", null, "\u767B\u5F55\u9700\u8981\u6ED1\u6761\u9A8C\u8BC1\u7801\u3002\u8BF7\u9009\u62E9\u9A8C\u8BC1\u65B9\u5F0F (\u5C06\u5728 10 \u79D2\u540E\u81EA\u52A8\u9009\u62E9 1)\uFF1A", -1)), uu = { class: "submit" }, tu = { key: 9 }, eu = ["href"], ou = /* @__PURE__ */ Q({
  __name: "settings",
  setup(i) {
    const g = k("manager.settings.local"), d = k("manager.settings.config"), C = k("manager.settings.current"), F = S(""), c = D(() => {
      if (g.value.name === "@koishijs/plugin-adapter-onebot")
        return `${d.value.platform || "onebot"}:${d.value.selfId}`;
    }), t = D(() => {
      var v;
      return (v = U.gocqhttp) == null ? void 0 : v[c.value];
    }), V = D(() => {
      if (!!t.value && t.value.status !== "init")
        return t.value.status === "offline" ? "error" : t.value.status === "success" ? "success" : "warning";
    }), w = A.object({
      gocqhttp: A.object({
        enabled: A.boolean().default(!1).description("\u662F\u5426\u81EA\u52A8\u521B\u5EFA go-cqhttp \u5B50\u8FDB\u7A0B\u3002"),
        password: A.string().role("secret").description("\u673A\u5668\u4EBA\u7684\u5BC6\u7801\u3002")
      }).description("go-cqhttp \u57FA\u7840\u8BBE\u7F6E")
    });
    return (v, e) => {
      const y = f("el-input"), B = f("k-button"), x = f("k-comment"), $ = f("k-form");
      return s(), l(_, null, [
        u(t) ? (s(), q(x, {
          key: 0,
          class: "gocqhttp",
          type: u(V)
        }, {
          default: p(() => [
            u(t).status === "offline" ? (s(), l("p", O, " \u672A\u8FDE\u63A5\u5230 go-cqhttp \u5B50\u8FDB\u7A0B\u3002 ")) : u(t).status === "init" ? (s(), l("p", R, " \u6B63\u5728\u521B\u5EFA go-cqhttp \u5B50\u8FDB\u7A0B\u2026\u2026 ")) : u(t).status === "success" ? (s(), l("p", T, " \u5DF2\u6210\u529F\u8FDE\u63A5 go-cqhttp \u5B50\u8FDB\u7A0B\u3002 ")) : u(t).status === "qrcode" ? (s(), l(_, { key: 3 }, [
              z,
              o("img", {
                class: "qrcode",
                src: u(t).image
              }, null, 8, G)
            ], 64)) : u(t).status === "captcha" ? (s(), l(_, { key: 4 }, [
              H,
              o("img", {
                src: u(t).image
              }, null, 8, J),
              o("div", L, [
                r(y, {
                  modelValue: F.value,
                  "onUpdate:modelValue": e[0] || (e[0] = (n) => F.value = n)
                }, null, 8, ["modelValue"]),
                r(B, {
                  onClick: e[1] || (e[1] = (n) => u(m)("gocqhttp/write", u(c), F.value))
                }, {
                  default: p(() => [
                    a("\u63D0\u4EA4")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(t).status === "sms" ? (s(), l(_, { key: 5 }, [
              M,
              o("div", P, [
                r(y, {
                  modelValue: F.value,
                  "onUpdate:modelValue": e[2] || (e[2] = (n) => F.value = n)
                }, null, 8, ["modelValue"]),
                r(B, {
                  onClick: e[3] || (e[3] = (n) => u(m)("gocqhttp/write", u(c), F.value))
                }, {
                  default: p(() => [
                    a("\u63D0\u4EA4")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(t).status === "sms-confirm" ? (s(), l(_, { key: 6 }, [
              o("p", null, "\u8D26\u53F7\u5DF2\u5F00\u542F\u8BBE\u5907\u9501\u3002\u70B9\u51FB\u786E\u8BA4\u5C06\u5411\u624B\u673A " + b(u(t).phone) + " \u53D1\u9001\u77ED\u4FE1\u9A8C\u8BC1\u7801\u3002", 1),
              o("div", W, [
                r(B, {
                  onClick: e[4] || (e[4] = (n) => u(m)("gocqhttp/write", u(c), ""))
                }, {
                  default: p(() => [
                    a("\u786E\u8BA4")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(t).status === "sms-or-qrcode" ? (s(), l(_, { key: 7 }, [
              X,
              o("div", Y, [
                r(B, {
                  onClick: e[5] || (e[5] = (n) => u(m)("gocqhttp/write", u(c), "1"))
                }, {
                  default: p(() => [
                    a("1. \u5411\u624B\u673A " + b(u(t).phone) + " \u53D1\u9001\u77ED\u4FE1\u9A8C\u8BC1\u7801", 1)
                  ]),
                  _: 1
                }),
                r(B, {
                  onClick: e[6] || (e[6] = (n) => u(m)("gocqhttp/write", u(c), "2"))
                }, {
                  default: p(() => [
                    a("2. \u4F7F\u7528\u624B\u673A\u767B\u5F55 QQ \u5E76\u626B\u7801\u9A8C\u8BC1")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(t).status === "slider-or-qrcode" ? (s(), l(_, { key: 8 }, [
              Z,
              o("div", uu, [
                r(B, {
                  disabled: "",
                  onClick: e[7] || (e[7] = (n) => u(m)("gocqhttp/write", u(c), "1"))
                }, {
                  default: p(() => [
                    a("1. \u4F7F\u7528\u6D4F\u89C8\u5668\u6293\u53D6\u6ED1\u6761\u5E76\u767B\u5F55 (\u6682\u4E0D\u652F\u6301)")
                  ]),
                  _: 1
                }),
                r(B, {
                  onClick: e[8] || (e[8] = (n) => u(m)("gocqhttp/write", u(c), "2"))
                }, {
                  default: p(() => [
                    a("2. \u4F7F\u7528\u624B\u673A\u767B\u5F55 QQ \u5E76\u626B\u7801\u9A8C\u8BC1 (\u9700\u8981\u624B\u673A\u548C Koishi \u5728\u540C\u4E00\u7F51\u7EDC\u4E0B)")
                  ]),
                  _: 1
                })
              ])
            ], 64)) : u(t).status === "slider" ? (s(), l("p", tu, [
              a("\u8D26\u53F7\u5DF2\u5F00\u542F\u8BBE\u5907\u9501\u3002\u8BF7\u70B9\u51FB"),
              o("a", {
                href: u(t).link,
                target: "_blank"
              }, "\u6B64\u94FE\u63A5", 8, eu),
              a("\u9A8C\u8BC1\u540E\u91CD\u542F Bot\u3002")
            ])) : h("", !0)
          ]),
          _: 1
        }, 8, ["type"])) : h("", !0),
        u(c) ? (s(), q($, {
          key: 1,
          modelValue: u(d),
          "onUpdate:modelValue": e[9] || (e[9] = (n) => I(d) ? d.value = n : null),
          initial: u(C).config,
          schema: u(w)
        }, null, 8, ["modelValue", "initial", "schema"])) : h("", !0)
      ], 64);
    };
  }
});
const su = (i, g) => {
  const d = i.__vccOpts || i;
  for (const [C, F] of g)
    d[C] = F;
  return d;
}, nu = /* @__PURE__ */ su(ou, [["__scopeId", "data-v-c331b3ef"]]), cu = K((i) => {
  i.slot({
    type: "market-settings",
    component: nu
  });
});
export {
  cu as default
};
