// @koishijs/plugin-locales/dist/index.js
import { store as x, send as q, icons as I } from "../client.js";
import { defineComponent as J, ref as j, watch as K, computed as E, resolveComponent as p, openBlock as d, createBlock as A, withCtx as r, createTextVNode as F, toDisplayString as B, createVNode as C, createElementBlock as g, Fragment as V, renderList as O, unref as $, createCommentVNode as Q, createElementVNode as w } from "../vue.js";
function G(l, o, s, n) {
  var c, y = !1, z = 0;
  function H() {
    c && clearTimeout(c);
  }
  function L() {
    H(), y = !0;
  }
  typeof o != "boolean" && (n = s, s = o, o = void 0);
  function b() {
    for (var M = arguments.length, N = new Array(M), k = 0; k < M; k++)
      N[k] = arguments[k];
    var t = this, e = Date.now() - z;
    if (y)
      return;
    function u() {
      z = Date.now(), s.apply(t, N);
    }
    function _() {
      c = void 0;
    }
    n && !c && u(), H(), n === void 0 && e > l ? u() : o !== !0 && (c = setTimeout(n ? _ : u, n === void 0 ? l - e : l));
  }
  return b.cancel = L, b;
}
function R(l, o, s) {
  return s === void 0 ? G(l, o, !1) : G(l, s, o !== !1);
}
const X = { class: "menu-item" }, Y = { class: "search" }, Z = { class: "locale" }, ee = /* @__PURE__ */ J({
  __name: "locales",
  setup(l) {
    const o = j(["zh", "en"]), s = j(null), n = j(""), c = j("");
    K(c, (t) => {
      s.value.filter(t);
    });
    function y(t, e) {
      return e.label.includes(c.value);
    }
    function z(t) {
      const e = [];
      return t.id === n.value && e.push("is-active"), e.join(" ");
    }
    function H(t) {
      n.value = t.id;
    }
    const L = E(() => {
      const t = {};
      for (const e in x.locales)
        !e || Object.assign(t, x.locales[e]);
      return Object.keys(t).filter((e) => !e.includes("._") && !e.includes("@"));
    }), b = E(() => L.value.filter((t) => t.startsWith(n.value + ".")).map((t) => t.slice(n.value.length + 1)).filter((t) => n.value.split(".").length >= 2 || !t.includes("."))), M = E(() => {
      var e;
      const t = [];
      for (const u of L.value) {
        const _ = u.split(".");
        let m = t;
        for (let f = 0; f < Math.min(_.length - 1, 2); f++) {
          const v = _[f];
          if (!v)
            break;
          const U = _.slice(0, f + 1).join(".");
          let h = m.find((D) => D.id === U);
          h || (h = { id: U, label: v }, m.push(h)), m = (e = h.children) != null ? e : h.children = [];
        }
      }
      return t;
    }), N = R(1e3, () => {
      const t = {};
      for (const e in x.locales)
        !e.startsWith("$") || (t[e.slice(1)] = x.locales[e]);
      q("l10n", t);
    });
    function k(t, e, u) {
      var m, f, v;
      const _ = (v = (m = x.locales)[f = "$" + t]) != null ? v : m[f] = {};
      _[`${n.value}.${e}`] = u, N();
    }
    return (t, e) => {
      const u = p("k-icon"), _ = p("el-checkbox"), m = p("el-checkbox-group"), f = p("el-dropdown"), v = p("el-input"), U = p("el-tree"), h = p("el-scrollbar"), D = p("k-content"), S = p("k-layout");
      return d(), A(S, { class: "page-locales" }, {
        header: r(() => [
          F(" \u672C\u5730\u5316 - " + B(n.value), 1)
        ]),
        menu: r(() => [
          C(f, {
            placement: "bottom",
            "popper-class": "k-dropdown"
          }, {
            dropdown: r(() => [
              C(m, {
                modelValue: o.value,
                "onUpdate:modelValue": e[0] || (e[0] = (i) => o.value = i)
              }, {
                default: r(() => [
                  (d(!0), g(V, null, O($(x).locales, (i, a) => (d(), g(V, { key: a }, [
                    a && !a.startsWith("$") ? (d(), A(_, {
                      key: 0,
                      label: a
                    }, {
                      default: r(() => [
                        F(B(a), 1)
                      ]),
                      _: 2
                    }, 1032, ["label"])) : Q("", !0)
                  ], 64))), 128))
                ]),
                _: 1
              }, 8, ["modelValue"])
            ]),
            default: r(() => [
              w("span", X, [
                C(u, {
                  class: "menu-icon",
                  name: "globe"
                })
              ])
            ]),
            _: 1
          })
        ]),
        left: r(() => [
          C(h, null, {
            default: r(() => [
              w("div", Y, [
                C(v, {
                  modelValue: c.value,
                  "onUpdate:modelValue": e[1] || (e[1] = (i) => c.value = i)
                }, {
                  suffix: r(() => [
                    C(u, { name: "search" })
                  ]),
                  _: 1
                }, 8, ["modelValue"])
              ]),
              C(U, {
                ref_key: "tree",
                ref: s,
                data: $(M),
                props: { class: z },
                "filter-node-method": y,
                "default-expand-all": !0,
                "expand-on-click-node": !1,
                onNodeClick: H
              }, null, 8, ["data", "props"])
            ]),
            _: 1
          })
        ]),
        default: r(() => [
          C(D, null, {
            default: r(() => [
              (d(!0), g(V, null, O($(b), (i) => (d(), g(V, { key: i }, [
                w("h3", null, B(i), 1),
                (d(!0), g(V, null, O(o.value, (a) => {
                  var W;
                  return d(), g("div", {
                    class: "translation",
                    key: a
                  }, [
                    w("span", Z, B(a), 1),
                    C(v, {
                      autosize: "",
                      type: "textarea",
                      modelValue: (W = $(x).locales["$" + a]) == null ? void 0 : W[`${n.value}.${i}`],
                      placeholder: $(x).locales[a][`${n.value}.${i}`],
                      "onUpdate:modelValue": (T) => k(a, i, T)
                    }, null, 8, ["modelValue", "placeholder", "onUpdate:modelValue"])
                  ]);
                }), 128))
              ], 64))), 128))
            ]),
            _: 1
          })
        ]),
        _: 1
      });
    };
  }
});
const P = (l, o) => {
  const s = l.__vccOpts || l;
  for (const [n, c] of o)
    s[n] = c;
  return s;
}, te = {}, ne = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, oe = /* @__PURE__ */ w("path", {
  fill: "currentColor",
  d: "M386.7 22.63C411.7-2.365 452.3-2.365 477.3 22.63L489.4 34.74C514.4 59.74 514.4 100.3 489.4 125.3L269 345.6C260.6 354.1 249.9 359.1 238.2 362.7L147.6 383.6C142.2 384.8 136.6 383.2 132.7 379.3C128.8 375.4 127.2 369.8 128.4 364.4L149.3 273.8C152 262.1 157.9 251.4 166.4 242.1L386.7 22.63zM454.6 45.26C442.1 32.76 421.9 32.76 409.4 45.26L382.6 72L440 129.4L466.7 102.6C479.2 90.13 479.2 69.87 466.7 57.37L454.6 45.26zM180.5 281L165.3 346.7L230.1 331.5C236.8 330.2 242.2 327.2 246.4 322.1L417.4 152L360 94.63L189 265.6C184.8 269.8 181.8 275.2 180.5 281V281zM208 64C216.8 64 224 71.16 224 80C224 88.84 216.8 96 208 96H80C53.49 96 32 117.5 32 144V432C32 458.5 53.49 480 80 480H368C394.5 480 416 458.5 416 432V304C416 295.2 423.2 288 432 288C440.8 288 448 295.2 448 304V432C448 476.2 412.2 512 368 512H80C35.82 512 0 476.2 0 432V144C0 99.82 35.82 64 80 64H208z"
}, null, -1), le = [
  oe
];
function se(l, o) {
  return d(), g("svg", ne, le);
}
const ce = /* @__PURE__ */ P(te, [["render", se]]), ae = {}, re = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, ue = /* @__PURE__ */ w("path", {
  fill: "currentColor",
  d: "M256 0C397.4 0 512 114.6 512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256C0 114.6 114.6 0 256 0zM256 464C263.4 464 282.1 456.8 303.6 415.6C312.4 397.9 319.1 376.4 325.6 352H186.4C192 376.4 199.6 397.9 208.4 415.6C229 456.8 248.6 464 256 464zM178.5 304H333.5C335.1 288.7 336 272.6 336 256C336 239.4 335.1 223.3 333.5 208H178.5C176.9 223.3 176 239.4 176 256C176 272.6 176.9 288.7 178.5 304V304zM325.6 160C319.1 135.6 312.4 114.1 303.6 96.45C282.1 55.22 263.4 48 256 48C248.6 48 229 55.22 208.4 96.45C199.6 114.1 192 135.6 186.4 160H325.6zM381.8 208C383.2 223.5 384 239.6 384 256C384 272.4 383.2 288.5 381.8 304H458.4C462.1 288.6 464 272.5 464 256C464 239.5 462.1 223.4 458.4 208H381.8zM342.1 66.61C356.2 92.26 367.4 124.1 374.7 160H440.6C419.2 118.9 384.4 85.88 342.1 66.61zM169.9 66.61C127.6 85.88 92.84 118.9 71.43 160H137.3C144.6 124.1 155.8 92.26 169.9 66.61V66.61zM48 256C48 272.5 49.93 288.6 53.57 304H130.2C128.8 288.5 128 272.4 128 256C128 239.6 128.8 223.5 130.2 208H53.57C49.93 223.4 48 239.5 48 256zM440.6 352H374.7C367.4 387.9 356.2 419.7 342.1 445.4C384.4 426.1 419.2 393.1 440.6 352zM137.3 352H71.43C92.84 393.1 127.6 426.1 169.9 445.4C155.8 419.7 144.6 387.9 137.3 352V352z"
}, null, -1), ie = [
  ue
];
function de(l, o) {
  return d(), g("svg", re, ie);
}
const _e = /* @__PURE__ */ P(ae, [["render", de]]);
I.register("activity:locales", ce);
I.register("globe", _e);
const Ce = (l) => {
  l.page({
    path: "/locales",
    name: "\u672C\u5730\u5316",
    icon: "activity:locales",
    order: 500,
    authority: 4,
    fields: ["locales"],
    component: ee
  });
};
export {
  Ce as default
};
