// @koishijs/plugin-market/dist/index.js
import { ref as ie, watch as pe, computed as q, defineComponent as ne, resolveComponent as M, openBlock as E, createBlock as T, unref as w, isRef as de, createSlots as Xr, withCtx as V, createElementBlock as R, createElementVNode as $, Fragment as j, renderList as te, toDisplayString as H, normalizeClass as ge, createVNode as G, createTextVNode as O, withModifiers as he, createCommentVNode as Q, pushScopeId as Wr, popScopeId as Yr, onActivated as Kr, nextTick as Zr, provide as Se, withDirectives as Sr, vModelText as Vr, reactive as Qr, withKeys as we, h as Ve } from "../vue.js";
import { store as k, valueMap as Bt, loading as Jr, socket as en, message as me, send as ue, createStorage as tn, router as fe, clone as xr, Schema as $e, config as rn, global as xe, icons as K, receive as nn } from "../client.js";
import { useRoute as Nr, useRouter as sn } from "../vue-router.js";
import { useWindowSize as un } from "../vueuse.js";
var Fe = { exports: {} }, Ne, Ot;
function Re() {
  if (Ot)
    return Ne;
  Ot = 1;
  const n = "2.0.0", h = 256, l = Number.MAX_SAFE_INTEGER || 9007199254740991;
  return Ne = {
    SEMVER_SPEC_VERSION: n,
    MAX_LENGTH: h,
    MAX_SAFE_INTEGER: l,
    MAX_SAFE_COMPONENT_LENGTH: 16
  }, Ne;
}
var Be, Tt;
function Ae() {
  return Tt || (Tt = 1, Be = typeof process == "object" && process.env && process.env.NODE_DEBUG && /\bsemver\b/i.test(process.env.NODE_DEBUG) ? (...h) => console.error("SEMVER", ...h) : () => {
  }), Be;
}
var Mt;
function Ce() {
  return Mt || (Mt = 1, function(n, h) {
    const { MAX_SAFE_COMPONENT_LENGTH: l } = Re(), o = Ae();
    h = n.exports = {};
    const a = h.re = [], e = h.src = [], r = h.t = {};
    let c = 0;
    const f = (p, d, _) => {
      const g = c++;
      o(p, g, d), r[p] = g, e[g] = d, a[g] = new RegExp(d, _ ? "g" : void 0);
    };
    f("NUMERICIDENTIFIER", "0|[1-9]\\d*"), f("NUMERICIDENTIFIERLOOSE", "[0-9]+"), f("NONNUMERICIDENTIFIER", "\\d*[a-zA-Z-][a-zA-Z0-9-]*"), f("MAINVERSION", `(${e[r.NUMERICIDENTIFIER]})\\.(${e[r.NUMERICIDENTIFIER]})\\.(${e[r.NUMERICIDENTIFIER]})`), f("MAINVERSIONLOOSE", `(${e[r.NUMERICIDENTIFIERLOOSE]})\\.(${e[r.NUMERICIDENTIFIERLOOSE]})\\.(${e[r.NUMERICIDENTIFIERLOOSE]})`), f("PRERELEASEIDENTIFIER", `(?:${e[r.NUMERICIDENTIFIER]}|${e[r.NONNUMERICIDENTIFIER]})`), f("PRERELEASEIDENTIFIERLOOSE", `(?:${e[r.NUMERICIDENTIFIERLOOSE]}|${e[r.NONNUMERICIDENTIFIER]})`), f("PRERELEASE", `(?:-(${e[r.PRERELEASEIDENTIFIER]}(?:\\.${e[r.PRERELEASEIDENTIFIER]})*))`), f("PRERELEASELOOSE", `(?:-?(${e[r.PRERELEASEIDENTIFIERLOOSE]}(?:\\.${e[r.PRERELEASEIDENTIFIERLOOSE]})*))`), f("BUILDIDENTIFIER", "[0-9A-Za-z-]+"), f("BUILD", `(?:\\+(${e[r.BUILDIDENTIFIER]}(?:\\.${e[r.BUILDIDENTIFIER]})*))`), f("FULLPLAIN", `v?${e[r.MAINVERSION]}${e[r.PRERELEASE]}?${e[r.BUILD]}?`), f("FULL", `^${e[r.FULLPLAIN]}$`), f("LOOSEPLAIN", `[v=\\s]*${e[r.MAINVERSIONLOOSE]}${e[r.PRERELEASELOOSE]}?${e[r.BUILD]}?`), f("LOOSE", `^${e[r.LOOSEPLAIN]}$`), f("GTLT", "((?:<|>)?=?)"), f("XRANGEIDENTIFIERLOOSE", `${e[r.NUMERICIDENTIFIERLOOSE]}|x|X|\\*`), f("XRANGEIDENTIFIER", `${e[r.NUMERICIDENTIFIER]}|x|X|\\*`), f("XRANGEPLAIN", `[v=\\s]*(${e[r.XRANGEIDENTIFIER]})(?:\\.(${e[r.XRANGEIDENTIFIER]})(?:\\.(${e[r.XRANGEIDENTIFIER]})(?:${e[r.PRERELEASE]})?${e[r.BUILD]}?)?)?`), f("XRANGEPLAINLOOSE", `[v=\\s]*(${e[r.XRANGEIDENTIFIERLOOSE]})(?:\\.(${e[r.XRANGEIDENTIFIERLOOSE]})(?:\\.(${e[r.XRANGEIDENTIFIERLOOSE]})(?:${e[r.PRERELEASELOOSE]})?${e[r.BUILD]}?)?)?`), f("XRANGE", `^${e[r.GTLT]}\\s*${e[r.XRANGEPLAIN]}$`), f("XRANGELOOSE", `^${e[r.GTLT]}\\s*${e[r.XRANGEPLAINLOOSE]}$`), f("COERCE", `(^|[^\\d])(\\d{1,${l}})(?:\\.(\\d{1,${l}}))?(?:\\.(\\d{1,${l}}))?(?:$|[^\\d])`), f("COERCERTL", e[r.COERCE], !0), f("LONETILDE", "(?:~>?)"), f("TILDETRIM", `(\\s*)${e[r.LONETILDE]}\\s+`, !0), h.tildeTrimReplace = "$1~", f("TILDE", `^${e[r.LONETILDE]}${e[r.XRANGEPLAIN]}$`), f("TILDELOOSE", `^${e[r.LONETILDE]}${e[r.XRANGEPLAINLOOSE]}$`), f("LONECARET", "(?:\\^)"), f("CARETTRIM", `(\\s*)${e[r.LONECARET]}\\s+`, !0), h.caretTrimReplace = "$1^", f("CARET", `^${e[r.LONECARET]}${e[r.XRANGEPLAIN]}$`), f("CARETLOOSE", `^${e[r.LONECARET]}${e[r.XRANGEPLAINLOOSE]}$`), f("COMPARATORLOOSE", `^${e[r.GTLT]}\\s*(${e[r.LOOSEPLAIN]})$|^$`), f("COMPARATOR", `^${e[r.GTLT]}\\s*(${e[r.FULLPLAIN]})$|^$`), f("COMPARATORTRIM", `(\\s*)${e[r.GTLT]}\\s*(${e[r.LOOSEPLAIN]}|${e[r.XRANGEPLAIN]})`, !0), h.comparatorTrimReplace = "$1$2$3", f("HYPHENRANGE", `^\\s*(${e[r.XRANGEPLAIN]})\\s+-\\s+(${e[r.XRANGEPLAIN]})\\s*$`), f("HYPHENRANGELOOSE", `^\\s*(${e[r.XRANGEPLAINLOOSE]})\\s+-\\s+(${e[r.XRANGEPLAINLOOSE]})\\s*$`), f("STAR", "(<|>)?=?\\s*\\*"), f("GTE0", "^\\s*>=\\s*0\\.0\\.0\\s*$"), f("GTE0PRE", "^\\s*>=\\s*0\\.0\\.0-0\\s*$");
  }(Fe, Fe.exports)), Fe.exports;
}
var Oe, qt;
function De() {
  if (qt)
    return Oe;
  qt = 1;
  const n = ["includePrerelease", "loose", "rtl"];
  return Oe = (l) => l ? typeof l != "object" ? { loose: !0 } : n.filter((o) => l[o]).reduce((o, a) => (o[a] = !0, o), {}) : {}, Oe;
}
var Te, zt;
function Br() {
  if (zt)
    return Te;
  zt = 1;
  const n = /^[0-9]+$/, h = (o, a) => {
    const e = n.test(o), r = n.test(a);
    return e && r && (o = +o, a = +a), o === a ? 0 : e && !r ? -1 : r && !e ? 1 : o < a ? -1 : 1;
  };
  return Te = {
    compareIdentifiers: h,
    rcompareIdentifiers: (o, a) => h(a, o)
  }, Te;
}
var Me, Ht;
function se() {
  if (Ht)
    return Me;
  Ht = 1;
  const n = Ae(), { MAX_LENGTH: h, MAX_SAFE_INTEGER: l } = Re(), { re: o, t: a } = Ce(), e = De(), { compareIdentifiers: r } = Br();
  class c {
    constructor(p, d) {
      if (d = e(d), p instanceof c) {
        if (p.loose === !!d.loose && p.includePrerelease === !!d.includePrerelease)
          return p;
        p = p.version;
      } else if (typeof p != "string")
        throw new TypeError(`Invalid Version: ${p}`);
      if (p.length > h)
        throw new TypeError(
          `version is longer than ${h} characters`
        );
      n("SemVer", p, d), this.options = d, this.loose = !!d.loose, this.includePrerelease = !!d.includePrerelease;
      const _ = p.trim().match(d.loose ? o[a.LOOSE] : o[a.FULL]);
      if (!_)
        throw new TypeError(`Invalid Version: ${p}`);
      if (this.raw = p, this.major = +_[1], this.minor = +_[2], this.patch = +_[3], this.major > l || this.major < 0)
        throw new TypeError("Invalid major version");
      if (this.minor > l || this.minor < 0)
        throw new TypeError("Invalid minor version");
      if (this.patch > l || this.patch < 0)
        throw new TypeError("Invalid patch version");
      _[4] ? this.prerelease = _[4].split(".").map((g) => {
        if (/^[0-9]+$/.test(g)) {
          const y = +g;
          if (y >= 0 && y < l)
            return y;
        }
        return g;
      }) : this.prerelease = [], this.build = _[5] ? _[5].split(".") : [], this.format();
    }
    format() {
      return this.version = `${this.major}.${this.minor}.${this.patch}`, this.prerelease.length && (this.version += `-${this.prerelease.join(".")}`), this.version;
    }
    toString() {
      return this.version;
    }
    compare(p) {
      if (n("SemVer.compare", this.version, this.options, p), !(p instanceof c)) {
        if (typeof p == "string" && p === this.version)
          return 0;
        p = new c(p, this.options);
      }
      return p.version === this.version ? 0 : this.compareMain(p) || this.comparePre(p);
    }
    compareMain(p) {
      return p instanceof c || (p = new c(p, this.options)), r(this.major, p.major) || r(this.minor, p.minor) || r(this.patch, p.patch);
    }
    comparePre(p) {
      if (p instanceof c || (p = new c(p, this.options)), this.prerelease.length && !p.prerelease.length)
        return -1;
      if (!this.prerelease.length && p.prerelease.length)
        return 1;
      if (!this.prerelease.length && !p.prerelease.length)
        return 0;
      let d = 0;
      do {
        const _ = this.prerelease[d], g = p.prerelease[d];
        if (n("prerelease compare", d, _, g), _ === void 0 && g === void 0)
          return 0;
        if (g === void 0)
          return 1;
        if (_ === void 0)
          return -1;
        if (_ === g)
          continue;
        return r(_, g);
      } while (++d);
    }
    compareBuild(p) {
      p instanceof c || (p = new c(p, this.options));
      let d = 0;
      do {
        const _ = this.build[d], g = p.build[d];
        if (n("prerelease compare", d, _, g), _ === void 0 && g === void 0)
          return 0;
        if (g === void 0)
          return 1;
        if (_ === void 0)
          return -1;
        if (_ === g)
          continue;
        return r(_, g);
      } while (++d);
    }
    inc(p, d) {
      switch (p) {
        case "premajor":
          this.prerelease.length = 0, this.patch = 0, this.minor = 0, this.major++, this.inc("pre", d);
          break;
        case "preminor":
          this.prerelease.length = 0, this.patch = 0, this.minor++, this.inc("pre", d);
          break;
        case "prepatch":
          this.prerelease.length = 0, this.inc("patch", d), this.inc("pre", d);
          break;
        case "prerelease":
          this.prerelease.length === 0 && this.inc("patch", d), this.inc("pre", d);
          break;
        case "major":
          (this.minor !== 0 || this.patch !== 0 || this.prerelease.length === 0) && this.major++, this.minor = 0, this.patch = 0, this.prerelease = [];
          break;
        case "minor":
          (this.patch !== 0 || this.prerelease.length === 0) && this.minor++, this.patch = 0, this.prerelease = [];
          break;
        case "patch":
          this.prerelease.length === 0 && this.patch++, this.prerelease = [];
          break;
        case "pre":
          if (this.prerelease.length === 0)
            this.prerelease = [0];
          else {
            let _ = this.prerelease.length;
            for (; --_ >= 0; )
              typeof this.prerelease[_] == "number" && (this.prerelease[_]++, _ = -2);
            _ === -1 && this.prerelease.push(0);
          }
          d && (r(this.prerelease[0], d) === 0 ? isNaN(this.prerelease[1]) && (this.prerelease = [d, 0]) : this.prerelease = [d, 0]);
          break;
        default:
          throw new Error(`invalid increment argument: ${p}`);
      }
      return this.format(), this.raw = this.version, this;
    }
  }
  return Me = c, Me;
}
var qe, Pt;
function Ee() {
  if (Pt)
    return qe;
  Pt = 1;
  const { MAX_LENGTH: n } = Re(), { re: h, t: l } = Ce(), o = se(), a = De();
  return qe = (r, c) => {
    if (c = a(c), r instanceof o)
      return r;
    if (typeof r != "string" || r.length > n || !(c.loose ? h[l.LOOSE] : h[l.FULL]).test(r))
      return null;
    try {
      return new o(r, c);
    } catch {
      return null;
    }
  }, qe;
}
var ze, jt;
function on() {
  if (jt)
    return ze;
  jt = 1;
  const n = Ee();
  return ze = (l, o) => {
    const a = n(l, o);
    return a ? a.version : null;
  }, ze;
}
var He, Gt;
function an() {
  if (Gt)
    return He;
  Gt = 1;
  const n = Ee();
  return He = (l, o) => {
    const a = n(l.trim().replace(/^[=v]+/, ""), o);
    return a ? a.version : null;
  }, He;
}
var Pe, Ut;
function ln() {
  if (Ut)
    return Pe;
  Ut = 1;
  const n = se();
  return Pe = (l, o, a, e) => {
    typeof a == "string" && (e = a, a = void 0);
    try {
      return new n(
        l instanceof n ? l.version : l,
        a
      ).inc(o, e).version;
    } catch {
      return null;
    }
  }, Pe;
}
var je, Xt;
function ae() {
  if (Xt)
    return je;
  Xt = 1;
  const n = se();
  return je = (l, o, a) => new n(l, a).compare(new n(o, a)), je;
}
var Ge, Wt;
function bt() {
  if (Wt)
    return Ge;
  Wt = 1;
  const n = ae();
  return Ge = (l, o, a) => n(l, o, a) === 0, Ge;
}
var Ue, Yt;
function cn() {
  if (Yt)
    return Ue;
  Yt = 1;
  const n = Ee(), h = bt();
  return Ue = (o, a) => {
    if (h(o, a))
      return null;
    {
      const e = n(o), r = n(a), c = e.prerelease.length || r.prerelease.length, f = c ? "pre" : "", p = c ? "prerelease" : "";
      for (const d in e)
        if ((d === "major" || d === "minor" || d === "patch") && e[d] !== r[d])
          return f + d;
      return p;
    }
  }, Ue;
}
var Xe, Kt;
function hn() {
  if (Kt)
    return Xe;
  Kt = 1;
  const n = se();
  return Xe = (l, o) => new n(l, o).major, Xe;
}
var We, Zt;
function fn() {
  if (Zt)
    return We;
  Zt = 1;
  const n = se();
  return We = (l, o) => new n(l, o).minor, We;
}
var Ye, Qt;
function dn() {
  if (Qt)
    return Ye;
  Qt = 1;
  const n = se();
  return Ye = (l, o) => new n(l, o).patch, Ye;
}
var Ke, Jt;
function pn() {
  if (Jt)
    return Ke;
  Jt = 1;
  const n = Ee();
  return Ke = (l, o) => {
    const a = n(l, o);
    return a && a.prerelease.length ? a.prerelease : null;
  }, Ke;
}
var Ze, er;
function _n() {
  if (er)
    return Ze;
  er = 1;
  const n = ae();
  return Ze = (l, o, a) => n(o, l, a), Ze;
}
var Qe, tr;
function vn() {
  if (tr)
    return Qe;
  tr = 1;
  const n = ae();
  return Qe = (l, o) => n(l, o, !0), Qe;
}
var Je, rr;
function kt() {
  if (rr)
    return Je;
  rr = 1;
  const n = se();
  return Je = (l, o, a) => {
    const e = new n(l, a), r = new n(o, a);
    return e.compare(r) || e.compareBuild(r);
  }, Je;
}
var et, nr;
function mn() {
  if (nr)
    return et;
  nr = 1;
  const n = kt();
  return et = (l, o) => l.sort((a, e) => n(a, e, o)), et;
}
var tt, sr;
function gn() {
  if (sr)
    return tt;
  sr = 1;
  const n = kt();
  return tt = (l, o) => l.sort((a, e) => n(e, a, o)), tt;
}
var rt, ur;
function be() {
  if (ur)
    return rt;
  ur = 1;
  const n = ae();
  return rt = (l, o, a) => n(l, o, a) > 0, rt;
}
var nt, ir;
function It() {
  if (ir)
    return nt;
  ir = 1;
  const n = ae();
  return nt = (l, o, a) => n(l, o, a) < 0, nt;
}
var st, or;
function Or() {
  if (or)
    return st;
  or = 1;
  const n = ae();
  return st = (l, o, a) => n(l, o, a) !== 0, st;
}
var ut, ar;
function St() {
  if (ar)
    return ut;
  ar = 1;
  const n = ae();
  return ut = (l, o, a) => n(l, o, a) >= 0, ut;
}
var it, lr;
function Vt() {
  if (lr)
    return it;
  lr = 1;
  const n = ae();
  return it = (l, o, a) => n(l, o, a) <= 0, it;
}
var ot, cr;
function Tr() {
  if (cr)
    return ot;
  cr = 1;
  const n = bt(), h = Or(), l = be(), o = St(), a = It(), e = Vt();
  return ot = (c, f, p, d) => {
    switch (f) {
      case "===":
        return typeof c == "object" && (c = c.version), typeof p == "object" && (p = p.version), c === p;
      case "!==":
        return typeof c == "object" && (c = c.version), typeof p == "object" && (p = p.version), c !== p;
      case "":
      case "=":
      case "==":
        return n(c, p, d);
      case "!=":
        return h(c, p, d);
      case ">":
        return l(c, p, d);
      case ">=":
        return o(c, p, d);
      case "<":
        return a(c, p, d);
      case "<=":
        return e(c, p, d);
      default:
        throw new TypeError(`Invalid operator: ${f}`);
    }
  }, ot;
}
var at, hr;
function Cn() {
  if (hr)
    return at;
  hr = 1;
  const n = se(), h = Ee(), { re: l, t: o } = Ce();
  return at = (e, r) => {
    if (e instanceof n)
      return e;
    if (typeof e == "number" && (e = String(e)), typeof e != "string")
      return null;
    r = r || {};
    let c = null;
    if (!r.rtl)
      c = e.match(l[o.COERCE]);
    else {
      let f;
      for (; (f = l[o.COERCERTL].exec(e)) && (!c || c.index + c[0].length !== e.length); )
        (!c || f.index + f[0].length !== c.index + c[0].length) && (c = f), l[o.COERCERTL].lastIndex = f.index + f[1].length + f[2].length;
      l[o.COERCERTL].lastIndex = -1;
    }
    return c === null ? null : h(`${c[2]}.${c[3] || "0"}.${c[4] || "0"}`, r);
  }, at;
}
var lt, fr;
function En() {
  return fr || (fr = 1, lt = function(n) {
    n.prototype[Symbol.iterator] = function* () {
      for (let h = this.head; h; h = h.next)
        yield h.value;
    };
  }), lt;
}
var ct, dr;
function $n() {
  if (dr)
    return ct;
  dr = 1, ct = n, n.Node = a, n.create = n;
  function n(e) {
    var r = this;
    if (r instanceof n || (r = new n()), r.tail = null, r.head = null, r.length = 0, e && typeof e.forEach == "function")
      e.forEach(function(p) {
        r.push(p);
      });
    else if (arguments.length > 0)
      for (var c = 0, f = arguments.length; c < f; c++)
        r.push(arguments[c]);
    return r;
  }
  n.prototype.removeNode = function(e) {
    if (e.list !== this)
      throw new Error("removing node which does not belong to this list");
    var r = e.next, c = e.prev;
    return r && (r.prev = c), c && (c.next = r), e === this.head && (this.head = r), e === this.tail && (this.tail = c), e.list.length--, e.next = null, e.prev = null, e.list = null, r;
  }, n.prototype.unshiftNode = function(e) {
    if (e !== this.head) {
      e.list && e.list.removeNode(e);
      var r = this.head;
      e.list = this, e.next = r, r && (r.prev = e), this.head = e, this.tail || (this.tail = e), this.length++;
    }
  }, n.prototype.pushNode = function(e) {
    if (e !== this.tail) {
      e.list && e.list.removeNode(e);
      var r = this.tail;
      e.list = this, e.prev = r, r && (r.next = e), this.tail = e, this.head || (this.head = e), this.length++;
    }
  }, n.prototype.push = function() {
    for (var e = 0, r = arguments.length; e < r; e++)
      l(this, arguments[e]);
    return this.length;
  }, n.prototype.unshift = function() {
    for (var e = 0, r = arguments.length; e < r; e++)
      o(this, arguments[e]);
    return this.length;
  }, n.prototype.pop = function() {
    if (!!this.tail) {
      var e = this.tail.value;
      return this.tail = this.tail.prev, this.tail ? this.tail.next = null : this.head = null, this.length--, e;
    }
  }, n.prototype.shift = function() {
    if (!!this.head) {
      var e = this.head.value;
      return this.head = this.head.next, this.head ? this.head.prev = null : this.tail = null, this.length--, e;
    }
  }, n.prototype.forEach = function(e, r) {
    r = r || this;
    for (var c = this.head, f = 0; c !== null; f++)
      e.call(r, c.value, f, this), c = c.next;
  }, n.prototype.forEachReverse = function(e, r) {
    r = r || this;
    for (var c = this.tail, f = this.length - 1; c !== null; f--)
      e.call(r, c.value, f, this), c = c.prev;
  }, n.prototype.get = function(e) {
    for (var r = 0, c = this.head; c !== null && r < e; r++)
      c = c.next;
    if (r === e && c !== null)
      return c.value;
  }, n.prototype.getReverse = function(e) {
    for (var r = 0, c = this.tail; c !== null && r < e; r++)
      c = c.prev;
    if (r === e && c !== null)
      return c.value;
  }, n.prototype.map = function(e, r) {
    r = r || this;
    for (var c = new n(), f = this.head; f !== null; )
      c.push(e.call(r, f.value, this)), f = f.next;
    return c;
  }, n.prototype.mapReverse = function(e, r) {
    r = r || this;
    for (var c = new n(), f = this.tail; f !== null; )
      c.push(e.call(r, f.value, this)), f = f.prev;
    return c;
  }, n.prototype.reduce = function(e, r) {
    var c, f = this.head;
    if (arguments.length > 1)
      c = r;
    else if (this.head)
      f = this.head.next, c = this.head.value;
    else
      throw new TypeError("Reduce of empty list with no initial value");
    for (var p = 0; f !== null; p++)
      c = e(c, f.value, p), f = f.next;
    return c;
  }, n.prototype.reduceReverse = function(e, r) {
    var c, f = this.tail;
    if (arguments.length > 1)
      c = r;
    else if (this.tail)
      f = this.tail.prev, c = this.tail.value;
    else
      throw new TypeError("Reduce of empty list with no initial value");
    for (var p = this.length - 1; f !== null; p--)
      c = e(c, f.value, p), f = f.prev;
    return c;
  }, n.prototype.toArray = function() {
    for (var e = new Array(this.length), r = 0, c = this.head; c !== null; r++)
      e[r] = c.value, c = c.next;
    return e;
  }, n.prototype.toArrayReverse = function() {
    for (var e = new Array(this.length), r = 0, c = this.tail; c !== null; r++)
      e[r] = c.value, c = c.prev;
    return e;
  }, n.prototype.slice = function(e, r) {
    r = r || this.length, r < 0 && (r += this.length), e = e || 0, e < 0 && (e += this.length);
    var c = new n();
    if (r < e || r < 0)
      return c;
    e < 0 && (e = 0), r > this.length && (r = this.length);
    for (var f = 0, p = this.head; p !== null && f < e; f++)
      p = p.next;
    for (; p !== null && f < r; f++, p = p.next)
      c.push(p.value);
    return c;
  }, n.prototype.sliceReverse = function(e, r) {
    r = r || this.length, r < 0 && (r += this.length), e = e || 0, e < 0 && (e += this.length);
    var c = new n();
    if (r < e || r < 0)
      return c;
    e < 0 && (e = 0), r > this.length && (r = this.length);
    for (var f = this.length, p = this.tail; p !== null && f > r; f--)
      p = p.prev;
    for (; p !== null && f > e; f--, p = p.prev)
      c.push(p.value);
    return c;
  }, n.prototype.splice = function(e, r, ...c) {
    e > this.length && (e = this.length - 1), e < 0 && (e = this.length + e);
    for (var f = 0, p = this.head; p !== null && f < e; f++)
      p = p.next;
    for (var d = [], f = 0; p && f < r; f++)
      d.push(p.value), p = this.removeNode(p);
    p === null && (p = this.tail), p !== this.head && p !== this.tail && (p = p.prev);
    for (var f = 0; f < c.length; f++)
      p = h(this, p, c[f]);
    return d;
  }, n.prototype.reverse = function() {
    for (var e = this.head, r = this.tail, c = e; c !== null; c = c.prev) {
      var f = c.prev;
      c.prev = c.next, c.next = f;
    }
    return this.head = r, this.tail = e, this;
  };
  function h(e, r, c) {
    var f = r === e.head ? new a(c, null, r, e) : new a(c, r, r.next, e);
    return f.next === null && (e.tail = f), f.prev === null && (e.head = f), e.length++, f;
  }
  function l(e, r) {
    e.tail = new a(r, e.tail, null, e), e.head || (e.head = e.tail), e.length++;
  }
  function o(e, r) {
    e.head = new a(r, null, e.head, e), e.tail || (e.tail = e.head), e.length++;
  }
  function a(e, r, c, f) {
    if (!(this instanceof a))
      return new a(e, r, c, f);
    this.list = f, this.value = e, r ? (r.next = this, this.prev = r) : this.prev = null, c ? (c.prev = this, this.next = c) : this.next = null;
  }
  try {
    En()(n);
  } catch {
  }
  return ct;
}
var ht, pr;
function yn() {
  if (pr)
    return ht;
  pr = 1;
  const n = $n(), h = Symbol("max"), l = Symbol("length"), o = Symbol("lengthCalculator"), a = Symbol("allowStale"), e = Symbol("maxAge"), r = Symbol("dispose"), c = Symbol("noDisposeOnSet"), f = Symbol("lruList"), p = Symbol("cache"), d = Symbol("updateAgeOnGet"), _ = () => 1;
  class g {
    constructor(t) {
      if (typeof t == "number" && (t = { max: t }), t || (t = {}), t.max && (typeof t.max != "number" || t.max < 0))
        throw new TypeError("max must be a non-negative number");
      this[h] = t.max || 1 / 0;
      const i = t.length || _;
      if (this[o] = typeof i != "function" ? _ : i, this[a] = t.stale || !1, t.maxAge && typeof t.maxAge != "number")
        throw new TypeError("maxAge must be a number");
      this[e] = t.maxAge || 0, this[r] = t.dispose, this[c] = t.noDisposeOnSet || !1, this[d] = t.updateAgeOnGet || !1, this.reset();
    }
    set max(t) {
      if (typeof t != "number" || t < 0)
        throw new TypeError("max must be a non-negative number");
      this[h] = t || 1 / 0, F(this);
    }
    get max() {
      return this[h];
    }
    set allowStale(t) {
      this[a] = !!t;
    }
    get allowStale() {
      return this[a];
    }
    set maxAge(t) {
      if (typeof t != "number")
        throw new TypeError("maxAge must be a non-negative number");
      this[e] = t, F(this);
    }
    get maxAge() {
      return this[e];
    }
    set lengthCalculator(t) {
      typeof t != "function" && (t = _), t !== this[o] && (this[o] = t, this[l] = 0, this[f].forEach((i) => {
        i.length = this[o](i.value, i.key), this[l] += i.length;
      })), F(this);
    }
    get lengthCalculator() {
      return this[o];
    }
    get length() {
      return this[l];
    }
    get itemCount() {
      return this[f].length;
    }
    rforEach(t, i) {
      i = i || this;
      for (let u = this[f].tail; u !== null; ) {
        const b = u.prev;
        v(this, t, u, i), u = b;
      }
    }
    forEach(t, i) {
      i = i || this;
      for (let u = this[f].head; u !== null; ) {
        const b = u.next;
        v(this, t, u, i), u = b;
      }
    }
    keys() {
      return this[f].toArray().map((t) => t.key);
    }
    values() {
      return this[f].toArray().map((t) => t.value);
    }
    reset() {
      this[r] && this[f] && this[f].length && this[f].forEach((t) => this[r](t.key, t.value)), this[p] = /* @__PURE__ */ new Map(), this[f] = new n(), this[l] = 0;
    }
    dump() {
      return this[f].map((t) => L(this, t) ? !1 : {
        k: t.key,
        v: t.value,
        e: t.now + (t.maxAge || 0)
      }).toArray().filter((t) => t);
    }
    dumpLru() {
      return this[f];
    }
    set(t, i, u) {
      if (u = u || this[e], u && typeof u != "number")
        throw new TypeError("maxAge must be a number");
      const b = u ? Date.now() : 0, P = this[o](i, t);
      if (this[p].has(t)) {
        if (P > this[h])
          return C(this, this[p].get(t)), !1;
        const D = this[p].get(t).value;
        return this[r] && (this[c] || this[r](t, D.value)), D.now = b, D.maxAge = u, D.value = i, this[l] += P - D.length, D.length = P, this.get(t), F(this), !0;
      }
      const W = new m(t, i, P, b, u);
      return W.length > this[h] ? (this[r] && this[r](t, i), !1) : (this[l] += W.length, this[f].unshift(W), this[p].set(t, this[f].head), F(this), !0);
    }
    has(t) {
      if (!this[p].has(t))
        return !1;
      const i = this[p].get(t).value;
      return !L(this, i);
    }
    get(t) {
      return y(this, t, !0);
    }
    peek(t) {
      return y(this, t, !1);
    }
    pop() {
      const t = this[f].tail;
      return t ? (C(this, t), t.value) : null;
    }
    del(t) {
      C(this, this[p].get(t));
    }
    load(t) {
      this.reset();
      const i = Date.now();
      for (let u = t.length - 1; u >= 0; u--) {
        const b = t[u], P = b.e || 0;
        if (P === 0)
          this.set(b.k, b.v);
        else {
          const W = P - i;
          W > 0 && this.set(b.k, b.v, W);
        }
      }
    }
    prune() {
      this[p].forEach((t, i) => y(this, i, !1));
    }
  }
  const y = (s, t, i) => {
    const u = s[p].get(t);
    if (u) {
      const b = u.value;
      if (L(s, b)) {
        if (C(s, u), !s[a])
          return;
      } else
        i && (s[d] && (u.value.now = Date.now()), s[f].unshiftNode(u));
      return b.value;
    }
  }, L = (s, t) => {
    if (!t || !t.maxAge && !s[e])
      return !1;
    const i = Date.now() - t.now;
    return t.maxAge ? i > t.maxAge : s[e] && i > s[e];
  }, F = (s) => {
    if (s[l] > s[h])
      for (let t = s[f].tail; s[l] > s[h] && t !== null; ) {
        const i = t.prev;
        C(s, t), t = i;
      }
  }, C = (s, t) => {
    if (t) {
      const i = t.value;
      s[r] && s[r](i.key, i.value), s[l] -= i.length, s[p].delete(i.key), s[f].removeNode(t);
    }
  };
  class m {
    constructor(t, i, u, b, P) {
      this.key = t, this.value = i, this.length = u, this.now = b, this.maxAge = P || 0;
    }
  }
  const v = (s, t, i, u) => {
    let b = i.value;
    L(s, b) && (C(s, i), s[a] || (b = void 0)), b && t.call(u, b.value, b.key, s);
  };
  return ht = g, ht;
}
var ft, _r;
function le() {
  if (_r)
    return ft;
  _r = 1;
  class n {
    constructor(A, N) {
      if (N = o(N), A instanceof n)
        return A.loose === !!N.loose && A.includePrerelease === !!N.includePrerelease ? A : new n(A.raw, N);
      if (A instanceof a)
        return this.raw = A.value, this.set = [[A]], this.format(), this;
      if (this.options = N, this.loose = !!N.loose, this.includePrerelease = !!N.includePrerelease, this.raw = A, this.set = A.split("||").map((S) => this.parseRange(S.trim())).filter((S) => S.length), !this.set.length)
        throw new TypeError(`Invalid SemVer Range: ${A}`);
      if (this.set.length > 1) {
        const S = this.set[0];
        if (this.set = this.set.filter((x) => !g(x[0])), this.set.length === 0)
          this.set = [S];
        else if (this.set.length > 1) {
          for (const x of this.set)
            if (x.length === 1 && y(x[0])) {
              this.set = [x];
              break;
            }
        }
      }
      this.format();
    }
    format() {
      return this.range = this.set.map((A) => A.join(" ").trim()).join("||").trim(), this.range;
    }
    toString() {
      return this.range;
    }
    parseRange(A) {
      A = A.trim();
      const S = `parseRange:${Object.keys(this.options).join(",")}:${A}`, x = l.get(S);
      if (x)
        return x;
      const I = this.options.loose, B = I ? c[f.HYPHENRANGELOOSE] : c[f.HYPHENRANGE];
      A = A.replace(B, W(this.options.includePrerelease)), e("hyphen replace", A), A = A.replace(c[f.COMPARATORTRIM], p), e("comparator trim", A), A = A.replace(c[f.TILDETRIM], d), A = A.replace(c[f.CARETTRIM], _), A = A.split(/\s+/).join(" ");
      let U = A.split(" ").map((Z) => F(Z, this.options)).join(" ").split(/\s+/).map((Z) => P(Z, this.options));
      I && (U = U.filter((Z) => (e("loose invalid filter", Z, this.options), !!Z.match(c[f.COMPARATORLOOSE])))), e("range list", U);
      const z = /* @__PURE__ */ new Map(), Y = U.map((Z) => new a(Z, this.options));
      for (const Z of Y) {
        if (g(Z))
          return [Z];
        z.set(Z.value, Z);
      }
      z.size > 1 && z.has("") && z.delete("");
      const re = [...z.values()];
      return l.set(S, re), re;
    }
    intersects(A, N) {
      if (!(A instanceof n))
        throw new TypeError("a Range is required");
      return this.set.some((S) => L(S, N) && A.set.some((x) => L(x, N) && S.every((I) => x.every((B) => I.intersects(B, N)))));
    }
    test(A) {
      if (!A)
        return !1;
      if (typeof A == "string")
        try {
          A = new r(A, this.options);
        } catch {
          return !1;
        }
      for (let N = 0; N < this.set.length; N++)
        if (ce(this.set[N], A, this.options))
          return !0;
      return !1;
    }
  }
  ft = n;
  const h = yn(), l = new h({ max: 1e3 }), o = De(), a = ke(), e = Ae(), r = se(), {
    re: c,
    t: f,
    comparatorTrimReplace: p,
    tildeTrimReplace: d,
    caretTrimReplace: _
  } = Ce(), g = (D) => D.value === "<0.0.0-0", y = (D) => D.value === "", L = (D, A) => {
    let N = !0;
    const S = D.slice();
    let x = S.pop();
    for (; N && S.length; )
      N = S.every((I) => x.intersects(I, A)), x = S.pop();
    return N;
  }, F = (D, A) => (e("comp", D, A), D = s(D, A), e("caret", D), D = m(D, A), e("tildes", D), D = i(D, A), e("xrange", D), D = b(D, A), e("stars", D), D), C = (D) => !D || D.toLowerCase() === "x" || D === "*", m = (D, A) => D.trim().split(/\s+/).map((N) => v(N, A)).join(" "), v = (D, A) => {
    const N = A.loose ? c[f.TILDELOOSE] : c[f.TILDE];
    return D.replace(N, (S, x, I, B, U) => {
      e("tilde", D, S, x, I, B, U);
      let z;
      return C(x) ? z = "" : C(I) ? z = `>=${x}.0.0 <${+x + 1}.0.0-0` : C(B) ? z = `>=${x}.${I}.0 <${x}.${+I + 1}.0-0` : U ? (e("replaceTilde pr", U), z = `>=${x}.${I}.${B}-${U} <${x}.${+I + 1}.0-0`) : z = `>=${x}.${I}.${B} <${x}.${+I + 1}.0-0`, e("tilde return", z), z;
    });
  }, s = (D, A) => D.trim().split(/\s+/).map((N) => t(N, A)).join(" "), t = (D, A) => {
    e("caret", D, A);
    const N = A.loose ? c[f.CARETLOOSE] : c[f.CARET], S = A.includePrerelease ? "-0" : "";
    return D.replace(N, (x, I, B, U, z) => {
      e("caret", D, x, I, B, U, z);
      let Y;
      return C(I) ? Y = "" : C(B) ? Y = `>=${I}.0.0${S} <${+I + 1}.0.0-0` : C(U) ? I === "0" ? Y = `>=${I}.${B}.0${S} <${I}.${+B + 1}.0-0` : Y = `>=${I}.${B}.0${S} <${+I + 1}.0.0-0` : z ? (e("replaceCaret pr", z), I === "0" ? B === "0" ? Y = `>=${I}.${B}.${U}-${z} <${I}.${B}.${+U + 1}-0` : Y = `>=${I}.${B}.${U}-${z} <${I}.${+B + 1}.0-0` : Y = `>=${I}.${B}.${U}-${z} <${+I + 1}.0.0-0`) : (e("no pr"), I === "0" ? B === "0" ? Y = `>=${I}.${B}.${U}${S} <${I}.${B}.${+U + 1}-0` : Y = `>=${I}.${B}.${U}${S} <${I}.${+B + 1}.0-0` : Y = `>=${I}.${B}.${U} <${+I + 1}.0.0-0`), e("caret return", Y), Y;
    });
  }, i = (D, A) => (e("replaceXRanges", D, A), D.split(/\s+/).map((N) => u(N, A)).join(" ")), u = (D, A) => {
    D = D.trim();
    const N = A.loose ? c[f.XRANGELOOSE] : c[f.XRANGE];
    return D.replace(N, (S, x, I, B, U, z) => {
      e("xRange", D, S, x, I, B, U, z);
      const Y = C(I), re = Y || C(B), Z = re || C(U), _e = Z;
      return x === "=" && _e && (x = ""), z = A.includePrerelease ? "-0" : "", Y ? x === ">" || x === "<" ? S = "<0.0.0-0" : S = "*" : x && _e ? (re && (B = 0), U = 0, x === ">" ? (x = ">=", re ? (I = +I + 1, B = 0, U = 0) : (B = +B + 1, U = 0)) : x === "<=" && (x = "<", re ? I = +I + 1 : B = +B + 1), x === "<" && (z = "-0"), S = `${x + I}.${B}.${U}${z}`) : re ? S = `>=${I}.0.0${z} <${+I + 1}.0.0-0` : Z && (S = `>=${I}.${B}.0${z} <${I}.${+B + 1}.0-0`), e("xRange return", S), S;
    });
  }, b = (D, A) => (e("replaceStars", D, A), D.trim().replace(c[f.STAR], "")), P = (D, A) => (e("replaceGTE0", D, A), D.trim().replace(c[A.includePrerelease ? f.GTE0PRE : f.GTE0], "")), W = (D) => (A, N, S, x, I, B, U, z, Y, re, Z, _e, Nt) => (C(S) ? N = "" : C(x) ? N = `>=${S}.0.0${D ? "-0" : ""}` : C(I) ? N = `>=${S}.${x}.0${D ? "-0" : ""}` : B ? N = `>=${N}` : N = `>=${N}${D ? "-0" : ""}`, C(Y) ? z = "" : C(re) ? z = `<${+Y + 1}.0.0-0` : C(Z) ? z = `<${Y}.${+re + 1}.0-0` : _e ? z = `<=${Y}.${re}.${Z}-${_e}` : D ? z = `<${Y}.${re}.${+Z + 1}-0` : z = `<=${z}`, `${N} ${z}`.trim()), ce = (D, A, N) => {
    for (let S = 0; S < D.length; S++)
      if (!D[S].test(A))
        return !1;
    if (A.prerelease.length && !N.includePrerelease) {
      for (let S = 0; S < D.length; S++)
        if (e(D[S].semver), D[S].semver !== a.ANY && D[S].semver.prerelease.length > 0) {
          const x = D[S].semver;
          if (x.major === A.major && x.minor === A.minor && x.patch === A.patch)
            return !0;
        }
      return !1;
    }
    return !0;
  };
  return ft;
}
var dt, vr;
function ke() {
  if (vr)
    return dt;
  vr = 1;
  const n = Symbol("SemVer ANY");
  class h {
    static get ANY() {
      return n;
    }
    constructor(d, _) {
      if (_ = l(_), d instanceof h) {
        if (d.loose === !!_.loose)
          return d;
        d = d.value;
      }
      r("comparator", d, _), this.options = _, this.loose = !!_.loose, this.parse(d), this.semver === n ? this.value = "" : this.value = this.operator + this.semver.version, r("comp", this);
    }
    parse(d) {
      const _ = this.options.loose ? o[a.COMPARATORLOOSE] : o[a.COMPARATOR], g = d.match(_);
      if (!g)
        throw new TypeError(`Invalid comparator: ${d}`);
      this.operator = g[1] !== void 0 ? g[1] : "", this.operator === "=" && (this.operator = ""), g[2] ? this.semver = new c(g[2], this.options.loose) : this.semver = n;
    }
    toString() {
      return this.value;
    }
    test(d) {
      if (r("Comparator.test", d, this.options.loose), this.semver === n || d === n)
        return !0;
      if (typeof d == "string")
        try {
          d = new c(d, this.options);
        } catch {
          return !1;
        }
      return e(d, this.operator, this.semver, this.options);
    }
    intersects(d, _) {
      if (!(d instanceof h))
        throw new TypeError("a Comparator is required");
      if ((!_ || typeof _ != "object") && (_ = {
        loose: !!_,
        includePrerelease: !1
      }), this.operator === "")
        return this.value === "" ? !0 : new f(d.value, _).test(this.value);
      if (d.operator === "")
        return d.value === "" ? !0 : new f(this.value, _).test(d.semver);
      const g = (this.operator === ">=" || this.operator === ">") && (d.operator === ">=" || d.operator === ">"), y = (this.operator === "<=" || this.operator === "<") && (d.operator === "<=" || d.operator === "<"), L = this.semver.version === d.semver.version, F = (this.operator === ">=" || this.operator === "<=") && (d.operator === ">=" || d.operator === "<="), C = e(this.semver, "<", d.semver, _) && (this.operator === ">=" || this.operator === ">") && (d.operator === "<=" || d.operator === "<"), m = e(this.semver, ">", d.semver, _) && (this.operator === "<=" || this.operator === "<") && (d.operator === ">=" || d.operator === ">");
      return g || y || L && F || C || m;
    }
  }
  dt = h;
  const l = De(), { re: o, t: a } = Ce(), e = Tr(), r = Ae(), c = se(), f = le();
  return dt;
}
var pt, mr;
function Ie() {
  if (mr)
    return pt;
  mr = 1;
  const n = le();
  return pt = (l, o, a) => {
    try {
      o = new n(o, a);
    } catch {
      return !1;
    }
    return o.test(l);
  }, pt;
}
var _t, gr;
function Ln() {
  if (gr)
    return _t;
  gr = 1;
  const n = le();
  return _t = (l, o) => new n(l, o).set.map((a) => a.map((e) => e.value).join(" ").trim().split(" ")), _t;
}
var vt, Cr;
function wn() {
  if (Cr)
    return vt;
  Cr = 1;
  const n = se(), h = le();
  return vt = (o, a, e) => {
    let r = null, c = null, f = null;
    try {
      f = new h(a, e);
    } catch {
      return null;
    }
    return o.forEach((p) => {
      f.test(p) && (!r || c.compare(p) === -1) && (r = p, c = new n(r, e));
    }), r;
  }, vt;
}
var mt, Er;
function Fn() {
  if (Er)
    return mt;
  Er = 1;
  const n = se(), h = le();
  return mt = (o, a, e) => {
    let r = null, c = null, f = null;
    try {
      f = new h(a, e);
    } catch {
      return null;
    }
    return o.forEach((p) => {
      f.test(p) && (!r || c.compare(p) === 1) && (r = p, c = new n(r, e));
    }), r;
  }, mt;
}
var gt, $r;
function Rn() {
  if ($r)
    return gt;
  $r = 1;
  const n = se(), h = le(), l = be();
  return gt = (a, e) => {
    a = new h(a, e);
    let r = new n("0.0.0");
    if (a.test(r) || (r = new n("0.0.0-0"), a.test(r)))
      return r;
    r = null;
    for (let c = 0; c < a.set.length; ++c) {
      const f = a.set[c];
      let p = null;
      f.forEach((d) => {
        const _ = new n(d.semver.version);
        switch (d.operator) {
          case ">":
            _.prerelease.length === 0 ? _.patch++ : _.prerelease.push(0), _.raw = _.format();
          case "":
          case ">=":
            (!p || l(_, p)) && (p = _);
            break;
          case "<":
          case "<=":
            break;
          default:
            throw new Error(`Unexpected operation: ${d.operator}`);
        }
      }), p && (!r || l(r, p)) && (r = p);
    }
    return r && a.test(r) ? r : null;
  }, gt;
}
var Ct, yr;
function An() {
  if (yr)
    return Ct;
  yr = 1;
  const n = le();
  return Ct = (l, o) => {
    try {
      return new n(l, o).range || "*";
    } catch {
      return null;
    }
  }, Ct;
}
var Et, Lr;
function xt() {
  if (Lr)
    return Et;
  Lr = 1;
  const n = se(), h = ke(), { ANY: l } = h, o = le(), a = Ie(), e = be(), r = It(), c = Vt(), f = St();
  return Et = (d, _, g, y) => {
    d = new n(d, y), _ = new o(_, y);
    let L, F, C, m, v;
    switch (g) {
      case ">":
        L = e, F = c, C = r, m = ">", v = ">=";
        break;
      case "<":
        L = r, F = f, C = e, m = "<", v = "<=";
        break;
      default:
        throw new TypeError('Must provide a hilo val of "<" or ">"');
    }
    if (a(d, _, y))
      return !1;
    for (let s = 0; s < _.set.length; ++s) {
      const t = _.set[s];
      let i = null, u = null;
      if (t.forEach((b) => {
        b.semver === l && (b = new h(">=0.0.0")), i = i || b, u = u || b, L(b.semver, i.semver, y) ? i = b : C(b.semver, u.semver, y) && (u = b);
      }), i.operator === m || i.operator === v || (!u.operator || u.operator === m) && F(d, u.semver))
        return !1;
      if (u.operator === v && C(d, u.semver))
        return !1;
    }
    return !0;
  }, Et;
}
var $t, wr;
function Dn() {
  if (wr)
    return $t;
  wr = 1;
  const n = xt();
  return $t = (l, o, a) => n(l, o, ">", a), $t;
}
var yt, Fr;
function bn() {
  if (Fr)
    return yt;
  Fr = 1;
  const n = xt();
  return yt = (l, o, a) => n(l, o, "<", a), yt;
}
var Lt, Rr;
function kn() {
  if (Rr)
    return Lt;
  Rr = 1;
  const n = le();
  return Lt = (l, o, a) => (l = new n(l, a), o = new n(o, a), l.intersects(o)), Lt;
}
var wt, Ar;
function In() {
  if (Ar)
    return wt;
  Ar = 1;
  const n = Ie(), h = ae();
  return wt = (l, o, a) => {
    const e = [];
    let r = null, c = null;
    const f = l.sort((g, y) => h(g, y, a));
    for (const g of f)
      n(g, o, a) ? (c = g, r || (r = g)) : (c && e.push([r, c]), c = null, r = null);
    r && e.push([r, null]);
    const p = [];
    for (const [g, y] of e)
      g === y ? p.push(g) : !y && g === f[0] ? p.push("*") : y ? g === f[0] ? p.push(`<=${y}`) : p.push(`${g} - ${y}`) : p.push(`>=${g}`);
    const d = p.join(" || "), _ = typeof o.raw == "string" ? o.raw : String(o);
    return d.length < _.length ? d : o;
  }, wt;
}
var Ft, Dr;
function Sn() {
  if (Dr)
    return Ft;
  Dr = 1;
  const n = le(), h = ke(), { ANY: l } = h, o = Ie(), a = ae(), e = (p, d, _ = {}) => {
    if (p === d)
      return !0;
    p = new n(p, _), d = new n(d, _);
    let g = !1;
    e:
      for (const y of p.set) {
        for (const L of d.set) {
          const F = r(y, L, _);
          if (g = g || F !== null, F)
            continue e;
        }
        if (g)
          return !1;
      }
    return !0;
  }, r = (p, d, _) => {
    if (p === d)
      return !0;
    if (p.length === 1 && p[0].semver === l) {
      if (d.length === 1 && d[0].semver === l)
        return !0;
      _.includePrerelease ? p = [new h(">=0.0.0-0")] : p = [new h(">=0.0.0")];
    }
    if (d.length === 1 && d[0].semver === l) {
      if (_.includePrerelease)
        return !0;
      d = [new h(">=0.0.0")];
    }
    const g = /* @__PURE__ */ new Set();
    let y, L;
    for (const u of p)
      u.operator === ">" || u.operator === ">=" ? y = c(y, u, _) : u.operator === "<" || u.operator === "<=" ? L = f(L, u, _) : g.add(u.semver);
    if (g.size > 1)
      return null;
    let F;
    if (y && L) {
      if (F = a(y.semver, L.semver, _), F > 0)
        return null;
      if (F === 0 && (y.operator !== ">=" || L.operator !== "<="))
        return null;
    }
    for (const u of g) {
      if (y && !o(u, String(y), _) || L && !o(u, String(L), _))
        return null;
      for (const b of d)
        if (!o(u, String(b), _))
          return !1;
      return !0;
    }
    let C, m, v, s, t = L && !_.includePrerelease && L.semver.prerelease.length ? L.semver : !1, i = y && !_.includePrerelease && y.semver.prerelease.length ? y.semver : !1;
    t && t.prerelease.length === 1 && L.operator === "<" && t.prerelease[0] === 0 && (t = !1);
    for (const u of d) {
      if (s = s || u.operator === ">" || u.operator === ">=", v = v || u.operator === "<" || u.operator === "<=", y) {
        if (i && u.semver.prerelease && u.semver.prerelease.length && u.semver.major === i.major && u.semver.minor === i.minor && u.semver.patch === i.patch && (i = !1), u.operator === ">" || u.operator === ">=") {
          if (C = c(y, u, _), C === u && C !== y)
            return !1;
        } else if (y.operator === ">=" && !o(y.semver, String(u), _))
          return !1;
      }
      if (L) {
        if (t && u.semver.prerelease && u.semver.prerelease.length && u.semver.major === t.major && u.semver.minor === t.minor && u.semver.patch === t.patch && (t = !1), u.operator === "<" || u.operator === "<=") {
          if (m = f(L, u, _), m === u && m !== L)
            return !1;
        } else if (L.operator === "<=" && !o(L.semver, String(u), _))
          return !1;
      }
      if (!u.operator && (L || y) && F !== 0)
        return !1;
    }
    return !(y && v && !L && F !== 0 || L && s && !y && F !== 0 || i || t);
  }, c = (p, d, _) => {
    if (!p)
      return d;
    const g = a(p.semver, d.semver, _);
    return g > 0 ? p : g < 0 || d.operator === ">" && p.operator === ">=" ? d : p;
  }, f = (p, d, _) => {
    if (!p)
      return d;
    const g = a(p.semver, d.semver, _);
    return g < 0 ? p : g > 0 || d.operator === "<" && p.operator === "<=" ? d : p;
  };
  return Ft = e, Ft;
}
var Rt, br;
function Vn() {
  if (br)
    return Rt;
  br = 1;
  const n = Ce(), h = Re(), l = se(), o = Br(), a = Ee(), e = on(), r = an(), c = ln(), f = cn(), p = hn(), d = fn(), _ = dn(), g = pn(), y = ae(), L = _n(), F = vn(), C = kt(), m = mn(), v = gn(), s = be(), t = It(), i = bt(), u = Or(), b = St(), P = Vt(), W = Tr(), ce = Cn(), D = ke(), A = le(), N = Ie(), S = Ln(), x = wn(), I = Fn(), B = Rn(), U = An(), z = xt(), Y = Dn(), re = bn(), Z = kn(), _e = In(), Nt = Sn();
  return Rt = {
    parse: a,
    valid: e,
    clean: r,
    inc: c,
    diff: f,
    major: p,
    minor: d,
    patch: _,
    prerelease: g,
    compare: y,
    rcompare: L,
    compareLoose: F,
    compareBuild: C,
    sort: m,
    rsort: v,
    gt: s,
    lt: t,
    eq: i,
    neq: u,
    gte: b,
    lte: P,
    cmp: W,
    coerce: ce,
    Comparator: D,
    Range: A,
    satisfies: N,
    toComparators: S,
    maxSatisfying: x,
    minSatisfying: I,
    minVersion: B,
    validRange: U,
    outside: z,
    gtr: Y,
    ltr: re,
    intersects: Z,
    simplifyRange: _e,
    subset: Nt,
    SemVer: l,
    re: n.re,
    src: n.src,
    tokens: n.t,
    SEMVER_SPEC_VERSION: h.SEMVER_SPEC_VERSION,
    compareIdentifiers: o.compareIdentifiers,
    rcompareIdentifiers: o.rcompareIdentifiers
  }, Rt;
}
var xn = Vn();
function Mr(n) {
  const { versions: h } = k.market.data[n] || k.dependencies[n];
  return Bt(h, (l) => {
    const o = Bt({ ...l.peerDependencies }, (e, r) => {
      var p, d;
      const c = ((p = k.dependencies[r]) == null ? void 0 : p.resolved) || ((d = k.packages[r]) == null ? void 0 : d.version), f = c ? xn.satisfies(c, e) ? "success" : "danger" : "warning";
      return { request: e, resolved: c, result: f };
    });
    let a = "success";
    for (const e of Object.values(o)) {
      if (e.result === "danger") {
        a = "danger";
        break;
      }
      a === "warning" && (a = "warning");
    }
    return { peers: o, result: a };
  });
}
const ve = ie(!1);
async function qr(n, h) {
  const l = Jr({
    text: "\u6B63\u5728\u66F4\u65B0\u4F9D\u8D56\u2026\u2026"
  }), o = pe(en, () => {
    me.success("\u5B89\u88C5\u6210\u529F\uFF01"), o(), l.close();
  });
  try {
    ve.value = !1, await ue("market/install", n) ? me.error("\u5B89\u88C5\u5931\u8D25\uFF01") : (me.success("\u5B89\u88C5\u6210\u529F\uFF01"), h == null || h());
  } catch {
    me.error("\u5B89\u88C5\u8D85\u65F6\uFF01");
  } finally {
    o(), l.close();
  }
}
const ee = tn("manager", 2, () => ({
  prefix: "^",
  override: {},
  showInstalled: !1,
  hideWorkspace: !0
}));
q(() => Object.values(ee.override).filter((n) => n !== void 0).length);
pe(() => k.dependencies, (n) => {
  var h;
  if (!!n)
    for (const l in ee.override)
      ee.override[l] ? ((h = n[l]) == null ? void 0 : h.request) === ee.override[l] && delete ee.override[l] : n[l] || delete ee.override[l];
}, { immediate: !0 });
const J = ie(""), ye = ie(!1), zr = (n) => {
  var h;
  return {
    keywords: [],
    peerDependencies: {},
    ...k.market.data[n],
    ...(h = k.packages) == null ? void 0 : h[n]
  };
}, Hr = q(() => ({
  icon: "refresh",
  label: "\u5237\u65B0",
  type: !k.market || k.market.progress < k.market.total ? "spin disabled" : "",
  action: () => ue("market/refresh")
})), Nn = ["id"], Bn = {
  key: 0,
  class: "padding"
}, On = /* @__PURE__ */ $("tr", null, [
  /* @__PURE__ */ $("td", null, "\u4F9D\u8D56\u540D\u79F0"),
  /* @__PURE__ */ $("td", null, "\u7248\u672C\u533A\u95F4"),
  /* @__PURE__ */ $("td", null, "\u5F53\u524D\u7248\u672C"),
  /* @__PURE__ */ $("td", null, "\u53EF\u7528\u6027")
], -1), Tn = { class: "name" }, Mn = { key: 0 }, qn = { key: 1 }, zn = { key: 2 }, Hn = { key: 2 }, Pn = ["onClick"], jn = { key: 0 }, Gn = { key: 3 }, Un = /* @__PURE__ */ $("div", { class: "left" }, null, -1), Xn = { class: "right" }, Wn = /* @__PURE__ */ ne({
  __name: "install",
  setup(n) {
    function h(y) {
      qr({ [J.value]: y }, () => {
        if (_.value.length)
          return;
        const L = d.value + ":" + Math.random().toString(36).slice(2, 8);
        ue("manager/unload", L, {});
      });
    }
    const l = ie(), o = q({
      get() {
        var y;
        return ((y = k.dependencies[J.value]) == null ? void 0 : y.request) === l.value ? l.value + " (\u5F53\u524D)" : l.value;
      },
      set(y) {
        l.value = y;
      }
    }), a = q(() => {
      var y;
      return l.value === ((y = k.dependencies[J.value]) == null ? void 0 : y.request);
    }), e = q(() => {
      var y;
      return (y = k.dependencies[J.value]) == null ? void 0 : y.resolved;
    }), r = q(() => {
      var y;
      return (y = k.packages) == null ? void 0 : y[J.value];
    }), c = q(() => {
      var y, L;
      return ((y = k.dependencies[J.value]) == null ? void 0 : y.workspace) || ((L = r.value) == null ? void 0 : L.workspace);
    }), f = q(() => {
      if (!(!J.value || c.value))
        return Mr(J.value);
    });
    pe(() => J.value, (y) => {
      var L;
      ve.value = !!y, y && (l.value = ee.override[J.value] || ((L = k.dependencies[J.value]) == null ? void 0 : L.request) || k.market.data[y].version);
    }, { immediate: !0 });
    function* p(y, L, F) {
      for (let C in L) {
        const m = L[C], v = !C.startsWith("~");
        v || (C = C.slice(1));
        const s = C.split(":")[0];
        s === y && (yield [F + C, v]), s === "group" && (yield* p(y, m, F + C + "/"));
      }
    }
    const d = q(() => J.value.replace(/(koishi-|^@koishijs\/)plugin-/, "")), _ = q(() => [...p(d.value, k.config.plugins, "")]);
    function g(y) {
      ve.value = !1, y === !0 && (y = d.value + ":" + Math.random().toString(36).slice(2, 8), ue("manager/unload", y, {})), fe.push("/plugins/" + y);
    }
    return (y, L) => {
      const F = M("el-option"), C = M("el-select"), m = M("k-icon"), v = M("el-scrollbar"), s = M("el-button"), t = M("el-dialog");
      return E(), T(t, {
        modelValue: w(ve),
        "onUpdate:modelValue": L[8] || (L[8] = (i) => de(ve) ? ve.value = i : null),
        class: "install-panel",
        onClosed: L[9] || (L[9] = (i) => J.value = "")
      }, Xr({
        default: V(() => {
          var i, u;
          return [
            !w(J) || w(c) || !Object.keys(w(f)[l.value].peers).length ? (E(), R("div", Bn)) : (E(), T(v, { key: 1 }, {
              default: V(() => [
                $("table", null, [
                  On,
                  (E(!0), R(j, null, te(w(f)[l.value].peers, ({ request: b, resolved: P, result: W }, ce) => (E(), R("tr", { key: ce }, [
                    $("td", Tn, H(ce), 1),
                    $("td", null, H(b), 1),
                    $("td", null, H(P), 1),
                    $("td", {
                      class: ge(["theme-color", W])
                    }, [
                      W === "warning" ? (E(), R("span", Mn, [
                        G(m, { name: "exclamation-full" }),
                        O("\u672A\u4E0B\u8F7D")
                      ])) : W === "success" ? (E(), R("span", qn, [
                        G(m, { name: "check-full" }),
                        O("\u5DF2\u4E0B\u8F7D")
                      ])) : (E(), R("span", zn, [
                        G(m, { name: "times-full" }),
                        O("\u4E0D\u517C\u5BB9")
                      ]))
                    ], 2)
                  ]))), 128))
                ])
              ]),
              _: 1
            })),
            w(r) && w(_).length ? (E(), R("div", Hn, [
              O(" \u70B9\u51FB\u4EE5\u524D\u5F80\u914D\u7F6E\u9875\u9762\uFF1A "),
              $("ul", null, [
                (E(!0), R(j, null, te(w(_), ([b, P]) => (E(), R("li", { key: b }, [
                  $("span", {
                    class: "link",
                    onClick: he((W) => g(b), ["stop"])
                  }, H(b), 9, Pn),
                  O(" (" + H(P ? "\u8FD0\u884C\u4E2D" : "\u95F2\u7F6E") + ") ", 1)
                ]))), 128)),
                (u = (i = w(k).packages) == null ? void 0 : i[w(J)]) != null && u.id ? Q("", !0) : (E(), R("li", jn, [
                  $("span", {
                    class: "link",
                    onClick: L[1] || (L[1] = he((b) => g(!0), ["stop"]))
                  }, "\u6DFB\u52A0\u65B0\u914D\u7F6E")
                ]))
              ])
            ])) : w(r) ? (E(), R("div", Gn, [
              $("span", {
                class: "link",
                onClick: L[2] || (L[2] = he((b) => g(!0), ["stop"]))
              }, "\u4F60\u5C1A\u672A\u914D\u7F6E\u6B64\u63D2\u4EF6\uFF0C\u70B9\u51FB\u7ACB\u5373\u914D\u7F6E\u3002")
            ])) : Q("", !0)
          ];
        }),
        _: 2
      }, [
        w(J) ? {
          name: "header",
          fn: V(({ titleId: i, titleClass: u }) => [
            $("span", {
              id: i,
              class: ge([u, ""])
            }, H(w(J).replace(/(koishi-|^@koishijs\/)plugin-/, "") + (w(c) ? " (\u5DE5\u4F5C\u533A)" : "")), 11, Nn),
            w(c) ? Q("", !0) : (E(), T(C, {
              key: 0,
              disabled: w(c),
              modelValue: w(o),
              "onUpdate:modelValue": L[0] || (L[0] = (b) => de(o) ? o.value = b : null)
            }, {
              default: V(() => [
                (E(!0), R(j, null, te(w(f), ({ result: b }, P) => (E(), T(F, {
                  key: P,
                  value: P
                }, {
                  default: V(() => [
                    O(H(P) + " ", 1),
                    P === w(e) ? (E(), R(j, { key: 0 }, [
                      O("(\u5F53\u524D)")
                    ], 64)) : Q("", !0),
                    $("span", {
                      class: ge([b, "theme-color", "dot-hint"])
                    }, null, 2)
                  ]),
                  _: 2
                }, 1032, ["value"]))), 128))
              ]),
              _: 1
            }, 8, ["disabled", "modelValue"]))
          ]),
          key: "0"
        } : void 0,
        w(J) ? {
          name: "footer",
          fn: V(() => [
            Un,
            $("div", Xn, [
              G(s, {
                onClick: L[3] || (L[3] = (i) => ve.value = !1)
              }, {
                default: V(() => [
                  O("\u53D6\u6D88")
                ]),
                _: 1
              }),
              w(c) ? (E(), R(j, { key: 0 }, [
                w(e) ? (E(), T(s, {
                  key: 0,
                  onClick: L[4] || (L[4] = (i) => h(""))
                }, {
                  default: V(() => [
                    O("\u79FB\u9664")
                  ]),
                  _: 1
                })) : (E(), T(s, {
                  key: 1,
                  onClick: L[5] || (L[5] = (i) => h(l.value)),
                  disabled: w(a)
                }, {
                  default: V(() => [
                    O("\u6DFB\u52A0")
                  ]),
                  _: 1
                }, 8, ["disabled"]))
              ], 64)) : (E(), R(j, { key: 1 }, [
                w(e) ? (E(), T(s, {
                  key: 0,
                  onClick: L[6] || (L[6] = (i) => h(""))
                }, {
                  default: V(() => [
                    O("\u79FB\u9664")
                  ]),
                  _: 1
                })) : Q("", !0),
                G(s, {
                  type: w(f)[l.value].result,
                  onClick: L[7] || (L[7] = (i) => h(l.value)),
                  disabled: w(a)
                }, {
                  default: V(() => [
                    O(H(w(e) ? "\u66F4\u65B0" : "\u5B89\u88C5"), 1)
                  ]),
                  _: 1
                }, 8, ["type", "disabled"])
              ], 64))
            ])
          ]),
          key: "1"
        } : void 0
      ]), 1032, ["modelValue"]);
    };
  }
});
const Yn = { class: "dep-package-view" }, Kn = { class: "name" }, Zn = { class: "current" }, Qn = /* @__PURE__ */ ne({
  __name: "package",
  props: {
    name: String
  },
  setup(n) {
    const h = n, l = q(() => k.dependencies[h.name]), o = q({
      get() {
        const e = ee.override[h.name];
        if (l.value.resolved !== e)
          return e === "" ? "\u79FB\u9664\u4F9D\u8D56" : e;
      },
      set(e) {
        l.value.resolved === e ? delete ee.override[h.name] : ee.override[h.name] = e, console.log(ee.override);
      }
    }), a = q(() => Mr(h.name));
    return (e, r) => {
      const c = M("el-option"), f = M("el-select"), p = M("el-button");
      return E(), R("tr", Yn, [
        $("td", Kn, H(n.name), 1),
        $("td", Zn, [
          w(l).workspace ? (E(), R(j, { key: 0 }, [
            O("\u5DE5\u4F5C\u533A")
          ], 64)) : (E(), R(j, { key: 1 }, [
            O(H(w(l).resolved) + " (" + H(w(l).resolved === w(l).latest ? "\u6700\u65B0" : "\u53EF\u66F4\u65B0") + ") ", 1)
          ], 64))
        ]),
        $("td", null, [
          w(l).workspace ? Q("", !0) : (E(), T(f, {
            key: 0,
            modelValue: w(o),
            "onUpdate:modelValue": r[0] || (r[0] = (d) => de(o) ? o.value = d : null)
          }, {
            default: V(() => [
              G(c, { value: "" }, {
                default: V(() => [
                  O("\u79FB\u9664\u4F9D\u8D56")
                ]),
                _: 1
              }),
              (E(!0), R(j, null, te(w(a), ({ result: d }, _) => (E(), T(c, {
                key: _,
                value: _
              }, {
                default: V(() => [
                  O(H(_) + " ", 1),
                  _ === w(l).resolved ? (E(), R(j, { key: 0 }, [
                    O("(\u5F53\u524D)")
                  ], 64)) : Q("", !0),
                  $("span", {
                    class: ge([d, "theme-color", "dot-hint"])
                  }, null, 2)
                ]),
                _: 2
              }, 1032, ["value"]))), 128))
            ]),
            _: 1
          }, 8, ["modelValue"]))
        ]),
        $("td", null, [
          w(l).workspace || w(l).versions ? (E(), T(p, {
            key: 0,
            onClick: r[1] || (r[1] = (d) => J.value = n.name)
          }, {
            default: V(() => [
              O("\u4FEE\u6539")
            ]),
            _: 1
          })) : (E(), R(j, { key: 1 }, [
            O(" \u7248\u672C\u83B7\u53D6\u5931\u8D25 ")
          ], 64))
        ])
      ]);
    };
  }
});
const X = (n, h) => {
  const l = n.__vccOpts || n;
  for (const [o, a] of h)
    l[o] = a;
  return l;
}, Jn = /* @__PURE__ */ X(Qn, [["__scopeId", "data-v-a69b4877"]]), es = /* @__PURE__ */ $("table", { class: "table-head" }, [
  /* @__PURE__ */ $("colgroup", null, [
    /* @__PURE__ */ $("col", { width: "auto" }),
    /* @__PURE__ */ $("col", { width: "15%" }),
    /* @__PURE__ */ $("col", { width: "30%" }),
    /* @__PURE__ */ $("col", { width: "15%" })
  ]),
  /* @__PURE__ */ $("thead", null, [
    /* @__PURE__ */ $("tr", null, [
      /* @__PURE__ */ $("th", null, "\u4F9D\u8D56\u540D\u79F0"),
      /* @__PURE__ */ $("th", null, "\u672C\u5730\u7248\u672C"),
      /* @__PURE__ */ $("th", null, "\u76EE\u6807\u7248\u672C"),
      /* @__PURE__ */ $("th", null, "\u64CD\u4F5C")
    ])
  ])
], -1), ts = { class: "table-body" }, rs = /* @__PURE__ */ $("colgroup", null, [
  /* @__PURE__ */ $("col", { width: "auto" }),
  /* @__PURE__ */ $("col", { width: "15%" }),
  /* @__PURE__ */ $("col", { width: "30%" }),
  /* @__PURE__ */ $("col", { width: "15%" })
], -1), ns = /* @__PURE__ */ ne({
  __name: "index",
  setup(n) {
    const h = q(() => Object.keys(k.dependencies).filter((o) => !k.dependencies[o].workspace).sort((o, a) => o > a ? 1 : -1)), l = q(() => [{
      icon: "check",
      label: "\u5E94\u7528\u66F4\u6539",
      disabled: !Object.keys(ee.override).length,
      async action() {
        return qr(ee.override);
      }
    }, Hr.value]);
    return (o, a) => {
      const e = M("el-scrollbar"), r = M("k-layout");
      return E(), T(r, {
        main: "page-deps",
        menu: w(l)
      }, {
        default: V(() => [
          es,
          G(e, { class: "body-container" }, {
            default: V(() => [
              $("table", ts, [
                rs,
                $("tbody", null, [
                  (E(!0), R(j, null, te(w(h), (c) => (E(), T(Jn, {
                    key: c,
                    name: c
                  }, null, 8, ["name"]))), 128))
                ])
              ])
            ]),
            _: 1
          })
        ]),
        _: 1
      }, 8, ["menu"]);
    };
  }
});
function ss(n, h) {
  var l;
  return (l = {
    ...h.versions[h.version],
    ...k.packages[h.name]
  }.manifest) == null ? void 0 : l.service.implements.includes(n);
}
function us(n) {
  function h(a, e) {
    if (a === "console") {
      o.console = !0;
      return;
    }
    const r = Object.values(k.market.data).filter((c) => ss(a, c)).map((c) => c.name);
    o.using[a] = { required: e, available: r };
  }
  const l = k.packages[n], o = { impl: [], using: {} };
  for (const a of l.manifest.service.implements)
    a !== "adapter" && o.impl.push(a);
  for (const a of l.manifest.service.required)
    h(a, !0);
  for (const a of l.manifest.service.optional)
    h(a, !1);
  return l.id && !l.forkable && (o.warning = !0), l.schema || (o.warning = !0), o;
}
const Pr = q(() => Object.fromEntries(Object.keys(k.packages).filter((n) => n).map((n) => [n, us(n)])));
function jr(n, h) {
  const l = [];
  for (let o in h) {
    if (o.startsWith("$"))
      continue;
    const a = h[o], e = { config: a, parent: n };
    o.startsWith("~") && (e.disabled = !0, o = o.slice(1)), e.label = o.split(":", 1)[0], e.alias = o.slice(e.label.length + 1), e.id = e.path = n.path + (n.path ? "/" : "") + o, o.startsWith("group:") && (e.children = jr(e, a)), l.push(e);
  }
  return l;
}
const oe = q(() => {
  const n = {
    label: "\u6240\u6709\u63D2\u4EF6",
    id: "",
    path: "",
    alias: "",
    config: k.config.plugins
  };
  n.children = jr(n, k.config.plugins);
  const h = {
    "@global": {
      label: "\u5168\u5C40\u8BBE\u7F6E",
      path: "@global",
      id: "@global",
      alias: "",
      config: k.config
    }
  }, l = [];
  function o(a) {
    var e, r;
    !((e = a.config) != null && e.$collapsed) && a.children && l.push(a.id), h[a.path] = a, (r = a.children) == null || r.forEach(o);
  }
  return o(n), { data: [n], paths: h, expanded: l };
});
function Gr(n, h) {
  if (n !== h) {
    for (const l of Object.keys(oe.value.paths)) {
      if (l !== n && !l.startsWith(n + "/"))
        continue;
      const o = oe.value.paths[l];
      o.path = h + l.slice(n.length), oe.value.paths[o.path] = o, delete oe.value.paths[l];
    }
    fe.replace("/plugins/" + h);
  }
}
function Le(n) {
  return n.split(/\/?(@[\w-]+\/[\w:-]+|[\w:-]+)\/?/).filter(Boolean);
}
function is(n, h, l) {
  const o = Math.random().toString(36).slice(2, 8);
  n === "@global" && (n = ""), n && (n += "/"), n += l + ":" + o, ue(`manager/${h}`, n), fe.replace("/plugins/" + n);
}
function os(n) {
  ue("manager/remove", n);
  const h = Le(n);
  h.pop(), fe.replace("/plugins/" + h.join("/"));
}
const as = /* @__PURE__ */ ne({
  __name: "global",
  props: {
    current: null,
    modelValue: null
  },
  emits: ["update:modelValue"],
  setup(n, { emit: h }) {
    const l = n, o = q({
      get: () => l.modelValue,
      set: (a) => h("update:modelValue", a)
    });
    return (a, e) => {
      const r = M("k-form");
      return E(), T(r, {
        schema: w(k).packages[""].schema,
        initial: n.current.config,
        modelValue: w(o),
        "onUpdate:modelValue": e[0] || (e[0] = (c) => de(o) ? o.value = c : null)
      }, null, 8, ["schema", "initial", "modelValue"]);
    };
  }
}), ls = (n) => (Wr("data-v-294bd254"), n = n(), Yr(), n), cs = {
  key: 0,
  class: "k-form"
}, hs = /* @__PURE__ */ ls(() => /* @__PURE__ */ $("h2", null, "\u8FC7\u6EE4\u5668", -1)), fs = /* @__PURE__ */ ne({
  __name: "group",
  props: {
    current: null,
    modelValue: null
  },
  setup(n) {
    const h = n, l = ie();
    return pe(() => h.current.config, (o) => {
      l.value = xr(o);
    }, { immediate: !0 }), $e.object({
      user: $e.array(String).description("\u7528\u6237\u5217\u8868"),
      channel: $e.array(String).description("\u9891\u9053\u5217\u8868"),
      guild: $e.array(String).description("\u7FA4\u7EC4\u5217\u8868"),
      platform: $e.array(String).description("\u5E73\u53F0\u5217\u8868")
    }).description("\u8FC7\u6EE4\u5668"), (o, a) => Object.keys(n.current.config.$filter || {}).length ? (E(), R("div", cs, [
      hs,
      $("ul", null, [
        (E(!0), R(j, null, te(n.current.config.$filter, (e, r) => (E(), R("li", null, H(r) + ": " + H(e), 1))), 256))
      ])
    ])) : Q("", !0);
  }
});
const ds = /* @__PURE__ */ X(fs, [["__scopeId", "data-v-294bd254"]]), ps = { class: "search" }, _s = { class: "item" }, vs = { class: "label" }, ms = /* @__PURE__ */ $("div", { class: "right" }, null, -1), gs = /* @__PURE__ */ ne({
  __name: "tree",
  props: {
    modelValue: null
  },
  emits: ["update:modelValue"],
  setup(n, { emit: h }) {
    const l = n, o = q({
      get: () => l.modelValue,
      set: (F) => h("update:modelValue", F)
    });
    function a(F, C) {
      return C.label.includes(L.value);
    }
    function e(F) {
      return F.data.path !== "";
    }
    function r(F, C, m) {
      if (m !== "inner")
        return C.data.path !== "";
      const v = Le(C.data.path);
      return v[v.length - 1].startsWith("group:");
    }
    function c(F) {
      o.value = F.path;
    }
    function f(F, C, m) {
      ue("manager/meta", F.path, { $collapsed: null });
    }
    function p(F, C, m) {
      ue("manager/meta", F.path, { $collapsed: !0 });
    }
    function d(F, C, m, v) {
      const s = m === "inner" ? C : C.parent, t = F.data.path, i = s.data.path, u = s.childNodes.findIndex((W) => W.data.path === t);
      ue("manager/teleport", t, i, u);
      const b = Le(t), P = i ? Le(i) : [];
      P.push(b.pop()), Gr(t, P.join("/"));
    }
    function _(F) {
      const C = [];
      return F.children && C.push("is-group"), F.disabled && C.push("is-disabled"), F.path === o.value && C.push("is-active"), C.join(" ");
    }
    const g = ie(null), y = ie(null), L = ie("");
    return pe(L, (F) => {
      y.value.filter(F);
    }), Kr(async () => {
      const F = g.value.$el;
      await Zr();
      const C = F.querySelector(".k-tab-item.active");
      !C || g.value.setScrollTop(C.offsetTop - (F.offsetHeight - C.offsetHeight) / 2);
    }), (F, C) => {
      const m = M("k-icon"), v = M("el-input"), s = M("k-tab-item"), t = M("el-tree"), i = M("el-scrollbar");
      return E(), T(i, {
        class: "plugin-tree",
        ref_key: "root",
        ref: g
      }, {
        default: V(() => [
          $("div", ps, [
            G(v, {
              modelValue: L.value,
              "onUpdate:modelValue": C[0] || (C[0] = (u) => L.value = u)
            }, {
              suffix: V(() => [
                G(m, { name: "search" })
              ]),
              _: 1
            }, 8, ["modelValue"])
          ]),
          G(s, {
            class: "k-tab-group-title",
            label: "@global",
            modelValue: w(o),
            "onUpdate:modelValue": C[1] || (C[1] = (u) => de(o) ? o.value = u : null)
          }, {
            default: V(() => [
              O("\u5168\u5C40\u8BBE\u7F6E")
            ]),
            _: 1
          }, 8, ["modelValue"]),
          G(t, {
            ref_key: "tree",
            ref: y,
            "node-key": "id",
            data: w(oe).data,
            draggable: !0,
            "auto-expand-parent": !1,
            "default-expanded-keys": w(oe).expanded,
            "expand-on-click-node": !1,
            "filter-node-method": a,
            props: { class: _ },
            "allow-drag": e,
            "allow-drop": r,
            onNodeClick: c,
            onNodeDrop: d,
            onNodeExpand: f,
            onNodeCollapse: p
          }, {
            default: V(({ node: u }) => [
              $("div", _s, [
                $("div", vs, H(u.label === "group" ? "\u5206\u7EC4\uFF1A" + u.data.alias : u.label || "\u5F85\u6DFB\u52A0"), 1),
                ms
              ])
            ]),
            _: 1
          }, 8, ["data", "default-expanded-keys", "props"])
        ]),
        _: 1
      }, 512);
    };
  }
});
const Cs = /* @__PURE__ */ ne({
  __name: "dep-link",
  props: {
    name: null
  },
  setup(n) {
    return (h, l) => (E(), R("span", {
      class: "link",
      onClick: l[0] || (l[0] = he((o) => J.value = n.name, ["stop"]))
    }, H(n.name), 1));
  }
}), Es = /* @__PURE__ */ $("p", null, "\u6B64\u63D2\u4EF6\u5DF2\u5728\u8FD0\u884C\u4E14\u4E0D\u53EF\u91CD\u7528\uFF0C\u542F\u7528\u53EF\u80FD\u4F1A\u5BFC\u81F4\u975E\u9884\u671F\u7684\u95EE\u9898\u3002", -1), $s = { key: 0 }, ys = ["onClick"], Ls = /* @__PURE__ */ ne({
  __name: "plugin",
  props: {
    current: null,
    modelValue: null
  },
  emits: ["update:modelValue"],
  setup(n, { emit: h }) {
    const l = n, o = q({
      get: () => l.modelValue,
      set: (g) => h("update:modelValue", g)
    }), a = q(() => {
      var C;
      if (!r.value)
        return {};
      const { required: g, optional: y, implements: L } = r.value.manifest.service, F = {};
      for (const m of [...g, ...y, ...L])
        F[m] = (C = k.services) == null ? void 0 : C[m];
      return F;
    }), e = q(() => {
      const { label: g, target: y } = l.current, L = y || g;
      if (L.includes("/")) {
        const [F, C] = L.split("/");
        return [`${F}/koishi-plugin-${C}`].find((m) => m in k.packages);
      }
      return [
        `@koishijs/plugin-${L}`,
        `koishi-plugin-${L}`
      ].find((F) => F in k.packages);
    }), r = q(() => k.packages[e.value]), c = q(() => k.market.data[e.value]), f = q(() => Pr.value[e.value]), p = q(() => r.value.workspace ? "\uFF0C\u8BF7\u68C0\u67E5\u6E90\u4EE3\u7801" : "\uFF0C\u8BF7\u8054\u7CFB\u63D2\u4EF6\u4F5C\u8005"), d = q(() => {
      var g;
      if (!(!((g = c.value) != null && g.versions) || r.value.workspace))
        return c.value.version !== r.value.version;
    });
    function _() {
      fe.push("/market?keyword=" + l.current.label);
    }
    return Se("manager.settings.local", r), Se("manager.settings.config", o), Se("manager.settings.current", q(() => l.current)), (g, y) => {
      const L = M("k-comment"), F = M("router-link"), C = M("k-slot"), m = M("k-form");
      return w(e) ? (E(), R(j, { key: 0 }, [
        w(r).id && !w(r).forkable && n.current.disabled ? (E(), T(L, {
          key: 0,
          type: "warning"
        }, {
          default: V(() => [
            Es
          ]),
          _: 1
        })) : Q("", !0),
        w(d) ? (E(), T(L, { key: 1 }, {
          default: V(() => [
            $("p", null, [
              O("\u5F53\u524D\u7684\u63D2\u4EF6\u7248\u672C\u4E0D\u662F\u6700\u65B0\uFF0C"),
              G(F, { to: "/dependencies" }, {
                default: V(() => [
                  O("\u70B9\u51FB\u524D\u5F80\u4F9D\u8D56\u7BA1\u7406")
                ]),
                _: 1
              }),
              O("\u3002")
            ])
          ]),
          _: 1
        })) : Q("", !0),
        !w(r).workspace && w(k).dependencies && !w(k).dependencies[w(e)] ? (E(), T(L, {
          key: 2,
          type: "warning"
        }, {
          default: V(() => [
            $("p", null, [
              O("\u5C1A\u672A\u5C06\u5F53\u524D\u63D2\u4EF6\u5217\u5165\u4F9D\u8D56\uFF0C"),
              $("a", {
                onClick: y[0] || (y[0] = (v) => w(ue)("market/install", { [w(e)]: w(r).version }))
              }, "\u70B9\u51FB\u6DFB\u52A0"),
              O("\u3002")
            ])
          ]),
          _: 1
        })) : Q("", !0),
        (E(!0), R(j, null, te(w(f).impl, (v) => (E(), R(j, { key: v }, [
          w(a)[v] && n.current.disabled ? (E(), T(L, {
            key: 0,
            type: "warning"
          }, {
            default: V(() => [
              $("p", null, "\u6B64\u63D2\u4EF6\u5C06\u4F1A\u63D0\u4F9B " + H(v) + " \u670D\u52A1\uFF0C\u4F46\u6B64\u670D\u52A1\u5DF2\u88AB\u5176\u4ED6\u63D2\u4EF6\u5B9E\u73B0\u3002", 1)
            ]),
            _: 2
          }, 1024)) : (E(), T(L, {
            key: 1,
            type: n.current.disabled ? "primary" : "success"
          }, {
            default: V(() => [
              $("p", null, "\u6B64\u63D2\u4EF6" + H(n.current.disabled ? "\u5C06\u4F1A\u63D0\u4F9B" : "\u63D0\u4F9B\u4E86") + " " + H(v) + " \u670D\u52A1\u3002", 1)
            ]),
            _: 2
          }, 1032, ["type"]))
        ], 64))), 128)),
        (E(!0), R(j, null, te(w(f).using, ({ required: v, available: s }, t) => (E(), T(L, {
          key: t,
          type: w(a)[t] ? "success" : v ? "warning" : "primary"
        }, {
          default: V(() => [
            $("p", null, H(v ? "\u5FC5\u9700" : "\u53EF\u9009") + "\u670D\u52A1\uFF1A" + H(t) + " " + H(w(a)[t] ? "(\u5DF2\u52A0\u8F7D)" : "(\u672A\u52A0\u8F7D\uFF0C\u542F\u7528\u4E0B\u5217\u4EFB\u4E00\u63D2\u4EF6\u53EF\u5B9E\u73B0\u6B64\u670D\u52A1)"), 1),
            w(a)[t] ? Q("", !0) : (E(), R("ul", $s, [
              (E(!0), R(j, null, te(s, (i) => (E(), R("li", null, [
                G(Cs, { name: i }, null, 8, ["name"]),
                O(" (\u70B9\u51FB" + H(i in w(k).packages ? "\u914D\u7F6E" : "\u6DFB\u52A0") + ") ", 1)
              ]))), 256))
            ]))
          ]),
          _: 2
        }, 1032, ["type"]))), 128)),
        G(C, { name: "market-settings" }),
        w(r).schema ? Q("", !0) : (E(), T(L, {
          key: 3,
          type: "warning"
        }, {
          default: V(() => [
            $("p", null, "\u6B64\u63D2\u4EF6\u672A\u58F0\u660E\u914D\u7F6E\u9879\uFF0C\u8FD9\u53EF\u80FD\u5E76\u975E\u9884\u671F\u884C\u4E3A" + H(w(p)) + "\u3002", 1)
          ]),
          _: 1
        })),
        G(m, {
          schema: w(r).schema,
          initial: n.current.config,
          modelValue: w(o),
          "onUpdate:modelValue": y[1] || (y[1] = (v) => de(o) ? o.value = v : null)
        }, {
          hint: V(() => [
            O(H(w(p)), 1)
          ]),
          _: 1
        }, 8, ["schema", "initial", "modelValue"])
      ], 64)) : n.current.label ? (E(), T(L, {
        key: 1,
        type: "error"
      }, {
        default: V(() => [
          $("p", null, [
            O("\u6B64\u63D2\u4EF6\u5C1A\u672A\u5B89\u88C5\uFF0C"),
            $("span", {
              class: "link",
              onClick: he(_, ["stop"])
            }, "\u70B9\u51FB\u524D\u5F80\u63D2\u4EF6\u5E02\u573A", 8, ys),
            O("\u3002")
          ])
        ]),
        _: 1
      })) : Q("", !0);
    };
  }
});
const ws = /* @__PURE__ */ ne({
  __name: "alias",
  props: {
    current: null
  },
  setup(n) {
    const h = n, l = ie();
    pe(() => h.current.alias, (e) => {
      l.value = e;
    }, { immediate: !0 });
    const o = q(() => {
      if (!l.value && h.current.label !== "group")
        return !1;
      if (l.value.match(/[:~+#?@&*]/))
        return !0;
      for (const e in oe.value.paths) {
        if (e === h.current.path)
          continue;
        if (oe.value.paths[e].alias === l.value)
          return !0;
      }
    });
    function a() {
      if (l.value === h.current.alias || o.value)
        return;
      h.current.alias = l.value, ue("manager/alias", h.current.path, l.value);
      const e = h.current.path, r = e.split("/"), c = r.pop();
      r.push(c.split(":", 1)[0] + (l.value ? ":" : "") + l.value), Gr(e, r.join("/"));
    }
    return (e, r) => (E(), R("span", {
      class: ge(["k-alias", { invalid: w(o) }])
    }, [
      O(" @ "),
      Sr($("input", {
        "onUpdate:modelValue": r[0] || (r[0] = (c) => l.value = c),
        onBlur: a
      }, null, 544), [
        [Vr, l.value]
      ])
    ], 2));
  }
});
const kr = /* @__PURE__ */ X(ws, [["__scopeId", "data-v-74100284"]]), Fs = { class: "label" }, Rs = /* @__PURE__ */ ne({
  __name: "index",
  setup(n) {
    const h = Nr(), l = sn(), o = q({
      get() {
        const d = h.path.slice(9);
        return d in oe.value.paths ? d : "@global";
      },
      set(d) {
        d in oe.value.paths || (d = "@global"), l.replace("/plugins/" + d);
      }
    }), a = ie(), e = ie();
    pe(() => oe.value.paths[o.value], (d) => {
      a.value = d, e.value = xr(d.config);
    }, { immediate: !0 });
    const r = q(() => {
      const { label: d, target: _ } = a.value, g = _ || d;
      if (g.includes("/")) {
        const [y, L] = g.split("/");
        return [`${y}/koishi-plugin-${L}`].find((F) => F in k.packages);
      }
      return [
        `@koishijs/plugin-${g}`,
        `koishi-plugin-${g}`
      ].find((y) => y in k.packages);
    }), c = q(() => {
      var _;
      const d = Pr.value[r.value];
      if (!!d) {
        if (d.warning && a.value.disabled)
          return "warning";
        for (const g in d.using)
          if ((_ = k.services) != null && _[g]) {
            if (d.impl.includes(g))
              return "warning";
          } else if (d.using[g].required)
            return "warning";
      }
    }), f = q(() => {
      const d = a.value.path === "@global", _ = a.value.children, g = !d && !_, y = a.value.disabled;
      return [{
        icon: y ? "play" : "stop",
        label: y ? "\u542F\u7528\u63D2\u4EF6" : "\u505C\u7528\u63D2\u4EF6",
        disabled: !g || !r.value,
        action: async () => {
          await p(y ? "reload" : "unload"), me.success(y ? "\u63D2\u4EF6\u5DF2\u542F\u7528\u3002" : "\u63D2\u4EF6\u5DF2\u505C\u7528\u3002");
        }
      }, {
        type: y ? c.value : "",
        icon: y ? "save" : "check",
        label: y ? "\u4FDD\u5B58\u914D\u7F6E" : "\u91CD\u8F7D\u914D\u7F6E",
        disabled: _ || !d && !r.value,
        async action() {
          d ? ue("manager/app-reload", e.value) : (await p(y ? "unload" : "reload"), me.success(y ? "\u914D\u7F6E\u5DF2\u4FDD\u5B58\u3002" : "\u914D\u7F6E\u5DF2\u91CD\u8F7D\u3002"));
        }
      }, {
        type: "error",
        icon: "trash-can",
        label: _ ? "\u79FB\u9664\u5206\u7EC4" : "\u79FB\u9664\u63D2\u4EF6",
        disabled: !a.value.path || d,
        action: () => os(a.value.path)
      }, {
        icon: "add-plugin",
        label: "\u6DFB\u52A0\u63D2\u4EF6",
        disabled: g,
        action: () => ye.value = !0
      }, {
        icon: "add-group",
        label: "\u6DFB\u52A0\u5206\u7EC4",
        disabled: g,
        action: () => is(a.value.path, "group", "group")
      }];
    });
    async function p(d) {
      if (await ue(`manager/${d}`, a.value.path, e.value, a.value.target), a.value.target) {
        const _ = Le(a.value.path);
        _[_.length - 1] = a.value.target, l.replace("/plugins/" + _.join("/"));
      }
    }
    return (d, _) => {
      const g = M("el-option"), y = M("el-select"), L = M("k-content"), F = M("k-layout");
      return E(), T(F, { menu: w(f) }, {
        header: V(() => [
          a.value.path === "@global" || !a.value.path ? (E(), R(j, { key: 0 }, [
            O(H(a.value.label), 1)
          ], 64)) : a.value.children ? (E(), R(j, { key: 1 }, [
            O(" \u5206\u7EC4 "),
            G(kr, { current: a.value }, null, 8, ["current"])
          ], 64)) : (E(), R(j, { key: 2 }, [
            a.value.label ? (E(), R(j, { key: 1 }, [
              $("span", Fs, H(a.value.label), 1),
              G(kr, { current: a.value }, null, 8, ["current"])
            ], 64)) : (E(), T(y, {
              key: 0,
              modelValue: a.value.target,
              "onUpdate:modelValue": _[0] || (_[0] = (C) => a.value.target = C),
              filterable: "",
              placeholder: "\u63D2\u4EF6\u9009\u62E9"
            }, {
              default: V(() => [
                (E(!0), R(j, null, te(Object.values(w(k).packages).slice(1).map((C) => C.shortname).sort(), (C) => (E(), T(g, {
                  key: C,
                  label: C,
                  value: C
                }, null, 8, ["label", "value"]))), 128))
              ]),
              _: 1
            }, 8, ["modelValue"]))
          ], 64))
        ]),
        left: V(() => [
          G(gs, {
            modelValue: w(o),
            "onUpdate:modelValue": _[1] || (_[1] = (C) => de(o) ? o.value = C : null)
          }, null, 8, ["modelValue"])
        ]),
        default: V(() => [
          (E(), T(L, {
            class: "plugin-view",
            key: w(o)
          }, {
            default: V(() => [
              a.value.path === "@global" ? (E(), T(as, {
                key: 0,
                current: a.value,
                modelValue: e.value,
                "onUpdate:modelValue": _[2] || (_[2] = (C) => e.value = C)
              }, null, 8, ["current", "modelValue"])) : a.value.children ? (E(), T(ds, {
                key: 1,
                modelValue: e.value,
                "onUpdate:modelValue": _[3] || (_[3] = (C) => e.value = C),
                current: a.value
              }, null, 8, ["modelValue", "current"])) : (E(), T(Ls, {
                key: 2,
                current: a.value,
                modelValue: e.value,
                "onUpdate:modelValue": _[4] || (_[4] = (C) => e.value = C)
              }, null, 8, ["current", "modelValue"]))
            ]),
            _: 1
          }))
        ]),
        _: 1
      }, 8, ["menu"]);
    };
  }
});
const Dt = {
  all: "\u6240\u6709\u63D2\u4EF6",
  other: "\u672A\u5206\u7C7B",
  console: "\u63A7\u5236\u53F0",
  business: "\u4E1A\u52A1\u529F\u80FD",
  storage: "\u5B58\u50A8\u670D\u52A1",
  tool: "\u57FA\u7840\u5DE5\u5177",
  game: "\u5A31\u4E50\u529F\u80FD"
};
function Ur(n) {
  return Dt[n] ? n : "other";
}
function As(n, h) {
  const { locales: l, service: o } = zr(n.name).manifest;
  if (h.startsWith("impl:"))
    return o.implements.includes(h.slice(5));
  if (h.startsWith("locale:"))
    return l.includes(h.slice(7));
  if (h.startsWith("using:")) {
    const a = h.slice(6);
    return o.required.includes(a) || o.optional.includes(a);
  } else {
    if (h.startsWith("email:"))
      return n.maintainers.some((a) => a.email === h.slice(6));
    if (h.startsWith("is:"))
      return h === "is:verified" ? n.verified : !0;
  }
  return n.shortname.includes(h) ? !0 : n.keywords.some((a) => !a.includes(":") && a.includes(h));
}
var At = { exports: {} }, Ir;
function Ds() {
  return Ir || (Ir = 1, function(n, h) {
    (function(l) {
      n.exports = l();
    })(function(l) {
      var o = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f"];
      function a(m, v) {
        var s = m[0], t = m[1], i = m[2], u = m[3];
        s += (t & i | ~t & u) + v[0] - 680876936 | 0, s = (s << 7 | s >>> 25) + t | 0, u += (s & t | ~s & i) + v[1] - 389564586 | 0, u = (u << 12 | u >>> 20) + s | 0, i += (u & s | ~u & t) + v[2] + 606105819 | 0, i = (i << 17 | i >>> 15) + u | 0, t += (i & u | ~i & s) + v[3] - 1044525330 | 0, t = (t << 22 | t >>> 10) + i | 0, s += (t & i | ~t & u) + v[4] - 176418897 | 0, s = (s << 7 | s >>> 25) + t | 0, u += (s & t | ~s & i) + v[5] + 1200080426 | 0, u = (u << 12 | u >>> 20) + s | 0, i += (u & s | ~u & t) + v[6] - 1473231341 | 0, i = (i << 17 | i >>> 15) + u | 0, t += (i & u | ~i & s) + v[7] - 45705983 | 0, t = (t << 22 | t >>> 10) + i | 0, s += (t & i | ~t & u) + v[8] + 1770035416 | 0, s = (s << 7 | s >>> 25) + t | 0, u += (s & t | ~s & i) + v[9] - 1958414417 | 0, u = (u << 12 | u >>> 20) + s | 0, i += (u & s | ~u & t) + v[10] - 42063 | 0, i = (i << 17 | i >>> 15) + u | 0, t += (i & u | ~i & s) + v[11] - 1990404162 | 0, t = (t << 22 | t >>> 10) + i | 0, s += (t & i | ~t & u) + v[12] + 1804603682 | 0, s = (s << 7 | s >>> 25) + t | 0, u += (s & t | ~s & i) + v[13] - 40341101 | 0, u = (u << 12 | u >>> 20) + s | 0, i += (u & s | ~u & t) + v[14] - 1502002290 | 0, i = (i << 17 | i >>> 15) + u | 0, t += (i & u | ~i & s) + v[15] + 1236535329 | 0, t = (t << 22 | t >>> 10) + i | 0, s += (t & u | i & ~u) + v[1] - 165796510 | 0, s = (s << 5 | s >>> 27) + t | 0, u += (s & i | t & ~i) + v[6] - 1069501632 | 0, u = (u << 9 | u >>> 23) + s | 0, i += (u & t | s & ~t) + v[11] + 643717713 | 0, i = (i << 14 | i >>> 18) + u | 0, t += (i & s | u & ~s) + v[0] - 373897302 | 0, t = (t << 20 | t >>> 12) + i | 0, s += (t & u | i & ~u) + v[5] - 701558691 | 0, s = (s << 5 | s >>> 27) + t | 0, u += (s & i | t & ~i) + v[10] + 38016083 | 0, u = (u << 9 | u >>> 23) + s | 0, i += (u & t | s & ~t) + v[15] - 660478335 | 0, i = (i << 14 | i >>> 18) + u | 0, t += (i & s | u & ~s) + v[4] - 405537848 | 0, t = (t << 20 | t >>> 12) + i | 0, s += (t & u | i & ~u) + v[9] + 568446438 | 0, s = (s << 5 | s >>> 27) + t | 0, u += (s & i | t & ~i) + v[14] - 1019803690 | 0, u = (u << 9 | u >>> 23) + s | 0, i += (u & t | s & ~t) + v[3] - 187363961 | 0, i = (i << 14 | i >>> 18) + u | 0, t += (i & s | u & ~s) + v[8] + 1163531501 | 0, t = (t << 20 | t >>> 12) + i | 0, s += (t & u | i & ~u) + v[13] - 1444681467 | 0, s = (s << 5 | s >>> 27) + t | 0, u += (s & i | t & ~i) + v[2] - 51403784 | 0, u = (u << 9 | u >>> 23) + s | 0, i += (u & t | s & ~t) + v[7] + 1735328473 | 0, i = (i << 14 | i >>> 18) + u | 0, t += (i & s | u & ~s) + v[12] - 1926607734 | 0, t = (t << 20 | t >>> 12) + i | 0, s += (t ^ i ^ u) + v[5] - 378558 | 0, s = (s << 4 | s >>> 28) + t | 0, u += (s ^ t ^ i) + v[8] - 2022574463 | 0, u = (u << 11 | u >>> 21) + s | 0, i += (u ^ s ^ t) + v[11] + 1839030562 | 0, i = (i << 16 | i >>> 16) + u | 0, t += (i ^ u ^ s) + v[14] - 35309556 | 0, t = (t << 23 | t >>> 9) + i | 0, s += (t ^ i ^ u) + v[1] - 1530992060 | 0, s = (s << 4 | s >>> 28) + t | 0, u += (s ^ t ^ i) + v[4] + 1272893353 | 0, u = (u << 11 | u >>> 21) + s | 0, i += (u ^ s ^ t) + v[7] - 155497632 | 0, i = (i << 16 | i >>> 16) + u | 0, t += (i ^ u ^ s) + v[10] - 1094730640 | 0, t = (t << 23 | t >>> 9) + i | 0, s += (t ^ i ^ u) + v[13] + 681279174 | 0, s = (s << 4 | s >>> 28) + t | 0, u += (s ^ t ^ i) + v[0] - 358537222 | 0, u = (u << 11 | u >>> 21) + s | 0, i += (u ^ s ^ t) + v[3] - 722521979 | 0, i = (i << 16 | i >>> 16) + u | 0, t += (i ^ u ^ s) + v[6] + 76029189 | 0, t = (t << 23 | t >>> 9) + i | 0, s += (t ^ i ^ u) + v[9] - 640364487 | 0, s = (s << 4 | s >>> 28) + t | 0, u += (s ^ t ^ i) + v[12] - 421815835 | 0, u = (u << 11 | u >>> 21) + s | 0, i += (u ^ s ^ t) + v[15] + 530742520 | 0, i = (i << 16 | i >>> 16) + u | 0, t += (i ^ u ^ s) + v[2] - 995338651 | 0, t = (t << 23 | t >>> 9) + i | 0, s += (i ^ (t | ~u)) + v[0] - 198630844 | 0, s = (s << 6 | s >>> 26) + t | 0, u += (t ^ (s | ~i)) + v[7] + 1126891415 | 0, u = (u << 10 | u >>> 22) + s | 0, i += (s ^ (u | ~t)) + v[14] - 1416354905 | 0, i = (i << 15 | i >>> 17) + u | 0, t += (u ^ (i | ~s)) + v[5] - 57434055 | 0, t = (t << 21 | t >>> 11) + i | 0, s += (i ^ (t | ~u)) + v[12] + 1700485571 | 0, s = (s << 6 | s >>> 26) + t | 0, u += (t ^ (s | ~i)) + v[3] - 1894986606 | 0, u = (u << 10 | u >>> 22) + s | 0, i += (s ^ (u | ~t)) + v[10] - 1051523 | 0, i = (i << 15 | i >>> 17) + u | 0, t += (u ^ (i | ~s)) + v[1] - 2054922799 | 0, t = (t << 21 | t >>> 11) + i | 0, s += (i ^ (t | ~u)) + v[8] + 1873313359 | 0, s = (s << 6 | s >>> 26) + t | 0, u += (t ^ (s | ~i)) + v[15] - 30611744 | 0, u = (u << 10 | u >>> 22) + s | 0, i += (s ^ (u | ~t)) + v[6] - 1560198380 | 0, i = (i << 15 | i >>> 17) + u | 0, t += (u ^ (i | ~s)) + v[13] + 1309151649 | 0, t = (t << 21 | t >>> 11) + i | 0, s += (i ^ (t | ~u)) + v[4] - 145523070 | 0, s = (s << 6 | s >>> 26) + t | 0, u += (t ^ (s | ~i)) + v[11] - 1120210379 | 0, u = (u << 10 | u >>> 22) + s | 0, i += (s ^ (u | ~t)) + v[2] + 718787259 | 0, i = (i << 15 | i >>> 17) + u | 0, t += (u ^ (i | ~s)) + v[9] - 343485551 | 0, t = (t << 21 | t >>> 11) + i | 0, m[0] = s + m[0] | 0, m[1] = t + m[1] | 0, m[2] = i + m[2] | 0, m[3] = u + m[3] | 0;
      }
      function e(m) {
        var v = [], s;
        for (s = 0; s < 64; s += 4)
          v[s >> 2] = m.charCodeAt(s) + (m.charCodeAt(s + 1) << 8) + (m.charCodeAt(s + 2) << 16) + (m.charCodeAt(s + 3) << 24);
        return v;
      }
      function r(m) {
        var v = [], s;
        for (s = 0; s < 64; s += 4)
          v[s >> 2] = m[s] + (m[s + 1] << 8) + (m[s + 2] << 16) + (m[s + 3] << 24);
        return v;
      }
      function c(m) {
        var v = m.length, s = [1732584193, -271733879, -1732584194, 271733878], t, i, u, b, P, W;
        for (t = 64; t <= v; t += 64)
          a(s, e(m.substring(t - 64, t)));
        for (m = m.substring(t - 64), i = m.length, u = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], t = 0; t < i; t += 1)
          u[t >> 2] |= m.charCodeAt(t) << (t % 4 << 3);
        if (u[t >> 2] |= 128 << (t % 4 << 3), t > 55)
          for (a(s, u), t = 0; t < 16; t += 1)
            u[t] = 0;
        return b = v * 8, b = b.toString(16).match(/(.*?)(.{0,8})$/), P = parseInt(b[2], 16), W = parseInt(b[1], 16) || 0, u[14] = P, u[15] = W, a(s, u), s;
      }
      function f(m) {
        var v = m.length, s = [1732584193, -271733879, -1732584194, 271733878], t, i, u, b, P, W;
        for (t = 64; t <= v; t += 64)
          a(s, r(m.subarray(t - 64, t)));
        for (m = t - 64 < v ? m.subarray(t - 64) : new Uint8Array(0), i = m.length, u = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], t = 0; t < i; t += 1)
          u[t >> 2] |= m[t] << (t % 4 << 3);
        if (u[t >> 2] |= 128 << (t % 4 << 3), t > 55)
          for (a(s, u), t = 0; t < 16; t += 1)
            u[t] = 0;
        return b = v * 8, b = b.toString(16).match(/(.*?)(.{0,8})$/), P = parseInt(b[2], 16), W = parseInt(b[1], 16) || 0, u[14] = P, u[15] = W, a(s, u), s;
      }
      function p(m) {
        var v = "", s;
        for (s = 0; s < 4; s += 1)
          v += o[m >> s * 8 + 4 & 15] + o[m >> s * 8 & 15];
        return v;
      }
      function d(m) {
        var v;
        for (v = 0; v < m.length; v += 1)
          m[v] = p(m[v]);
        return m.join("");
      }
      d(c("hello")), typeof ArrayBuffer < "u" && !ArrayBuffer.prototype.slice && function() {
        function m(v, s) {
          return v = v | 0 || 0, v < 0 ? Math.max(v + s, 0) : Math.min(v, s);
        }
        ArrayBuffer.prototype.slice = function(v, s) {
          var t = this.byteLength, i = m(v, t), u = t, b, P, W, ce;
          return s !== l && (u = m(s, t)), i > u ? new ArrayBuffer(0) : (b = u - i, P = new ArrayBuffer(b), W = new Uint8Array(P), ce = new Uint8Array(this, i, b), W.set(ce), P);
        };
      }();
      function _(m) {
        return /[\u0080-\uFFFF]/.test(m) && (m = unescape(encodeURIComponent(m))), m;
      }
      function g(m, v) {
        var s = m.length, t = new ArrayBuffer(s), i = new Uint8Array(t), u;
        for (u = 0; u < s; u += 1)
          i[u] = m.charCodeAt(u);
        return v ? i : t;
      }
      function y(m) {
        return String.fromCharCode.apply(null, new Uint8Array(m));
      }
      function L(m, v, s) {
        var t = new Uint8Array(m.byteLength + v.byteLength);
        return t.set(new Uint8Array(m)), t.set(new Uint8Array(v), m.byteLength), s ? t : t.buffer;
      }
      function F(m) {
        var v = [], s = m.length, t;
        for (t = 0; t < s - 1; t += 2)
          v.push(parseInt(m.substr(t, 2), 16));
        return String.fromCharCode.apply(String, v);
      }
      function C() {
        this.reset();
      }
      return C.prototype.append = function(m) {
        return this.appendBinary(_(m)), this;
      }, C.prototype.appendBinary = function(m) {
        this._buff += m, this._length += m.length;
        var v = this._buff.length, s;
        for (s = 64; s <= v; s += 64)
          a(this._hash, e(this._buff.substring(s - 64, s)));
        return this._buff = this._buff.substring(s - 64), this;
      }, C.prototype.end = function(m) {
        var v = this._buff, s = v.length, t, i = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], u;
        for (t = 0; t < s; t += 1)
          i[t >> 2] |= v.charCodeAt(t) << (t % 4 << 3);
        return this._finish(i, s), u = d(this._hash), m && (u = F(u)), this.reset(), u;
      }, C.prototype.reset = function() {
        return this._buff = "", this._length = 0, this._hash = [1732584193, -271733879, -1732584194, 271733878], this;
      }, C.prototype.getState = function() {
        return {
          buff: this._buff,
          length: this._length,
          hash: this._hash.slice()
        };
      }, C.prototype.setState = function(m) {
        return this._buff = m.buff, this._length = m.length, this._hash = m.hash, this;
      }, C.prototype.destroy = function() {
        delete this._hash, delete this._buff, delete this._length;
      }, C.prototype._finish = function(m, v) {
        var s = v, t, i, u;
        if (m[s >> 2] |= 128 << (s % 4 << 3), s > 55)
          for (a(this._hash, m), s = 0; s < 16; s += 1)
            m[s] = 0;
        t = this._length * 8, t = t.toString(16).match(/(.*?)(.{0,8})$/), i = parseInt(t[2], 16), u = parseInt(t[1], 16) || 0, m[14] = i, m[15] = u, a(this._hash, m);
      }, C.hash = function(m, v) {
        return C.hashBinary(_(m), v);
      }, C.hashBinary = function(m, v) {
        var s = c(m), t = d(s);
        return v ? F(t) : t;
      }, C.ArrayBuffer = function() {
        this.reset();
      }, C.ArrayBuffer.prototype.append = function(m) {
        var v = L(this._buff.buffer, m, !0), s = v.length, t;
        for (this._length += m.byteLength, t = 64; t <= s; t += 64)
          a(this._hash, r(v.subarray(t - 64, t)));
        return this._buff = t - 64 < s ? new Uint8Array(v.buffer.slice(t - 64)) : new Uint8Array(0), this;
      }, C.ArrayBuffer.prototype.end = function(m) {
        var v = this._buff, s = v.length, t = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], i, u;
        for (i = 0; i < s; i += 1)
          t[i >> 2] |= v[i] << (i % 4 << 3);
        return this._finish(t, s), u = d(this._hash), m && (u = F(u)), this.reset(), u;
      }, C.ArrayBuffer.prototype.reset = function() {
        return this._buff = new Uint8Array(0), this._length = 0, this._hash = [1732584193, -271733879, -1732584194, 271733878], this;
      }, C.ArrayBuffer.prototype.getState = function() {
        var m = C.prototype.getState.call(this);
        return m.buff = y(m.buff), m;
      }, C.ArrayBuffer.prototype.setState = function(m) {
        return m.buff = g(m.buff, !0), C.prototype.setState.call(this, m);
      }, C.ArrayBuffer.prototype.destroy = C.prototype.destroy, C.ArrayBuffer.prototype._finish = C.prototype._finish, C.ArrayBuffer.hash = function(m, v) {
        var s = f(new Uint8Array(m)), t = d(s);
        return v ? F(t) : t;
      }, C;
    });
  }(At)), At.exports;
}
var bs = Ds();
const ks = { class: "k-card market-view" }, Is = { class: "header" }, Ss = { class: "left" }, Vs = { class: "right" }, xs = ["href"], Ns = { class: "verified" }, Bs = { class: "rating" }, Os = { class: "footer" }, Ts = ["href"], Ms = /* @__PURE__ */ $("span", { class: "spacer" }, null, -1), qs = ["href"], zs = /* @__PURE__ */ $("span", { class: "spacer" }, null, -1), Hs = /* @__PURE__ */ $("span", { class: "spacer" }, null, -1), Ps = /* @__PURE__ */ $("span", { class: "spacer grow" }, null, -1), js = { class: "avatars" }, Gs = ["onClick"], Us = ["src"], Xs = /* @__PURE__ */ ne({
  __name: "package",
  props: {
    data: {}
  },
  emits: ["query", "click"],
  setup(n) {
    const h = n, l = q(() => zr(h.data.name));
    function o() {
      J.value = h.data.name;
    }
    const a = q(() => Math.min(Math.max((h.data.score.final - 0.25) * 10, 0), 5));
    function e(f) {
      return (k.market.gravatar || "https://s.gravatar.com") + "/avatar/" + bs.hash(f) + "?d=mp";
    }
    function r(f) {
      return f >= 100 ? +f.toFixed() : +f.toFixed(1);
    }
    function c(f) {
      return f >= (1 << 20) * 1e3 ? r(f / (1 << 30)) + " GB" : f >= (1 << 10) * 1e3 ? r(f / (1 << 20)) + " MB" : r(f / (1 << 10)) + " KB";
    }
    return (f, p) => {
      const d = M("k-icon"), _ = M("el-tooltip"), g = M("k-button"), y = M("k-markdown");
      return E(), R("section", ks, [
        $("div", Is, [
          $("div", Ss, [
            G(d, {
              name: "category:" + w(Ur)(n.data.manifest.category)
            }, null, 8, ["name"])
          ]),
          $("div", Vs, [
            $("h2", null, [
              $("a", {
                href: n.data.links.homepage,
                target: "_blank",
                rel: "noopener noreferrer"
              }, H(n.data.shortname), 9, xs),
              n.data.verified ? (E(), T(_, {
                key: 0,
                content: "\u5B98\u65B9\u8BA4\u8BC1",
                placement: "right"
              }, {
                default: V(() => [
                  $("span", Ns, [
                    G(d, { name: "verified" })
                  ])
                ]),
                _: 1
              })) : Q("", !0)
            ]),
            G(_, {
              content: w(a).toFixed(1),
              placement: "right"
            }, {
              default: V(() => [
                $("div", Bs, [
                  (E(!0), R(j, null, te(Array(5).fill(null), (L, F) => (E(), T(d, {
                    key: F,
                    name: F + 0.5 < w(a) ? "star-full" : "star-empty"
                  }, null, 8, ["name"]))), 128))
                ])
              ]),
              _: 1
            }, 8, ["content"]),
            w(k).packages ? (E(), R(j, { key: 0 }, [
              w(k).packages[n.data.name] ? (E(), T(g, {
                key: 0,
                type: "success",
                solid: "",
                class: "right",
                onClick: o
              }, {
                default: V(() => [
                  O("\u4FEE\u6539")
                ]),
                _: 1
              })) : (E(), T(g, {
                key: 1,
                disabled: w(rn).static,
                solid: "",
                class: "right",
                onClick: o
              }, {
                default: V(() => [
                  O("\u6DFB\u52A0")
                ]),
                _: 1
              }, 8, ["disabled"]))
            ], 64)) : Q("", !0)
          ])
        ]),
        G(y, {
          inline: "",
          class: "desc",
          source: w(l).manifest.description.zh || w(l).manifest.description.en
        }, null, 8, ["source"]),
        $("div", Os, [
          $("a", {
            class: "shrink",
            href: n.data.links.npm,
            target: "_blank",
            rel: "noopener noreferrer"
          }, [
            G(d, { name: "tag" }),
            O(H(n.data.version), 1)
          ], 8, Ts),
          n.data.installSize ? (E(), R(j, { key: 0 }, [
            Ms,
            $("a", {
              href: n.data.links.size,
              target: "_blank",
              rel: "noopener noreferrer"
            }, [
              G(d, { name: "file-archive" }),
              O(H(c(n.data.installSize)), 1)
            ], 8, qs)
          ], 64)) : Q("", !0),
          n.data.downloads ? (E(), R(j, { key: 1 }, [
            zs,
            $("span", null, [
              G(d, { name: "download" }),
              O(H(n.data.downloads.lastMonth), 1)
            ])
          ], 64)) : Q("", !0),
          !n.data.installSize && !n.data.downloads ? (E(), R(j, { key: 2 }, [
            Hs,
            $("span", null, [
              G(d, { name: "balance" }),
              O(H(n.data.license), 1)
            ])
          ], 64)) : Q("", !0),
          Ps,
          $("div", js, [
            (E(!0), R(j, null, te(n.data.maintainers, ({ email: L, username: F }) => (E(), T(_, {
              key: L,
              content: F
            }, {
              default: V(() => [
                $("a", {
                  onClick: (C) => f.$emit("query", "email:" + L)
                }, [
                  $("img", {
                    src: e(L)
                  }, null, 8, Us)
                ], 8, Gs)
              ]),
              _: 2
            }, 1032, ["content"]))), 128))
          ])
        ])
      ]);
    };
  }
});
const Ws = { key: 0 }, Ys = /* @__PURE__ */ $("div", { class: "el-loading-spinner" }, [
  /* @__PURE__ */ $("svg", {
    class: "circular",
    viewBox: "25 25 50 50"
  }, [
    /* @__PURE__ */ $("circle", {
      class: "path",
      cx: "50",
      cy: "50",
      r: "20",
      fill: "none"
    })
  ]),
  /* @__PURE__ */ $("p", { class: "el-loading-text" }, "\u6B63\u5728\u52A0\u8F7D\u63D2\u4EF6\u5E02\u573A\u2026\u2026")
], -1), Ks = [
  Ys
], Zs = { class: "search-box" }, Qs = ["onKeydown", "onKeypress"], Js = { class: "market-filter" }, e1 = { class: "market-container" }, t1 = /* @__PURE__ */ $("p", null, "\u65E0\u6CD5\u8FDE\u63A5\u5230\u63D2\u4EF6\u5E02\u573A\u3002\u8FD9\u53EF\u80FD\u662F\u4EE5\u4E0B\u539F\u56E0\u5BFC\u81F4\u7684\uFF1A", -1), r1 = /* @__PURE__ */ $("ul", null, [
  /* @__PURE__ */ $("li", null, "\u65E0\u6CD5\u8FDE\u63A5\u5230\u7F51\u7EDC\uFF0C\u8BF7\u68C0\u67E5\u4F60\u7684\u7F51\u7EDC\u8FDE\u63A5\u548C\u4EE3\u7406\u8BBE\u7F6E"),
  /* @__PURE__ */ $("li", null, "\u60A8\u6240\u7528\u7684 registry \u4E0D\u652F\u6301\u641C\u7D22\u529F\u80FD (\u5982 npmmirror)\uFF0C\u8BF7\u8003\u8651\u8FDB\u884C\u66F4\u6362")
], -1), n1 = /* @__PURE__ */ ne({
  __name: "index",
  setup(n) {
    const h = Nr(), { keyword: l } = h.query, o = Qr(Array.isArray(l) ? l : (l || "").split(" "));
    o[o.length - 1] && o.push("");
    const a = q(() => o.filter((_) => _));
    pe(o, () => {
      const { keyword: _, ...g } = h.query, y = a.value.join(" ");
      y ? fe.replace({ query: { keyword: y, ...g } }) : fe.replace({ query: g });
    }, { deep: !0 });
    function e(_) {
      const g = o[o.length - 1];
      !g || (o.slice(0, -1).includes(g) && o.pop(), o.push(""));
    }
    function r(_) {
      o[o.length - 1] = "";
    }
    function c(_) {
      o[o.length - 1] === "" && o.length > 1 && (_.preventDefault(), o.splice(o.length - 2, 1));
    }
    function f(_) {
      o[o.length - 1] || o.pop(), o.includes(_) || o.push(_), o.push("");
    }
    const p = q(() => Object.values(k.market.data).filter((_) => o.every((g) => As(_, g))).filter((_) => xe.static ? !ee.showInstalled || k.packages[_.name] : ee.showInstalled || !k.packages[_.name]).sort((_, g) => (xe.static ? +g.portable - +_.portable : 0) || g.score.final - _.score.final)), d = q(() => [Hr.value]);
    return (_, g) => {
      const y = M("k-badge"), L = M("k-icon"), F = M("el-checkbox"), C = M("el-scrollbar"), m = M("k-comment"), v = M("k-layout");
      return E(), T(v, {
        main: "darker",
        class: "page-market",
        menu: w(d)
      }, {
        default: V(() => [
          w(k).market ? w(k).market.total ? (E(), T(C, { key: 1 }, {
            default: V(() => [
              $("div", Zs, [
                (E(!0), R(j, null, te(o.slice(0, -1), (s, t) => (E(), T(y, {
                  type: "success",
                  key: t,
                  onClick: (i) => o.splice(t, 1)
                }, {
                  default: V(() => [
                    O(H(s), 1)
                  ]),
                  _: 2
                }, 1032, ["onClick"]))), 128)),
                Sr($("input", {
                  placeholder: "\u8F93\u5165\u60F3\u8981\u67E5\u8BE2\u7684\u63D2\u4EF6\u540D",
                  "onUpdate:modelValue": g[0] || (g[0] = (s) => o[o.length - 1] = s),
                  onBlur: e,
                  onKeydown: [
                    we(r, ["escape"]),
                    we(c, ["backspace"])
                  ],
                  onKeypress: [
                    we(he(e, ["prevent"]), ["enter"]),
                    we(he(e, ["prevent"]), ["space"])
                  ]
                }, null, 40, Qs), [
                  [Vr, o[o.length - 1]]
                ]),
                G(L, { name: "search" })
              ]),
              $("div", Js, [
                O(" \u5171\u641C\u7D22\u5230 " + H(w(a).length ? w(p).length + " / " : "") + H(Object.keys(w(k).market.data).length) + " \u4E2A\u63D2\u4EF6\u3002 ", 1),
                w(k).packages ? (E(), T(F, {
                  key: 0,
                  modelValue: w(ee).showInstalled,
                  "onUpdate:modelValue": g[1] || (g[1] = (s) => w(ee).showInstalled = s)
                }, {
                  default: V(() => [
                    O(H(w(xe).static ? "\u53EA\u663E\u793A\u53EF\u7528\u63D2\u4EF6" : "\u663E\u793A\u5DF2\u4E0B\u8F7D\u7684\u63D2\u4EF6"), 1)
                  ]),
                  _: 1
                }, 8, ["modelValue"])) : Q("", !0)
              ]),
              $("div", e1, [
                (E(!0), R(j, null, te(w(p), (s) => (E(), T(Xs, {
                  key: s.name,
                  data: s,
                  onQuery: f
                }, null, 8, ["data"]))), 128))
              ])
            ]),
            _: 1
          })) : (E(), T(m, {
            key: 2,
            type: "error",
            class: "market-error"
          }, {
            default: V(() => [
              t1,
              r1
            ]),
            _: 1
          })) : (E(), R("div", Ws, Ks))
        ]),
        _: 1
      }, 8, ["menu"]);
    };
  }
});
const s1 = { class: "tabs" }, u1 = ["onClick"], i1 = { class: "title" }, o1 = { class: "content" }, a1 = ["onClick"], l1 = /* @__PURE__ */ ne({
  __name: "select",
  setup(n) {
    const h = ie("all"), { width: l } = un();
    function o(a) {
      ye.value = !1, a += ":" + Math.random().toString(36).slice(2, 8), ue("manager/unload", a, {}), fe.push("/plugins/" + a);
    }
    return (a, e) => {
      const r = M("k-icon"), c = M("k-markdown"), f = M("el-scrollbar"), p = M("el-dialog");
      return E(), T(p, {
        modelValue: w(ye),
        "onUpdate:modelValue": e[0] || (e[0] = (d) => de(ye) ? ye.value = d : null),
        class: "plugin-select",
        title: w(l) <= 768 ? w(Dt)[h.value] : "\u9009\u62E9\u63D2\u4EF6"
      }, {
        default: V(() => [
          $("div", s1, [
            (E(!0), R(j, null, te(w(Dt), (d, _) => (E(), R("span", {
              class: ge(["tab-item", { active: h.value === _ }]),
              key: _,
              onClick: he((g) => h.value = _, ["stop"])
            }, [
              G(r, {
                name: "category:" + _
              }, null, 8, ["name"]),
              $("span", i1, H(d), 1)
            ], 10, u1))), 128))
          ]),
          $("div", o1, [
            G(f, null, {
              default: V(() => [
                (E(!0), R(j, null, te(w(k).packages, ({ name: d, shortname: _, manifest: g }) => (E(), R(j, null, [
                  d && (h.value === "all" || w(Ur)(g.category) === h.value) ? (E(), R("div", {
                    key: 0,
                    class: "package",
                    onClick: he((y) => o(_), ["stop"])
                  }, [
                    $("h3", null, H(_), 1),
                    G(c, {
                      inline: "",
                      class: "desc",
                      source: g.description.zh || g.description.en
                    }, null, 8, ["source"])
                  ], 8, a1)) : Q("", !0)
                ], 64))), 256))
              ]),
              _: 1
            })
          ])
        ]),
        _: 1
      }, 8, ["modelValue", "title"]);
    };
  }
});
const c1 = /* @__PURE__ */ ne({
  __name: "progress",
  setup(n) {
    const h = q(() => !k.user || k.user.authority < 4 ? !1 : !k.market || k.market.total > k.market.progress), l = q(() => k.market ? 100 * k.market.progress / k.market.total : 50);
    return (o, a) => {
      const e = M("el-progress"), r = M("k-status");
      return w(h) ? (E(), T(r, { key: 0 }, {
        default: V(() => [
          G(e, {
            indeterminate: !w(k).market,
            percentage: w(l)
          }, {
            default: V(() => [
              O(" \u6B63\u5728\u52A0\u8F7D\u63D2\u4EF6\u5E02\u573A ")
            ]),
            _: 1
          }, 8, ["indeterminate", "percentage"])
        ]),
        _: 1
      })) : Q("", !0);
    };
  }
});
const h1 = {}, f1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, d1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M320 77.23L559.6 49.48C573.7 47.85 587.2 55.65 592.7 68.66L624.7 143.3C634.6 166.2 621.3 192.4 596.1 198L437.8 234.8C418.8 239.1 399.2 231.7 387.9 215.9L320 120.9L252.1 215.9C240.8 231.7 221.2 239.1 202.3 234.8L43.04 198C18.75 192.4 5.449 166.2 15.27 143.3L47.26 68.66C52.84 55.65 66.3 47.85 80.36 49.48L320 77.23zM44.68 155.9C42.72 160.5 45.38 165.7 50.24 166.9L209.4 203.6C215.7 205.1 222.3 202.6 226.1 197.3L291.2 106.1L76.68 81.27L44.68 155.9zM413.9 197.3C417.7 202.6 424.3 205.1 430.6 203.6L589.8 166.9C594.6 165.7 597.3 160.5 595.3 155.9L563.3 81.27L348.8 106.1L413.9 197.3zM544 236.7L576 227.6V378.5C576 400.5 561 419.7 539.6 425.1L331.6 477.1C324 479 316 479 308.4 477.1L100.4 425.1C78.99 419.7 64 400.5 64 378.5V227.6L96 236.7V378.5C96 385.9 100.1 392.3 108.1 394L304 443V207.1C304 199.2 311.2 191.1 320 191.1C328.8 191.1 336 199.2 336 207.1V443L531.9 394C539 392.3 544 385.9 544 378.5L544 236.7z"
}, null, -1), p1 = [
  d1
];
function _1(n, h) {
  return E(), R("svg", f1, p1);
}
const v1 = /* @__PURE__ */ X(h1, [["render", _1]]), m1 = {}, g1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, C1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M448 0h-128c-17.67 0-32 14.33-32 32v24c0 17.67 14.33 32 32 32h16c8.836 0 16 7.162 16 16c0 8.836-7.164 16-16 16H320c-17.67 0-32 14.33-32 32V176c0 17.67 14.33 32 32 32h32C352 234.5 373.5 256 400 256C426.5 256 448 234.5 448 208h32c17.67 0 32-14.33 32-32V64C512 28.65 483.3 0 448 0zM480 176h-48C423.2 176 416 183.2 416 192v16C416 216.8 408.8 224 400 224S384 216.8 384 208V192c0-8.838-7.164-16-16-16H320V152h16c26.51 0 48-21.49 48-48s-21.49-48-48-48H320V32h128c17.67 0 32 14.33 32 32V176zM416 272h-64c-8.836 0-16 7.162-16 16v32c0 8.834-7.164 16-16 16s-16-7.166-16-16V288c0-8.838-7.164-16-16-16H240V96c0-17.67-14.33-32-32-32H64C28.65 64 0 92.65 0 128v160v160c0 35.35 28.65 64 64 64h160h160c35.35 0 64-28.65 64-64V304C448 286.3 433.7 272 416 272zM32 128c0-17.67 14.33-32 32-32h144v176h-32V256c0-26.51-21.49-48-48-48S80 229.5 80 256v16H32V128zM208 480H64c-17.67 0-32-14.33-32-32V304h64c8.836 0 16-7.164 16-16V256c0-8.838 7.164-16 16-16S144 247.2 144 256v32c0 8.836 7.164 16 16 16h48v32H192c-26.51 0-48 21.49-48 48c0 26.51 21.49 48 48 48h16V480zM416 448c0 17.67-14.33 32-32 32H240v-64c0-8.836-7.164-16-16-16H192c-8.836 0-16-7.166-16-16c0-8.836 7.164-16 16-16h32c8.836 0 16-7.166 16-16V304h32V320c0 26.51 21.49 48 48 48c26.51 0 48-21.49 48-48V304H416V448z"
}, null, -1), E1 = [
  C1
];
function $1(n, h) {
  return E(), R("svg", g1, E1);
}
const y1 = /* @__PURE__ */ X(m1, [["render", $1]]), L1 = {}, w1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, F1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M96 0C104.8 0 112 7.164 112 16V112H80V16C80 7.164 87.16 0 96 0zM288 0C296.8 0 304 7.164 304 16V112H272V16C272 7.164 279.2 0 288 0zM368 144C376.8 144 384 151.2 384 160C384 168.8 376.8 176 368 176H352V240C352 322.1 288.9 391.2 208 399.2V496C208 504.8 200.8 512 192 512C183.2 512 176 504.8 176 496V399.2C95.15 391.2 32 322.1 32 240V176H16C7.164 176 0 168.8 0 160C0 151.2 7.164 144 16 144H368zM192 368C262.7 368 320 310.7 320 240V176H64V240C64 310.7 121.3 368 192 368z"
}, null, -1), R1 = [
  F1
];
function A1(n, h) {
  return E(), R("svg", w1, R1);
}
const D1 = /* @__PURE__ */ X(L1, [["render", A1]]), b1 = {}, k1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, I1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M64 64C81.67 64 96 78.33 96 96C96 113.7 81.67 128 64 128C46.33 128 32 113.7 32 96C32 78.33 46.33 64 64 64zM496 80C504.8 80 512 87.16 512 96C512 104.8 504.8 112 496 112H176C167.2 112 160 104.8 160 96C160 87.16 167.2 80 176 80H496zM496 240C504.8 240 512 247.2 512 256C512 264.8 504.8 272 496 272H176C167.2 272 160 264.8 160 256C160 247.2 167.2 240 176 240H496zM496 400C504.8 400 512 407.2 512 416C512 424.8 504.8 432 496 432H176C167.2 432 160 424.8 160 416C160 407.2 167.2 400 176 400H496zM64 288C46.33 288 32 273.7 32 256C32 238.3 46.33 224 64 224C81.67 224 96 238.3 96 256C96 273.7 81.67 288 64 288zM64 384C81.67 384 96 398.3 96 416C96 433.7 81.67 448 64 448C46.33 448 32 433.7 32 416C32 398.3 46.33 384 64 384z"
}, null, -1), S1 = [
  I1
];
function V1(n, h) {
  return E(), R("svg", k1, S1);
}
const x1 = /* @__PURE__ */ X(b1, [["render", V1]]), N1 = {}, B1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, O1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M448 96h-64V48C384 21.53 362.5 0 336 0h-160C149.5 0 128 21.53 128 48V96H64C28.65 96 0 124.7 0 160v256c0 35.35 28.65 64 64 64h384c35.35 0 64-28.65 64-64V160C512 124.7 483.3 96 448 96zM160 48C160 39.17 167.2 32 176 32h160C344.8 32 352 39.17 352 48V96H160V48zM480 416c0 17.64-14.36 32-32 32H64c-17.64 0-32-14.36-32-32V288h144v64c0 8.844 7.156 16 16 16h128c8.844 0 16-7.156 16-16V288H480V416zM208 336V288h96v48H208zM480 256H32V160c0-17.64 14.36-32 32-32h384c17.64 0 32 14.36 32 32V256z"
}, null, -1), T1 = [
  O1
];
function M1(n, h) {
  return E(), R("svg", B1, T1);
}
const q1 = /* @__PURE__ */ X(N1, [["render", M1]]), z1 = {}, H1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, P1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M144 96C152.8 96 160 103.2 160 112C160 120.8 152.8 128 144 128H80C71.16 128 64 120.8 64 112C64 103.2 71.16 96 80 96H144zM64 176C64 167.2 71.16 160 80 160H144C152.8 160 160 167.2 160 176C160 184.8 152.8 192 144 192H80C71.16 192 64 184.8 64 176zM144 224C152.8 224 160 231.2 160 240C160 248.8 152.8 256 144 256H80C71.16 256 64 248.8 64 240C64 231.2 71.16 224 80 224H144zM448 32C483.3 32 512 60.65 512 96V416C512 451.3 483.3 480 448 480H64C28.65 480 0 451.3 0 416V96C0 60.65 28.65 32 64 32H448zM448 64H224V448H448C465.7 448 480 433.7 480 416V96C480 78.33 465.7 64 448 64zM32 96V416C32 433.7 46.33 448 64 448H192V64H64C46.33 64 32 78.33 32 96z"
}, null, -1), j1 = [
  P1
];
function G1(n, h) {
  return E(), R("svg", H1, j1);
}
const U1 = /* @__PURE__ */ X(z1, [["render", G1]]), X1 = {}, W1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, Y1 = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M638.6 368.2l-33.38-211.6c-4.422-25.12-16.52-47.82-34.05-65.77s-40.5-31.18-66.62-37.36c-25.12-5.93-58.4-11.3-91.79-15.19S345.8 32 319.1 32C294.2 32 260.6 34.41 227.2 38.29S160.6 47.56 135.5 53.48C109.3 59.66 86.36 72.88 68.83 90.83S39.2 131.5 34.78 156.6l-33.38 211.6c-5.156 29.26 4.195 57.21 22.67 77.84c18.48 20.63 46.08 33.95 77.42 33.95c24.84 0 48.29-7.512 67.49-20.6c19.2-13.09 34.15-31.75 42.01-54.04l7.5-21.36h203l7.5 21.36c7.859 22.29 22.81 40.95 42.01 54.04c19.2 13.09 42.64 20.6 67.49 20.6c31.34 0 58.95-13.31 77.42-33.95C634.4 425.4 643.8 397.5 638.6 368.2zM592.1 424.7C578.8 439.5 559.3 448 538.5 448c-17.92 0-35.03-5.203-49.46-15.04c-14.1-9.615-24.43-22.84-29.84-38.2l-7.5-21.36c-4.502-12.82-16.61-21.4-30.19-21.4h-203c-13.59 0-25.69 8.58-30.19 21.4l-7.486 21.32c-5.43 15.4-15.76 28.62-29.86 38.24C136.5 442.8 119.4 448 101.5 448c-20.79 0-40.32-8.49-53.58-23.3c-12.92-14.42-18.24-32.51-14.9-51.5l33.28-211C69.52 143.9 78.31 126.9 91.72 113.2c13.69-14.02 31.37-23.9 51.13-28.56C165.1 79.17 197.3 74 230.9 70.08C264.6 66.16 296.2 64 320 64c23.75 0 55.38 2.16 89.05 6.08c33.67 3.922 64.96 9.088 88.1 14.55c.2656 .0625 .5352 .123 .8047 .1777c19.42 4.738 36.8 14.54 50.31 28.39c13.42 13.75 22.21 30.68 25.34 48.41l33.47 212.2C610.3 392.2 605 410.3 592.1 424.7zM432 248c-6.627 0-12.63 2.684-16.97 7.029c-4.344 4.344-7.029 10.35-7.029 16.97s2.685 12.62 7.029 16.96c4.344 4.346 10.34 7.038 16.97 7.038c6.629 0 12.63-2.692 16.97-7.038c4.344-4.344 7.029-10.34 7.029-16.96s-2.686-12.63-7.029-16.97C444.6 250.7 438.6 248 432 248zM496 152c-6.627 0-12.63 2.685-16.97 7.031c-4.344 4.344-7.029 10.35-7.029 16.97s2.685 12.62 7.029 16.96c4.344 4.346 10.34 7.036 16.97 7.036c6.629 0 12.63-2.69 16.97-7.036c4.344-4.344 7.029-10.34 7.029-16.96s-2.686-12.63-7.029-16.97C508.6 154.7 502.6 152 496 152zM256 208H208V160c0-8.844-7.18-16-16.02-16S176 151.2 176 160v48H128C119.2 208 111.1 215.2 111.1 224S119.2 240 128 240h48V288c0 8.844 7.132 16 15.98 16S208 296.8 208 288V240H256c8.844 0 15.98-7.156 15.98-16S264.8 208 256 208z"
}, null, -1), K1 = [
  Y1
];
function Z1(n, h) {
  return E(), R("svg", W1, K1);
}
const Q1 = /* @__PURE__ */ X(X1, [["render", Z1]]), J1 = {}, eu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 576 512"
}, tu = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M440 84.99V212.8L440.1 212.9L544.1 249.1C563.2 256.8 576 274.9 576 295.2V421.4C576 440.8 564.3 458.3 546.4 465.7L442.4 508.8C430.6 513.6 417.4 513.6 405.6 508.8L288 460.1L170.4 508.8C158.6 513.6 145.4 513.6 133.6 508.8L29.65 465.7C11.7 458.3 0 440.8 0 421.4V295.2C0 274.9 12.76 256.8 31.87 249.1L135.9 212.9L136 212.8V84.99C136 64.7 148.8 46.6 167.9 39.78L271.9 2.669C282.3-1.054 293.7-1.054 304.1 2.669L408.1 39.78C427.2 46.6 440 64.7 440 84.99H440zM293.4 32.81C289.9 31.57 286.1 31.57 282.6 32.81L178.6 69.92C176.3 70.75 174.2 72.1 172.6 73.79L287.1 118L403.4 73.79C401.8 72.1 399.7 70.75 397.4 69.92L293.4 32.81zM168.1 212.9L271.1 249.9V146.2L167.1 106.3V212.8L168.1 212.9zM304 249.9L407.9 212.9L408 212.8V106.3L304 146.2V249.9zM159.1 348.6V478.4L271.1 432.1V308.2L159.1 348.6zM41.88 436.2L127.1 471.8V348.1L31.1 308.9V421.4C31.1 427.8 35.9 433.7 41.88 436.2V436.2zM416 348.6L304 308.2V432.1L416 478.4V348.6zM448 471.8L534.1 436.2C540.1 433.7 544 427.8 544 421.4V308.9L448 348.1V471.8zM157.4 243C153.9 241.8 150.1 241.8 146.6 243L44.55 279.4L144.4 320.2L258.5 279.1L157.4 243zM317.5 279.1L431.6 320.2L531.4 279.4L429.4 243C425.9 241.8 422.1 241.8 418.6 243L317.5 279.1z"
}, null, -1), ru = [
  tu
];
function nu(n, h) {
  return E(), R("svg", eu, ru);
}
const su = /* @__PURE__ */ X(J1, [["render", nu]]), uu = {}, iu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 448 512"
}, ou = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M224 512C100.3 512 0 476.2 0 432V80C0 35.82 100.3 0 224 0C347.7 0 448 35.82 448 80V432C448 476.2 347.7 512 224 512zM416 80.45C415.7 79.69 414.4 77.27 409.8 73.31C402.4 67.11 389.9 60.09 371.6 53.57C335.4 40.62 283.2 32 224 32C164.8 32 112.6 40.62 76.37 53.57C58.1 60.09 45.59 67.11 38.25 73.31C33.55 77.27 32.29 79.69 32 80.45V182.1C46.47 192.7 69.9 202.8 100.9 210.4C135.5 218.9 177.1 224 224 224C270 224 312.5 218.9 347.1 210.4C378.1 202.8 401.5 192.7 416 182.1V80.45zM416 219.5C398.8 228.4 377.9 235.8 354.8 241.5C317.3 250.7 272.2 256 224 256C175.8 256 130.7 250.7 93.22 241.5C70.11 235.8 49.18 228.4 32 219.5V310.1C46.47 320.7 69.9 330.8 100.9 338.4C135.5 346.9 177.1 352 224 352C270 352 312.5 346.9 347.1 338.4C378.1 330.8 401.5 320.7 416 310.1V219.5zM38.25 438.7C45.59 444.9 58.1 451.9 76.37 458.4C112.6 471.4 164.8 480 224 480C283.2 480 335.4 471.4 371.6 458.4C389.9 451.9 402.4 444.9 409.8 438.7C414.4 434.7 415.7 432.3 416 431.6V347.5C398.8 356.4 377.9 363.8 354.8 369.5C317.3 378.7 272.2 384 224 384C175.8 384 130.7 378.7 93.22 369.5C70.11 363.8 49.18 356.4 32 347.5V431.6C32.29 432.3 33.55 434.7 38.25 438.7zM416 431.4C416.1 431.3 416.1 431.3 416.1 431.3L416 431.4zM31.96 431.4C31.94 431.3 31.93 431.3 31.92 431.3L31.96 431.4zM31.96 80.56C31.93 80.65 31.92 80.7 31.92 80.7L31.96 80.56zM416.1 80.7C416.1 80.7 416.1 80.65 416 80.56z"
}, null, -1), au = [
  ou
];
function lu(n, h) {
  return E(), R("svg", iu, au);
}
const cu = /* @__PURE__ */ X(uu, [["render", lu]]), hu = {}, fu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, du = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M225.1 378.1c-5.346-9.277-17.67-11.48-25.24-3.91l-91.57 91.55c-15.01 15-38.9 18.79-56.54 6.984c-23.71-15.88-26.04-48.76-6.967-67.97l134.8-134.8c6.242-6.238 6.229-16.36-.0313-22.59C173.3 241.2 163.2 241.2 156.1 247.4l-134.7 134.7c-29.62 29.62-29.62 77.86 0 107.5c16.64 16.78 39.93 24.7 63.89 21.57c17.43-2.273 33.27-11.36 45.7-23.79l90.54-90.52c4.883-4.883 6.32-12.52 2.857-18.49C225.2 378.3 225.1 378.2 225.1 378.1zM259.8 88.66c4.697-7.82 10.35-15.14 16.98-21.8c22.75-22.62 52.62-34.87 83.87-34.87c6.875 0 13.75 .625 20.62 1.75l-69.25 69.48l13.75 82.98l83.12 13.75l69.37-69.23c6.625 38.12-5.373 76.73-33.12 104.5c-4.432 4.434-9.207 8.426-14.23 12c-7.775 5.527-8.24 16.98-1.475 23.7c5.625 5.594 14.65 6.258 21.07 1.602c6.045-4.383 11.82-9.293 17.26-14.81c37.75-37.74 52.62-91.35 39.62-143.2c-2.25-9.492-9.625-16.99-19.12-19.62s-19.62 0-26.62 7l-63.87 63.86L353.5 158.3l-7.373-44.12l63.87-63.98c6.873-6.871 9.623-16.99 6.998-26.49c-2.623-9.496-10.25-16.74-19.75-19.12c-51.62-12.88-105.2 1.875-143.1 39.61c-8.27 8.348-15.16 17.61-20.98 27.39c-3.791 6.367-3.049 14.45 2.191 19.68C242.5 98.44 254.6 97.3 259.8 88.66zM63.1 431.9c0 8.746 7.25 15.1 16 15.1s15.1-7.25 15.1-15.1s-7.248-15.1-15.1-15.1S63.1 423.1 63.1 431.9zM502.6 389.4l-124.4-124.5c-15.62-15.62-36.12-23.37-56.62-23.37c-15.38 0-30.88 4.371-44.12 13.25L191.1 169.3l.002-73.35l-127.1-95.99L0 63.1l96 127.1l73.37-.0289l85.5 85.48c-20.62 31.12-17.12 73.23 10.25 100.7l124.5 124.5C395.7 508.8 403.1 512 412.2 512c8.123 0 16.37-3.125 22.5-9.375l67.1-67.98C515.1 422.1 515.1 401.9 502.6 389.4zM159.1 159.1h-47.1L42.25 66.98L66.1 42.24l92.1 69.73V159.1zM412.1 479.9l-124.4-124.5c-9.123-9.125-14.12-21.12-14.12-33.87c0-12.88 5-24.87 14.12-33.87c9-9.125 21-14.12 33.88-14.12c12.75 0 24.88 4.996 33.87 14.12l124.5 124.3L412.1 479.9z"
}, null, -1), pu = [
  du
];
function _u(n, h) {
  return E(), R("svg", fu, pu);
}
const vu = /* @__PURE__ */ X(hu, [["render", _u]]), mu = {}, gu = {
  class: "k-icon balance",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, Cu = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M256 336h-.02c0-16.18 1.34-8.73-85.05-181.51-17.65-35.29-68.19-35.36-85.87 0C-2.06 328.75.02 320.33.02 336H0c0 44.18 57.31 80 128 80s128-35.82 128-80zM128 176l72 144H56l72-144zm511.98 160c0-16.18 1.34-8.73-85.05-181.51-17.65-35.29-68.19-35.36-85.87 0-87.12 174.26-85.04 165.84-85.04 181.51H384c0 44.18 57.31 80 128 80s128-35.82 128-80h-.02zM440 320l72-144 72 144H440zm88 128H352V153.25c23.51-10.29 41.16-31.48 46.39-57.25H528c8.84 0 16-7.16 16-16V48c0-8.84-7.16-16-16-16H383.64C369.04 12.68 346.09 0 320 0s-49.04 12.68-63.64 32H112c-8.84 0-16 7.16-16 16v32c0 8.84 7.16 16 16 16h129.61c5.23 25.76 22.87 46.96 46.39 57.25V448H112c-8.84 0-16 7.16-16 16v32c0 8.84 7.16 16 16 16h416c8.84 0 16-7.16 16-16v-32c0-8.84-7.16-16-16-16z"
}, null, -1), Eu = [
  Cu
];
function $u(n, h) {
  return E(), R("svg", gu, Eu);
}
const yu = /* @__PURE__ */ X(mu, [["render", $u]]), Lu = {}, wu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Fu = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M480 352h-133.5l-45.25 45.25C289.2 409.3 273.1 416 256 416s-33.16-6.656-45.25-18.75L165.5 352H32c-17.67 0-32 14.33-32 32v96c0 17.67 14.33 32 32 32h448c17.67 0 32-14.33 32-32v-96C512 366.3 497.7 352 480 352zM432 456c-13.2 0-24-10.8-24-24c0-13.2 10.8-24 24-24s24 10.8 24 24C456 445.2 445.2 456 432 456zM233.4 374.6C239.6 380.9 247.8 384 256 384s16.38-3.125 22.62-9.375l128-128c12.49-12.5 12.49-32.75 0-45.25c-12.5-12.5-32.76-12.5-45.25 0L288 274.8V32c0-17.67-14.33-32-32-32C238.3 0 224 14.33 224 32v242.8L150.6 201.4c-12.49-12.5-32.75-12.5-45.25 0c-12.49 12.5-12.49 32.75 0 45.25L233.4 374.6z"
}, null, -1), Ru = [
  Fu
];
function Au(n, h) {
  return E(), R("svg", wu, Ru);
}
const Du = /* @__PURE__ */ X(Lu, [["render", Au]]), bu = {}, ku = {
  class: "k-icon refresh",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Iu = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M496 40v160C496 213.3 485.3 224 472 224h-160C298.8 224 288 213.3 288 200s10.75-24 24-24h100.5C382.8 118.3 322.5 80 256 80C158.1 80 80 158.1 80 256s78.97 176 176 176c41.09 0 81.09-14.47 112.6-40.75c10.16-8.5 25.31-7.156 33.81 3.062c8.5 10.19 7.125 25.31-3.062 33.81c-40.16 33.44-91.17 51.77-143.5 51.77C132.4 479.9 32 379.5 32 256s100.4-223.9 223.9-223.9c79.85 0 152.4 43.46 192.1 109.1V40c0-13.25 10.75-24 24-24S496 26.75 496 40z"
}, null, -1), Su = [
  Iu
];
function Vu(n, h) {
  return E(), R("svg", ku, Su);
}
const xu = /* @__PURE__ */ X(bu, [["render", Vu]]), Nu = {}, Bu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 576 512"
}, Ou = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M287.9 0C297.1 0 305.5 5.25 309.5 13.52L378.1 154.8L531.4 177.5C540.4 178.8 547.8 185.1 550.7 193.7C553.5 202.4 551.2 211.9 544.8 218.2L433.6 328.4L459.9 483.9C461.4 492.9 457.7 502.1 450.2 507.4C442.8 512.7 432.1 513.4 424.9 509.1L287.9 435.9L150.1 509.1C142.9 513.4 133.1 512.7 125.6 507.4C118.2 502.1 114.5 492.9 115.1 483.9L142.2 328.4L31.11 218.2C24.65 211.9 22.36 202.4 25.2 193.7C28.03 185.1 35.5 178.8 44.49 177.5L197.7 154.8L266.3 13.52C270.4 5.249 278.7 0 287.9 0L287.9 0zM287.9 78.95L235.4 187.2C231.9 194.3 225.1 199.3 217.3 200.5L98.98 217.9L184.9 303C190.4 308.5 192.9 316.4 191.6 324.1L171.4 443.7L276.6 387.5C283.7 383.7 292.2 383.7 299.2 387.5L404.4 443.7L384.2 324.1C382.9 316.4 385.5 308.5 391 303L476.9 217.9L358.6 200.5C350.7 199.3 343.9 194.3 340.5 187.2L287.9 78.95z"
}, null, -1), Tu = [
  Ou
];
function Mu(n, h) {
  return E(), R("svg", Bu, Tu);
}
const qu = /* @__PURE__ */ X(Nu, [["render", Mu]]), zu = {}, Hu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 576 512"
}, Pu = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M381.2 150.3L524.9 171.5C536.8 173.2 546.8 181.6 550.6 193.1C554.4 204.7 551.3 217.3 542.7 225.9L438.5 328.1L463.1 474.7C465.1 486.7 460.2 498.9 450.2 506C440.3 513.1 427.2 514 416.5 508.3L288.1 439.8L159.8 508.3C149 514 135.9 513.1 126 506C116.1 498.9 111.1 486.7 113.2 474.7L137.8 328.1L33.58 225.9C24.97 217.3 21.91 204.7 25.69 193.1C29.46 181.6 39.43 173.2 51.42 171.5L195 150.3L259.4 17.97C264.7 6.954 275.9-.0391 288.1-.0391C300.4-.0391 311.6 6.954 316.9 17.97L381.2 150.3z"
}, null, -1), ju = [
  Pu
];
function Gu(n, h) {
  return E(), R("svg", Hu, ju);
}
const Uu = /* @__PURE__ */ X(zu, [["render", Gu]]), Xu = {}, Wu = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 576 512"
}, Yu = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M378.1 154.8L531.4 177.5C540.4 178.8 547.8 185.1 550.7 193.7C553.5 202.4 551.2 211.9 544.8 218.2L433.6 328.4L459.9 483.9C461.4 492.9 457.7 502.1 450.2 507.4C442.8 512.7 432.1 513.4 424.9 509.1L287.9 435.9L150.1 509.1C142.9 513.4 133.1 512.7 125.6 507.4C118.2 502.1 114.5 492.9 115.1 483.9L142.2 328.4L31.11 218.2C24.65 211.9 22.36 202.4 25.2 193.7C28.03 185.1 35.5 178.8 44.49 177.5L197.7 154.8L266.3 13.52C270.4 5.249 278.7 0 287.9 0C297.1 0 305.5 5.25 309.5 13.52L378.1 154.8zM287.1 384.7C291.9 384.7 295.7 385.6 299.2 387.5L404.4 443.7L384.2 324.1C382.9 316.4 385.5 308.5 391 303L476.9 217.9L358.6 200.5C350.7 199.3 343.9 194.3 340.5 187.2L287.1 79.09L287.1 384.7z"
}, null, -1), Ku = [
  Yu
];
function Zu(n, h) {
  return E(), R("svg", Wu, Ku);
}
const Qu = /* @__PURE__ */ X(Xu, [["render", Zu]]), Ju = {}, ei = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, ti = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M256 0C292.8 0 324.8 20.7 340.9 51.1C373.8 40.1 410.1 48.96 437 74.98C463 101 470.1 138.2 460.9 171.1C491.3 187.2 512 219.2 512 256C512 292.8 491.3 324.8 460.9 340.9C471 373.8 463 410.1 437 437C410.1 463 373.8 470.1 340.9 460.9C324.8 491.3 292.8 512 256 512C219.2 512 187.2 491.3 171.1 460.9C138.2 471 101 463 74.98 437C48.96 410.1 41 373.8 51.1 340.9C20.7 324.8 0 292.8 0 256C0 219.2 20.7 187.2 51.1 171.1C40.1 138.2 48.96 101 74.98 74.98C101 48.96 138.2 41 171.1 51.1C187.2 20.7 219.2 0 256 0V0zM352.1 224.1C362.3 215.6 362.3 200.4 352.1 191C343.6 181.7 328.4 181.7 319 191L224 286.1L184.1 247C175.6 237.7 160.4 237.7 151 247C141.7 256.4 141.7 271.6 151 280.1L207 336.1C216.4 346.3 231.6 346.3 240.1 336.1L352.1 224.1z"
}, null, -1), ri = [
  ti
];
function ni(n, h) {
  return E(), R("svg", ei, ri);
}
const si = /* @__PURE__ */ X(Ju, [["render", ni]]), ui = {}, ii = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, oi = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M452 60H488C499 60 508 68.95 508 80C508 91.05 499 100 488 100H452V136C452 147 443 156 432 156C420.1 156 412 147 412 136V100H376C364.1 100 356 91.05 356 80C356 68.95 364.1 60 376 60H412V24C412 12.95 420.1 4 432 4C443 4 452 12.95 452 24V60zM280.8 133.1L490.1 222.8C503.4 228.5 512 241.5 512 256C512 270.5 503.4 283.5 490.1 289.2L280.8 378.9C272.1 382.3 264.5 384 256 384C247.5 384 239 382.3 231.2 378.9L21.9 289.2C8.614 283.5 0 270.5 0 256C0 241.5 8.614 228.5 21.9 222.8L231.2 133.1C239 129.7 247.5 128 256 128C264.5 128 272.1 129.7 280.8 133.1V133.1zM66.26 256L250.1 334.8C251.1 335.6 253.1 336 256 336C258 336 260 335.6 261.9 334.8L445.7 256L261.9 177.2C260 176.4 258 176 256 176C253.1 176 251.1 176.4 250.1 177.2L66.26 256zM59.53 350.6C66.15 359.9 65.24 372.3 57.1 380.5L250.1 462.8C251.1 463.6 253.1 464 256 464C258 464 260 463.6 261.9 462.8L452 381.3C447.4 374.4 446.5 365.2 450.5 357.3C456.5 345.4 470.9 340.6 482.7 346.5L491.8 351.1C504.2 357.3 512 369.9 512 383.7C512 398.4 503.3 411.6 489.9 417.3L280.8 506.9C272.1 510.3 264.5 512 256 512C247.5 512 239 510.3 231.2 506.9L22.81 417.6C8.971 411.7 .001 398.1 .001 383C.001 370.9 5.87 359.5 15.76 352.4L26.05 345C36.84 337.3 51.83 339.8 59.53 350.6L59.53 350.6z"
}, null, -1), ai = [
  oi
];
function li(n, h) {
  return E(), R("svg", ii, ai);
}
const ci = /* @__PURE__ */ X(ui, [["render", li]]), hi = {}, fi = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, di = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: `M144,0c13.3,0,24,10.8,24,24v88h-48V24C120,10.8,130.8,0,144,0z M336,0c13.3,0,24,10.8,24,24v88h-48V24
      C312,10.8,322.7,0,336,0z M408,144c13.3,0,24,10.7,24,24s-10.7,24-24,24h-8v64c0,80.2-59.9,146.6-136,158.2V512h-48v-97.8
      c-77-11.6-136-78-136-158.2v-64h-8c-13.2,0-24-10.7-24-24s10.8-24,24-24H408z M240,368c61.9,0,112-50.1,112-112v-64H128v64
      C128,317.9,178.1,368,240,368z`
}, null, -1), pi = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: `M444,416h36c11,0,20,9,20,20s-9,20-20,20h-36v36c0,11-9,20-20,20c-11.9,0-20-9-20-20v-36h-36c-11.9,0-20-9-20-20
      s8.1-20,20-20h36v-36c0-11,8.1-20,20-20c11,0,20,9,20,20V416z`
}, null, -1), _i = [
  di,
  pi
];
function vi(n, h) {
  return E(), R("svg", fi, _i);
}
const mi = /* @__PURE__ */ X(hi, [["render", vi]]), gi = {}, Ci = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 448 512"
}, Ei = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M160 400C160 408.8 152.8 416 144 416C135.2 416 128 408.8 128 400V192C128 183.2 135.2 176 144 176C152.8 176 160 183.2 160 192V400zM240 400C240 408.8 232.8 416 224 416C215.2 416 208 408.8 208 400V192C208 183.2 215.2 176 224 176C232.8 176 240 183.2 240 192V400zM320 400C320 408.8 312.8 416 304 416C295.2 416 288 408.8 288 400V192C288 183.2 295.2 176 304 176C312.8 176 320 183.2 320 192V400zM317.5 24.94L354.2 80H424C437.3 80 448 90.75 448 104C448 117.3 437.3 128 424 128H416V432C416 476.2 380.2 512 336 512H112C67.82 512 32 476.2 32 432V128H24C10.75 128 0 117.3 0 104C0 90.75 10.75 80 24 80H93.82L130.5 24.94C140.9 9.357 158.4 0 177.1 0H270.9C289.6 0 307.1 9.358 317.5 24.94H317.5zM151.5 80H296.5L277.5 51.56C276 49.34 273.5 48 270.9 48H177.1C174.5 48 171.1 49.34 170.5 51.56L151.5 80zM80 432C80 449.7 94.33 464 112 464H336C353.7 464 368 449.7 368 432V128H80V432z"
}, null, -1), $i = [
  Ei
];
function yi(n, h) {
  return E(), R("svg", Ci, $i);
}
const Li = /* @__PURE__ */ X(gi, [["render", yi]]), wi = {}, Fi = {
  class: "k-icon check",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 448 512"
}, Ri = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M438.6 105.4C451.1 117.9 451.1 138.1 438.6 150.6L182.6 406.6C170.1 419.1 149.9 419.1 137.4 406.6L9.372 278.6C-3.124 266.1-3.124 245.9 9.372 233.4C21.87 220.9 42.13 220.9 54.63 233.4L159.1 338.7L393.4 105.4C405.9 92.88 426.1 92.88 438.6 105.4H438.6z"
}, null, -1), Ai = [
  Ri
];
function Di(n, h) {
  return E(), R("svg", Fi, Ai);
}
const bi = /* @__PURE__ */ X(wi, [["render", Di]]), ki = {}, Ii = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, Si = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M24.52 38.13C39.66 29.64 58.21 29.99 73.03 39.04L361 215C375.3 223.8 384 239.3 384 256C384 272.7 375.3 288.2 361 296.1L73.03 472.1C58.21 482 39.66 482.4 24.52 473.9C9.377 465.4 0 449.4 0 432V80C0 62.64 9.377 46.63 24.52 38.13V38.13zM48 432L336 256L48 80V432z"
}, null, -1), Vi = [
  Si
];
function xi(n, h) {
  return E(), R("svg", Ii, Vi);
}
const Ni = /* @__PURE__ */ X(ki, [["render", xi]]), Bi = {}, Oi = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, Ti = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M320 64H64c-35.35 0-64 28.65-64 64v255.1c0 35.35 28.65 64 64 64H320c35.35 0 64-28.65 64-64V128C384 92.65 355.3 64 320 64zM336 384c0 8.822-7.178 16-16 16H64c-8.822 0-16-7.178-16-16V128c0-8.822 7.178-16 16-16h256c8.822 0 16 7.178 16 16V384z"
}, null, -1), Mi = [
  Ti
];
function qi(n, h) {
  return E(), R("svg", Oi, Mi);
}
const zi = /* @__PURE__ */ X(Bi, [["render", qi]]), Hi = {}, Pi = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 448 512"
}, ji = /* @__PURE__ */ $("path", {
  fill: "currentColor",
  d: "M224 256c-35.2 0-64 28.8-64 64c0 35.2 28.8 64 64 64c35.2 0 64-28.8 64-64C288 284.8 259.2 256 224 256zM433.1 129.1l-83.9-83.9C341.1 37.06 328.8 32 316.1 32H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h320c35.35 0 64-28.65 64-64V163.9C448 151.2 442.9 138.9 433.1 129.1zM128 80h144V160H128V80zM400 416c0 8.836-7.164 16-16 16H64c-8.836 0-16-7.164-16-16V96c0-8.838 7.164-16 16-16h16v104c0 13.25 10.75 24 24 24h192C309.3 208 320 197.3 320 184V83.88l78.25 78.25C399.4 163.2 400 164.8 400 166.3V416z"
}, null, -1), Gi = [
  ji
];
function Ui(n, h) {
  return E(), R("svg", Pi, Gi);
}
const Xi = /* @__PURE__ */ X(Hi, [["render", Ui]]);
K.register("activity:deps", v1);
K.register("activity:market", y1);
K.register("activity:plugin", D1);
K.register("category:all", x1);
K.register("category:business", q1);
K.register("category:console", U1);
K.register("category:game", Q1);
K.register("category:other", su);
K.register("category:storage", cu);
K.register("category:tool", vu);
K.register("balance", yu);
K.register("download", Du);
K.register("refresh", xu);
K.register("star-empty", qu);
K.register("star-full", Uu);
K.register("star-half", Qu);
K.register("verified", si);
K.register("add-plugin", mi);
K.register("add-group", ci);
K.register("trash-can", Li);
K.register("check", bi);
K.register("play", Ni);
K.register("stop", zi);
K.register("save", Xi);
nn("market/patch", (n) => {
  var h;
  k.market = {
    ...n,
    data: {
      ...(h = k.market) == null ? void 0 : h.data,
      ...n.data
    }
  };
});
const Qi = (n) => {
  n.slot({
    type: "welcome-choice",
    component: ne(() => () => Ve("div", {
      class: "choice",
      onClick: () => fe.push("/market")
    }, [
      Ve("h2", "\u6D4F\u89C8\u63D2\u4EF6"),
      Ve("p", "\u6D4F\u89C8\u63D2\u4EF6\u5E02\u573A\u4E2D\u7684\u63D2\u4EF6\uFF0C\u5E76\u6839\u636E\u81EA\u5DF1\u7684\u9700\u8981\u5B89\u88C5\u548C\u914D\u7F6E\u3002")
    ]))
  }), n.slot({
    type: "global",
    component: Wn
  }), n.slot({
    type: "global",
    component: l1
  }), n.slot({
    type: "status-right",
    component: c1,
    order: 10
  }), n.page({
    path: "/plugins/:name*",
    name: "\u63D2\u4EF6\u914D\u7F6E",
    icon: "activity:plugin",
    order: 630,
    authority: 4,
    fields: ["config", "packages"],
    component: Rs
  }), n.page({
    path: "/market",
    name: "\u63D2\u4EF6\u5E02\u573A",
    icon: "activity:market",
    order: 620,
    authority: 4,
    component: n1
  }), n.page({
    path: "/dependencies",
    name: "\u4F9D\u8D56\u7BA1\u7406",
    icon: "activity:deps",
    order: 610,
    authority: 4,
    fields: ["dependencies"],
    component: ns
  });
};
export {
  Qi as default
};
