// koishi-plugin-nonebot/lib/internal/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parameter = exports.kwarg = exports.unwrap = exports.isPyProxy = exports.extractText = void 0;
const koishi_1 = require("koishi");
function extractText(elements) {
    return koishi_1.h.transform(elements, {
        default: false,
        text: true,
    }).join('');
}
exports.extractText = extractText;
function isPyProxy(obj) {
    return !!(obj === null || obj === void 0 ? void 0 : obj.toJs);
}
exports.isPyProxy = isPyProxy;
function unwrap(obj) {
    if (!isPyProxy(obj))
        return obj;
    if (obj.type === 'Message') {
        return obj.toJs().map(item => item.internal);
    }
    else if (obj.type === 'MessageSegment') {
        return obj.internal;
    }
    return obj.toJs();
}
exports.unwrap = unwrap;
function kwarg(name, args, index = 0) {
    var _a;
    if (args.length > index + 1)
        return unwrap(args[index]);
    if (args.length <= index)
        return unwrap((_a = args[args.length - 1]) === null || _a === void 0 ? void 0 : _a[name]);
    if (typeof args[index] !== 'object')
        return args[index];
    if (isPyProxy(args[index]))
        return unwrap(args[index]);
    return unwrap(args[index][name]);
}
exports.kwarg = kwarg;
var Parameter;
(function (Parameter) {
    let Kind;
    (function (Kind) {
        Kind[Kind["POSITIONAL_ONLY"] = 0] = "POSITIONAL_ONLY";
        Kind[Kind["POSITIONAL_OR_KEYWORD"] = 1] = "POSITIONAL_OR_KEYWORD";
        Kind[Kind["VAR_POSITIONAL"] = 2] = "VAR_POSITIONAL";
        Kind[Kind["KEYWORD_ONLY"] = 3] = "KEYWORD_ONLY";
        Kind[Kind["VAR_KEYWORD"] = 4] = "VAR_KEYWORD";
    })(Kind = Parameter.Kind || (Parameter.Kind = {}));
})(Parameter = exports.Parameter || (exports.Parameter = {}));
