// aragami/dist/src/base-driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseDriver = void 0;
class BaseDriver {
    usingKey(baseKey, key) {
        return `${baseKey}:${key}`;
    }
    async get(baseKey, key) {
        return;
    }
    async has(baseKey, key) {
        return !!(await this.get(baseKey, key));
    }
    async set(baseKey, key, value, ttl) { }
    async del(baseKey, key) {
        return false;
    }
    async keys(baseKey, prefix) {
        return [];
    }
    async values(baseKey, prefix) {
        const keys = await this.keys(baseKey, prefix);
        return Promise.all(keys.map((key) => this.get(baseKey, key)));
    }
    async entries(baseKey, prefix) {
        const keys = await this.keys(baseKey, prefix);
        return Promise.all(keys.map(async (key) => [key, await this.get(baseKey, key)]));
    }
    async clear(baseKey, prefix) {
        const keys = await this.keys(baseKey, prefix);
        await Promise.all(keys.map((key) => this.del(baseKey, key)));
    }
    lock(keys, cb) {
        return cb();
    }
    async destroy() { }
}
exports.BaseDriver = BaseDriver;
