// @koishijs/plugin-analytics/dist/index.js
import { Card as c, store as v, icons as i, defineExtension as x } from "../client.js";
import { defineAsyncComponent as $, h as g, resolveComponent as A, openBlock as a, createElementBlock as l, createElementVNode as C } from "../vue.js";
const w = $(() => import("./echarts.f084e4a0.mjs"));
function m({ title: e, fields: o, options: t }) {
  return c.create(() => {
    const r = t(v);
    if (!!r)
      return g(A("k-card"), { class: "frameless", title: e }, () => g(w, { option: r, autoresize: !0 }));
  }, o);
}
var u;
((e) => {
  e.item = (o) => ({
    trigger: "item",
    formatter: o
  }), e.axis = (o) => ({
    trigger: "axis",
    axisPointer: {
      type: "cross"
    },
    formatter: o
  });
})(u || (u = {}));
const b = (e) => {
  e.slot({
    type: "chart",
    component: m({
      title: "\u6307\u4EE4\u8C03\u7528\u9891\u7387",
      fields: ["stats"],
      options({ stats: o }) {
        const t = Object.entries(o.commands).sort((r, n) => n[1] - r[1]).map(([r, n]) => ({ name: r, value: n }));
        if (!!t.length)
          return {
            tooltip: u.item(({ data: r }) => {
              const n = [r.name];
              return n.push(`\u65E5\u5747\u8C03\u7528\uFF1A${r.value}`), n.join("<br>");
            }),
            series: [{
              type: "pie",
              data: t,
              radius: ["35%", "65%"],
              minShowLabelAngle: 3
            }]
          };
      }
    })
  });
}, z = (e) => {
  e.slot({
    type: "chart",
    component: m({
      title: "\u5404\u7FA4\u53D1\u8A00\u6570\u91CF",
      fields: ["stats"],
      options({ stats: o }) {
        if (!!o.guilds.length)
          return {
            tooltip: u.item(({ data: t }) => {
              const r = [t.name];
              return r.push(`\u5E73\u53F0\uFF1A${t.platform}`), t.memberCount && r.push(`\u4EBA\u6570\uFF1A${t.memberCount}`), t.assignee && r.push(`\u63A5\u5165\uFF1A${t.assignee}`), r.push(`\u65E5\u5747\u53D1\u8A00\uFF1A${+t.value.toFixed(1)}`), r.push(`\u6628\u65E5\u53D1\u8A00\uFF1A${+t.last.toFixed(1)}`), r.join("<br>");
            }),
            series: [{
              type: "pie",
              data: o.guilds.sort((t, r) => r.value - t.value),
              radius: ["35%", "65%"],
              minShowLabelAngle: 3
            }]
          };
      }
    })
  });
}, S = "\u65E5\u4E00\u4E8C\u4E09\u56DB\u4E94\u516D", E = (e) => {
  e.slot({
    type: "chart",
    component: m({
      title: "\u5386\u53F2\u53D1\u8A00\u6570\u91CF",
      fields: ["stats"],
      options({ stats: o }) {
        if (!!Object.keys(o.history).length)
          return {
            tooltip: u.axis(([{ name: t, value: r }]) => {
              const n = new Date(t).getDay();
              return `${t} \u661F\u671F${S[n]}<br>\u53D1\u8A00\u6570\u91CF\uFF1A${r}`;
            }),
            xAxis: {
              type: "category",
              data: Object.keys(o.history).reverse()
            },
            yAxis: {
              type: "value"
            },
            series: {
              type: "line",
              smooth: !0,
              data: Object.values(o.history).reverse()
            }
          };
      }
    })
  });
}, f = (e) => `${(e - 0.5).toFixed()}:00-${(e + 0.5).toFixed()}:00`, D = (e) => {
  e.slot({
    type: "chart",
    component: m({
      title: "\u6BCF\u5C0F\u65F6\u53D1\u8A00\u6570\u91CF",
      fields: ["stats"],
      options({ stats: o }) {
        return {
          tooltip: u.axis((t) => {
            const [{ data: [r], dataIndex: n }] = t, s = o.hours[n], h = [
              `${f(r)}`,
              `\u6D88\u606F\u603B\u91CF\uFF1A${+(s.total || 0).toFixed(1)}`
            ];
            return t.reverse().forEach(({ seriesName: F, marker: y }, d) => {
              const _ = d === 0 ? s.command : d === 1 ? s.dialogue : Math.max(0, s.total - s.command - s.dialogue);
              !_ || h.push(`${y}${F}\uFF1A${+_.toFixed(1)}`);
            }), h.join("<br>");
          }),
          xAxis: {
            type: "value",
            min: 0,
            max: 24,
            minInterval: 1,
            maxInterval: 4,
            axisLabel: {
              formatter: (t) => t + ":00"
            },
            axisPointer: {
              label: {
                formatter: ({ value: t }) => f(t)
              }
            }
          },
          yAxis: {
            type: "value"
          },
          series: [{
            name: "\u5176\u4ED6",
            data: o.hours.map((t, r) => [r + 0.5, t.total || 0]),
            type: "bar",
            stack: "1",
            itemStyle: {
              color: "rgb(255,219,92)"
            }
          }, {
            name: "\u6559\u5B66",
            data: o.hours.map((t, r) => [r + 0.5, (t.command || 0) + (t.dialogue || 0)]),
            type: "bar",
            stack: "1",
            itemStyle: {
              color: "rgb(103,224,227)"
            }
          }, {
            name: "\u6307\u4EE4",
            data: o.hours.map((t, r) => [r + 0.5, t.command || 0]),
            type: "bar",
            stack: "1",
            itemStyle: {
              color: "rgb(55,162,218)"
            }
          }]
        };
      }
    })
  });
}, M = (e) => {
  e.plugin(E), e.plugin(D), e.plugin(z), e.plugin(b);
}, p = (e, o) => {
  const t = e.__vccOpts || e;
  for (const [r, n] of o)
    t[r] = n;
  return t;
}, V = {}, H = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, k = /* @__PURE__ */ C("path", {
  fill: "currentColor",
  d: "M255.1 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 0 232.4 0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84.02L255.1 96zM255.1 141.3L221.4 106.6C196.1 81.31 160 69.77 124.7 75.66C71.21 84.58 31.1 130.9 31.1 185.1V190.9C31.1 223.6 45.55 254.7 69.42 277L250.1 445.7C251.7 447.2 253.8 448 255.1 448C258.2 448 260.3 447.2 261.9 445.7L442.6 277C466.4 254.7 480 223.6 480 190.9V185.1C480 130.9 440.8 84.58 387.3 75.66C351.1 69.77 315.9 81.31 290.6 106.6L255.1 141.3z"
}, null, -1), B = [
  k
];
function L(e, o) {
  return a(), l("svg", H, B);
}
const j = /* @__PURE__ */ p(V, [["render", L]]), O = {}, I = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, G = /* @__PURE__ */ C("path", {
  fill: "currentColor",
  d: "M256 0C397.4 0 512 114.6 512 256C512 397.4 397.4 512 256 512C172.2 512 97.87 471.8 51.19 409.6C45.88 402.5 47.31 392.5 54.37 387.2C61.44 381.9 71.47 383.3 76.78 390.4C117.7 444.8 182.7 480 256 480C379.7 480 480 379.7 480 256C480 132.3 379.7 32 256 32C166.7 32 89.51 84.3 53.55 160H144C152.8 160 160 167.2 160 176C160 184.8 152.8 192 144 192H16C7.164 192 0 184.8 0 176V48C0 39.16 7.164 32 16 32C24.84 32 32 39.16 32 48V131.1C75.66 53.29 159.6 0 256 0zM256 128C264.8 128 272 135.2 272 144V249.4L347.3 324.7C353.6 330.9 353.6 341.1 347.3 347.3C341.1 353.6 330.9 353.6 324.7 347.3L244.7 267.3C241.7 264.3 239.1 260.2 239.1 256V144C239.1 135.2 247.2 128 255.1 128H256z"
}, null, -1), P = [
  G
];
function U(e, o) {
  return a(), l("svg", I, P);
}
const N = /* @__PURE__ */ p(O, [["render", U]]), q = {}, J = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, K = /* @__PURE__ */ C("path", {
  fill: "currentColor",
  d: "M448 32h-384c-35.35 0-64 28.65-64 64v320c0 35.35 28.65 64 64 64h384c35.35 0 64-28.65 64-64V96C512 60.65 483.3 32 448 32zM480 416c0 17.64-14.36 32-32 32H64c-17.64 0-32-14.36-32-32v-96c0-17.64 14.36-32 32-32h384C465.6 288 480 302.4 480 320V416zM480 264.9C470.6 259.4 459.7 256 448 256H64C52.28 256 41.45 259.4 32 264.9V96c0-17.64 14.36-32 32-32h384c17.64 0 32 14.36 32 32V264.9zM320 344c-13.25 0-24 10.74-24 24c0 13.25 10.75 24 24 24c13.26 0 24-10.75 24-24C344 354.7 333.3 344 320 344zM416 344c-13.25 0-24 10.74-24 24c0 13.25 10.75 24 24 24c13.26 0 24-10.75 24-24C440 354.7 429.3 344 416 344z"
}, null, -1), Q = [
  K
];
function R(e, o) {
  return a(), l("svg", J, Q);
}
const T = /* @__PURE__ */ p(q, [["render", R]]), W = {}, X = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, Y = /* @__PURE__ */ C("path", {
  fill: "currentColor",
  d: "M319.9 320c57.41 0 103.1-46.56 103.1-104c0-57.44-46.54-104-103.1-104c-57.41 0-103.1 46.56-103.1 104C215.9 273.4 262.5 320 319.9 320zM319.9 144c39.68 0 71.96 32.3 71.96 72S359.5 288 319.9 288S247.9 255.7 247.9 216S280.2 144 319.9 144zM369.9 352H270.1C191.6 352 128 411.7 128 485.3C128 500.1 140.7 512 156.4 512h327.2C499.3 512 512 500.1 512 485.3C512 411.7 448.4 352 369.9 352zM160.2 480c3.021-53.41 51.19-96 109.1-96H369.9c58.78 0 106.9 42.59 109.1 96H160.2zM512 160c44.18 0 80-35.82 80-80S556.2 0 512 0c-44.18 0-80 35.82-80 80S467.8 160 512 160zM512 32c26.47 0 48 21.53 48 48S538.5 128 512 128s-48-21.53-48-48S485.5 32 512 32zM128 160c44.18 0 80-35.82 80-80S172.2 0 128 0C83.82 0 48 35.82 48 80S83.82 160 128 160zM128 32c26.47 0 48 21.53 48 48S154.5 128 128 128S80 106.5 80 80S101.5 32 128 32zM561.1 192H496c-11.16 0-22.08 2.5-32.47 7.438c-7.984 3.797-11.39 13.34-7.594 21.31s13.38 11.39 21.31 7.594C483.3 225.5 489.6 224 496 224h65.08C586.1 224 608 246.7 608 274.7V288c0 8.844 7.156 16 16 16S640 296.8 640 288V274.7C640 229.1 604.6 192 561.1 192zM162.8 228.3c7.938 3.797 17.53 .375 21.31-7.594c3.797-7.969 .3906-17.52-7.594-21.31C166.1 194.5 155.2 192 144 192H78.92C35.41 192 0 229.1 0 274.7V288c0 8.844 7.156 16 16 16S32 296.8 32 288V274.7C32 246.7 53.05 224 78.92 224H144C150.4 224 156.7 225.5 162.8 228.3z"
}, null, -1), Z = [
  Y
];
function t1(e, o) {
  return a(), l("svg", X, Z);
}
const e1 = /* @__PURE__ */ p(W, [["render", t1]]);
i.register("heart", j);
i.register("history", N);
i.register("storage", T);
i.register("users", e1);
const n1 = x((e) => {
  e.plugin(M), e.slot({
    type: "numeric",
    component: c.numeric({
      title: "\u8FD1\u671F\u6D88\u606F\u9891\u7387",
      icon: "history",
      fields: ["stats"],
      content({ stats: o }) {
        return Object.values(o.botSend).reduce((t, r) => t + r, 0).toFixed(1) + " / d";
      }
    })
  }), e.slot({
    type: "numeric",
    component: c.numeric({
      title: "\u8D44\u6E90\u670D\u52A1\u5668",
      icon: "storage",
      type: "size",
      fields: ["meta"],
      content: ({ meta: o }) => o.assetSize
    })
  }), e.slot({
    type: "numeric",
    component: c.numeric({
      title: "\u6D3B\u8DC3\u7528\u6237\u6570\u91CF",
      icon: "heart",
      fields: ["meta"],
      content: ({ meta: o }) => o.activeUsers
    })
  }), e.slot({
    type: "numeric",
    component: c.numeric({
      title: "\u6D3B\u8DC3\u7FA4\u7EC4\u6570\u91CF",
      icon: "users",
      fields: ["meta"],
      content: ({ meta: o }) => o.activeGuilds
    })
  });
});
export {
  n1 as default
};
