// @koishijs/plugin-logger/dist/index.js
import { defineComponent as B, ref as I, onMounted as L, watch as j, nextTick as R, resolveComponent as C, openBlock as v, createBlock as M, withCtx as E, createVNode as U, createElementVNode as S, createElementBlock as w, Fragment as q, renderList as A, unref as H, normalizeClass as z } from "../vue.js";
import { store as k, icons as N } from "../client.js";
function F(h, u) {
  for (var a = 0; a < u.length; a++) {
    const n = u[a];
    if (typeof n != "string" && !Array.isArray(n)) {
      for (const g in n)
        if (g !== "default" && !(g in h)) {
          const o = Object.getOwnPropertyDescriptor(n, g);
          o && Object.defineProperty(h, g, o.get ? o : {
            enumerable: !0,
            get: () => n[g]
          });
        }
    }
  }
  return Object.freeze(Object.defineProperty(h, Symbol.toStringTag, { value: "Module" }));
}
var G = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, y = {}, O;
function V() {
  return O || (O = 1, function(h) {
    (function(u, a) {
      if (typeof h.nodeName != "string")
        a(h);
      else {
        var n = {};
        a(n), u.AnsiUp = n.default;
      }
    })(G, function(u) {
      var a = this && this.__makeTemplateObject || function(s, e) {
        return Object.defineProperty ? Object.defineProperty(s, "raw", { value: e }) : s.raw = e, s;
      }, n;
      (function(s) {
        s[s.EOS = 0] = "EOS", s[s.Text = 1] = "Text", s[s.Incomplete = 2] = "Incomplete", s[s.ESC = 3] = "ESC", s[s.Unknown = 4] = "Unknown", s[s.SGR = 5] = "SGR", s[s.OSCURL = 6] = "OSCURL";
      })(n || (n = {}));
      var g = function() {
        function s() {
          this.VERSION = "5.1.0", this.setup_palettes(), this._use_classes = !1, this.bold = !1, this.italic = !1, this.underline = !1, this.fg = this.bg = null, this._buffer = "", this._url_whitelist = { http: 1, https: 1 };
        }
        return Object.defineProperty(s.prototype, "use_classes", {
          get: function() {
            return this._use_classes;
          },
          set: function(e) {
            this._use_classes = e;
          },
          enumerable: !1,
          configurable: !0
        }), Object.defineProperty(s.prototype, "url_whitelist", {
          get: function() {
            return this._url_whitelist;
          },
          set: function(e) {
            this._url_whitelist = e;
          },
          enumerable: !1,
          configurable: !0
        }), s.prototype.setup_palettes = function() {
          var e = this;
          this.ansi_colors = [
            [
              { rgb: [0, 0, 0], class_name: "ansi-black" },
              { rgb: [187, 0, 0], class_name: "ansi-red" },
              { rgb: [0, 187, 0], class_name: "ansi-green" },
              { rgb: [187, 187, 0], class_name: "ansi-yellow" },
              { rgb: [0, 0, 187], class_name: "ansi-blue" },
              { rgb: [187, 0, 187], class_name: "ansi-magenta" },
              { rgb: [0, 187, 187], class_name: "ansi-cyan" },
              { rgb: [255, 255, 255], class_name: "ansi-white" }
            ],
            [
              { rgb: [85, 85, 85], class_name: "ansi-bright-black" },
              { rgb: [255, 85, 85], class_name: "ansi-bright-red" },
              { rgb: [0, 255, 0], class_name: "ansi-bright-green" },
              { rgb: [255, 255, 85], class_name: "ansi-bright-yellow" },
              { rgb: [85, 85, 255], class_name: "ansi-bright-blue" },
              { rgb: [255, 85, 255], class_name: "ansi-bright-magenta" },
              { rgb: [85, 255, 255], class_name: "ansi-bright-cyan" },
              { rgb: [255, 255, 255], class_name: "ansi-bright-white" }
            ]
          ], this.palette_256 = [], this.ansi_colors.forEach(function(b) {
            b.forEach(function(x) {
              e.palette_256.push(x);
            });
          });
          for (var t = [0, 95, 135, 175, 215, 255], r = 0; r < 6; ++r)
            for (var i = 0; i < 6; ++i)
              for (var l = 0; l < 6; ++l) {
                var f = { rgb: [t[r], t[i], t[l]], class_name: "truecolor" };
                this.palette_256.push(f);
              }
          for (var c = 8, _ = 0; _ < 24; ++_, c += 10) {
            var p = { rgb: [c, c, c], class_name: "truecolor" };
            this.palette_256.push(p);
          }
        }, s.prototype.escape_txt_for_html = function(e) {
          return e.replace(/[&<>"']/gm, function(t) {
            if (t === "&")
              return "&amp;";
            if (t === "<")
              return "&lt;";
            if (t === ">")
              return "&gt;";
            if (t === '"')
              return "&quot;";
            if (t === "'")
              return "&#x27;";
          });
        }, s.prototype.append_buffer = function(e) {
          var t = this._buffer + e;
          this._buffer = t;
        }, s.prototype.get_next_packet = function() {
          var e = {
            kind: n.EOS,
            text: "",
            url: ""
          }, t = this._buffer.length;
          if (t == 0)
            return e;
          var r = this._buffer.indexOf("\x1B");
          if (r == -1)
            return e.kind = n.Text, e.text = this._buffer, this._buffer = "", e;
          if (r > 0)
            return e.kind = n.Text, e.text = this._buffer.slice(0, r), this._buffer = this._buffer.slice(r), e;
          if (r == 0) {
            if (t == 1)
              return e.kind = n.Incomplete, e;
            var i = this._buffer.charAt(1);
            if (i != "[" && i != "]")
              return e.kind = n.ESC, e.text = this._buffer.slice(0, 1), this._buffer = this._buffer.slice(1), e;
            if (i == "[") {
              this._csi_regex || (this._csi_regex = o(a([`
                        ^                           # beginning of line
                                                    #
                                                    # First attempt
                        (?:                         # legal sequence
                          \x1B[                      # CSI
                          ([<-?]?)              # private-mode char
                          ([d;]*)                    # any digits or semicolons
                          ([ -/]?               # an intermediate modifier
                          [@-~])                # the command
                        )
                        |                           # alternate (second attempt)
                        (?:                         # illegal sequence
                          \x1B[                      # CSI
                          [ -~]*                # anything legal
                          ([\0-:])              # anything illegal
                        )
                    `], [`
                        ^                           # beginning of line
                                                    #
                                                    # First attempt
                        (?:                         # legal sequence
                          \\x1b\\[                      # CSI
                          ([\\x3c-\\x3f]?)              # private-mode char
                          ([\\d;]*)                    # any digits or semicolons
                          ([\\x20-\\x2f]?               # an intermediate modifier
                          [\\x40-\\x7e])                # the command
                        )
                        |                           # alternate (second attempt)
                        (?:                         # illegal sequence
                          \\x1b\\[                      # CSI
                          [\\x20-\\x7e]*                # anything legal
                          ([\\x00-\\x1f:])              # anything illegal
                        )
                    `])));
              var l = this._buffer.match(this._csi_regex);
              if (l === null)
                return e.kind = n.Incomplete, e;
              if (l[4])
                return e.kind = n.ESC, e.text = this._buffer.slice(0, 1), this._buffer = this._buffer.slice(1), e;
              l[1] != "" || l[3] != "m" ? e.kind = n.Unknown : e.kind = n.SGR, e.text = l[2];
              var f = l[0].length;
              return this._buffer = this._buffer.slice(f), e;
            }
            if (i == "]") {
              if (t < 4)
                return e.kind = n.Incomplete, e;
              if (this._buffer.charAt(2) != "8" || this._buffer.charAt(3) != ";")
                return e.kind = n.ESC, e.text = this._buffer.slice(0, 1), this._buffer = this._buffer.slice(1), e;
              this._osc_st || (this._osc_st = d(a([`
                        (?:                         # legal sequence
                          (\x1B\\)                    # ESC                           |                           # alternate
                          (\x07)                      # BEL (what xterm did)
                        )
                        |                           # alternate (second attempt)
                        (                           # illegal sequence
                          [\0-]                 # anything illegal
                          |                           # alternate
                          [\b-]                 # anything illegal
                          |                           # alternate
                          [-]                 # anything illegal
                        )
                    `], [`
                        (?:                         # legal sequence
                          (\\x1b\\\\)                    # ESC \\
                          |                           # alternate
                          (\\x07)                      # BEL (what xterm did)
                        )
                        |                           # alternate (second attempt)
                        (                           # illegal sequence
                          [\\x00-\\x06]                 # anything illegal
                          |                           # alternate
                          [\\x08-\\x1a]                 # anything illegal
                          |                           # alternate
                          [\\x1c-\\x1f]                 # anything illegal
                        )
                    `]))), this._osc_st.lastIndex = 0;
              {
                var c = this._osc_st.exec(this._buffer);
                if (c === null)
                  return e.kind = n.Incomplete, e;
                if (c[3])
                  return e.kind = n.ESC, e.text = this._buffer.slice(0, 1), this._buffer = this._buffer.slice(1), e;
              }
              {
                var _ = this._osc_st.exec(this._buffer);
                if (_ === null)
                  return e.kind = n.Incomplete, e;
                if (_[3])
                  return e.kind = n.ESC, e.text = this._buffer.slice(0, 1), this._buffer = this._buffer.slice(1), e;
              }
              this._osc_regex || (this._osc_regex = o(a([`
                        ^                           # beginning of line
                                                    #
                        \x1B]8;                    # OSC Hyperlink
                        [ -:<-~]*       # params (excluding ;)
                        ;                           # end of params
                        ([!-~]{0,512})        # URL capture
                        (?:                         # ST
                          (?:\x1B\\)                  # ESC                           |                           # alternate
                          (?:\x07)                    # BEL (what xterm did)
                        )
                        ([ -~]+)              # TEXT capture
                        \x1B]8;;                   # OSC Hyperlink End
                        (?:                         # ST
                          (?:\x1B\\)                  # ESC                           |                           # alternate
                          (?:\x07)                    # BEL (what xterm did)
                        )
                    `], [`
                        ^                           # beginning of line
                                                    #
                        \\x1b\\]8;                    # OSC Hyperlink
                        [\\x20-\\x3a\\x3c-\\x7e]*       # params (excluding ;)
                        ;                           # end of params
                        ([\\x21-\\x7e]{0,512})        # URL capture
                        (?:                         # ST
                          (?:\\x1b\\\\)                  # ESC \\
                          |                           # alternate
                          (?:\\x07)                    # BEL (what xterm did)
                        )
                        ([\\x20-\\x7e]+)              # TEXT capture
                        \\x1b\\]8;;                   # OSC Hyperlink End
                        (?:                         # ST
                          (?:\\x1b\\\\)                  # ESC \\
                          |                           # alternate
                          (?:\\x07)                    # BEL (what xterm did)
                        )
                    `])));
              var l = this._buffer.match(this._osc_regex);
              if (l === null)
                return e.kind = n.ESC, e.text = this._buffer.slice(0, 1), this._buffer = this._buffer.slice(1), e;
              e.kind = n.OSCURL, e.url = l[1], e.text = l[2];
              var f = l[0].length;
              return this._buffer = this._buffer.slice(f), e;
            }
          }
        }, s.prototype.ansi_to_html = function(e) {
          this.append_buffer(e);
          for (var t = []; ; ) {
            var r = this.get_next_packet();
            if (r.kind == n.EOS || r.kind == n.Incomplete)
              break;
            r.kind == n.ESC || r.kind == n.Unknown || (r.kind == n.Text ? t.push(this.transform_to_html(this.with_state(r))) : r.kind == n.SGR ? this.process_ansi(r) : r.kind == n.OSCURL && t.push(this.process_hyperlink(r)));
          }
          return t.join("");
        }, s.prototype.with_state = function(e) {
          return { bold: this.bold, italic: this.italic, underline: this.underline, fg: this.fg, bg: this.bg, text: e.text };
        }, s.prototype.process_ansi = function(e) {
          for (var t = e.text.split(";"); t.length > 0; ) {
            var r = t.shift(), i = parseInt(r, 10);
            if (isNaN(i) || i === 0)
              this.fg = this.bg = null, this.bold = !1, this.italic = !1, this.underline = !1;
            else if (i === 1)
              this.bold = !0;
            else if (i === 3)
              this.italic = !0;
            else if (i === 4)
              this.underline = !0;
            else if (i === 22)
              this.bold = !1;
            else if (i === 23)
              this.italic = !1;
            else if (i === 24)
              this.underline = !1;
            else if (i === 39)
              this.fg = null;
            else if (i === 49)
              this.bg = null;
            else if (i >= 30 && i < 38)
              this.fg = this.ansi_colors[0][i - 30];
            else if (i >= 40 && i < 48)
              this.bg = this.ansi_colors[0][i - 40];
            else if (i >= 90 && i < 98)
              this.fg = this.ansi_colors[1][i - 90];
            else if (i >= 100 && i < 108)
              this.bg = this.ansi_colors[1][i - 100];
            else if ((i === 38 || i === 48) && t.length > 0) {
              var l = i === 38, f = t.shift();
              if (f === "5" && t.length > 0) {
                var c = parseInt(t.shift(), 10);
                c >= 0 && c <= 255 && (l ? this.fg = this.palette_256[c] : this.bg = this.palette_256[c]);
              }
              if (f === "2" && t.length > 2) {
                var _ = parseInt(t.shift(), 10), p = parseInt(t.shift(), 10), b = parseInt(t.shift(), 10);
                if (_ >= 0 && _ <= 255 && p >= 0 && p <= 255 && b >= 0 && b <= 255) {
                  var x = { rgb: [_, p, b], class_name: "truecolor" };
                  l ? this.fg = x : this.bg = x;
                }
              }
            }
          }
        }, s.prototype.transform_to_html = function(e) {
          var t = e.text;
          if (t.length === 0 || (t = this.escape_txt_for_html(t), !e.bold && !e.italic && !e.underline && e.fg === null && e.bg === null))
            return t;
          var r = [], i = [], l = e.fg, f = e.bg;
          e.bold && r.push("font-weight:bold"), e.italic && r.push("font-style:italic"), e.underline && r.push("text-decoration:underline"), this._use_classes ? (l && (l.class_name !== "truecolor" ? i.push(l.class_name + "-fg") : r.push("color:rgb(" + l.rgb.join(",") + ")")), f && (f.class_name !== "truecolor" ? i.push(f.class_name + "-bg") : r.push("background-color:rgb(" + f.rgb.join(",") + ")"))) : (l && r.push("color:rgb(" + l.rgb.join(",") + ")"), f && r.push("background-color:rgb(" + f.rgb + ")"));
          var c = "", _ = "";
          return i.length && (c = ' class="' + i.join(" ") + '"'), r.length && (_ = ' style="' + r.join(";") + '"'), "<span" + _ + c + ">" + t + "</span>";
        }, s.prototype.process_hyperlink = function(e) {
          var t = e.url.split(":");
          if (t.length < 1 || !this._url_whitelist[t[0]])
            return "";
          var r = '<a href="' + this.escape_txt_for_html(e.url) + '">' + this.escape_txt_for_html(e.text) + "</a>";
          return r;
        }, s;
      }();
      function o(s) {
        var e = s.raw[0], t = /^\s+|\s+\n|\s*#[\s\S]*?\n|\n/gm, r = e.replace(t, "");
        return new RegExp(r);
      }
      function d(s) {
        var e = s.raw[0], t = /^\s+|\s+\n|\s*#[\s\S]*?\n|\n/gm, r = e.replace(t, "");
        return new RegExp(r, "g");
      }
      Object.defineProperty(u, "__esModule", { value: !0 }), u.default = g;
    });
  }(y)), y;
}
var m = V();
const $ = /* @__PURE__ */ F({
  __proto__: null,
  default: m
}, [m]), D = ["innerHTML"], X = /* @__PURE__ */ B({
  __name: "index",
  setup(h) {
    const u = I(), a = "app\x1B[0m \x1B[38;5;15;1mKoishi/";
    console.log(m, $);
    const n = new m.default();
    function g(o) {
      return n.ansi_to_html(o);
    }
    return L(() => {
      const o = u.value.parentElement.parentElement.parentElement;
      o.scrollTop = o.scrollHeight;
    }), j(() => k.logs.length, async () => {
      const o = u.value.parentElement.parentElement.parentElement, { scrollTop: d, clientHeight: s, scrollHeight: e } = o;
      Math.abs(d + s - e) < 1 && (await R(), o.scrollTop = e);
    }), (o, d) => {
      const s = C("el-scrollbar"), e = C("k-layout");
      return v(), M(e, null, {
        default: E(() => [
          U(s, { class: "container" }, {
            default: E(() => [
              S("div", {
                ref_key: "root",
                ref: u,
                class: "logs"
              }, [
                (v(!0), w(q, null, A(H(k).logs, (t) => (v(), w("div", {
                  class: z(["line", { start: t.includes(a) }])
                }, [
                  S("code", {
                    innerHTML: g(t)
                  }, null, 8, D)
                ], 2))), 256))
              ], 512)
            ]),
            _: 1
          })
        ]),
        _: 1
      });
    };
  }
});
const T = (h, u) => {
  const a = h.__vccOpts || h;
  for (const [n, g] of u)
    a[n] = g;
  return a;
}, J = /* @__PURE__ */ T(X, [["__scopeId", "data-v-58b54103"]]);
const Q = {}, W = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, Y = /* @__PURE__ */ S("path", {
  fill: "currentColor",
  d: "M304 336h-128C167.2 336 160 343.2 160 352s7.156 16 16 16h128c8.844 0 16-7.156 16-16S312.8 336 304 336zM304 240h-96C199.2 240 192 247.2 192 256s7.156 16 16 16h96C312.8 272 320 264.8 320 256S312.8 240 304 240zM96 328c-13.25 0-24 10.74-24 24c0 13.25 10.75 24 24 24c13.26 0 24-10.75 24-24C120 338.7 109.3 328 96 328zM320 64c-8.844 0-16 7.156-16 16S311.2 96 320 96c17.64 0 32 14.34 32 32v320c0 17.66-14.36 32-32 32H64c-17.64 0-32-14.34-32-32V128c0-17.66 14.36-32 32-32c8.844 0 16-7.156 16-16S72.84 64 64 64C28.7 64 0 92.72 0 128v320c0 35.28 28.7 64 64 64h256c35.3 0 64-28.72 64-64V128C384 92.72 355.3 64 320 64zM112 128h160C280.8 128 288 120.8 288 112S280.8 96 272 96h-24.88C252.6 86.55 256 75.72 256 64c0-35.35-28.65-64-64-64S128 28.65 128 64c0 11.72 3.379 22.55 8.877 32H112C103.2 96 96 103.2 96 112S103.2 128 112 128zM192 32c17.64 0 32 14.36 32 32s-14.36 32-32 32S160 81.64 160 64S174.4 32 192 32zM68.25 229.2c-6 6.5-5.594 16.59 .9062 22.59l34.66 32C106.9 286.6 110.8 288 114.7 288c4.203 0 8.406-1.656 11.55-4.938l61.33-64c6.125-6.375 5.906-16.5-.4844-22.62c-6.359-6.094-16.48-5.906-22.61 .5L113.1 249.6L90.84 228.3C84.39 222.3 74.22 222.6 68.25 229.2z"
}, null, -1), Z = [
  Y
];
function P(h, u) {
  return v(), w("svg", W, Z);
}
const K = /* @__PURE__ */ T(Q, [["render", P]]);
N.register("activity:logs", K);
const ne = (h) => {
  h.page({
    path: "/logs",
    name: "\u65E5\u5FD7",
    icon: "activity:logs",
    order: 0,
    authority: 4,
    fields: ["logs"],
    component: J
  });
};
export {
  ne as default
};
