// koishi-plugin-changelog/dist/index.js
import { store, icons } from "../client.js";
import { defineComponent, resolveComponent, openBlock, createBlock, withCtx, createVNode, createElementBlock, Fragment, renderList, unref, createElementVNode, toDisplayString } from "../vue.js";
var changelog_vue_vue_type_style_index_0_lang = "";
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  name: "changelog",
  setup(__props) {
    function transform(source) {
      return source.replace(/^[^#]+/, "").replace(/^#+ Other Changes[\s\S]+/gm, "").replace(/^## /gm, "### ").replace(/#(\d+)\b/g, (_, id) => {
        return `[#${id}](https://github.com/koishijs/koishi/issues/${id})`;
      }).replace(/\(([\w-]+\/[\w-]+@)?[0-9a-f]{40}(, ([\w-]+\/[\w-]+@)?[0-9a-f]{40})*\)/g, (match) => {
        return `(${match.slice(1, -1).split(", ").map((hash) => {
          if (hash.includes("@")) {
            const s = hash.split("@");
            return `[${s[0]}@\`${s[1].slice(0, 7)}\`](https://github.com/${s[0]}/commit/${s[1]})`;
          }
          return `[\`${hash.slice(0, 7)}\`](https://github.com/koishijs/koishi/commit/${hash})`;
        }).join(", ")})`;
      });
    }
    return (_ctx, _cache) => {
      const _component_k_markdown = resolveComponent("k-markdown");
      const _component_k_card = resolveComponent("k-card");
      const _component_el_scrollbar = resolveComponent("el-scrollbar");
      const _component_k_layout = resolveComponent("k-layout");
      return openBlock(), createBlock(_component_k_layout, {
        main: "darker",
        class: "page-changelog"
      }, {
        default: withCtx(() => [
          createVNode(_component_el_scrollbar, null, {
            default: withCtx(() => [
              (openBlock(true), createElementBlock(Fragment, null, renderList(unref(store).releases, ({ name, body, tag_name }) => {
                return openBlock(), createBlock(_component_k_card, {
                  key: tag_name,
                  class: "xxx"
                }, {
                  default: withCtx(() => [
                    createElementVNode("h2", null, toDisplayString(name), 1),
                    createVNode(_component_k_markdown, {
                      source: transform(body)
                    }, null, 8, ["source"])
                  ]),
                  _: 2
                }, 1024);
              }), 128))
            ]),
            _: 1
          })
        ]),
        _: 1
      });
    };
  }
});
var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};
const _sfc_main = {};
const _hoisted_1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
};
const _hoisted_2 = /* @__PURE__ */ createElementVNode("path", {
  fill: "currentColor",
  d: "M256 0C114.6 0 0 114.6 0 256s114.6 256 256 256s256-114.6 256-256S397.4 0 256 0zM256 480c-123.5 0-224-100.5-224-224s100.5-224 224-224s224 100.5 224 224S379.5 480 256 480zM256 184c13.25 0 24-10.74 24-24c0-13.25-10.75-24-24-24S232 146.7 232 160C232 173.3 242.7 184 256 184zM304 352h-32V240C272 231.2 264.8 224 256 224H224C215.2 224 208 231.2 208 240S215.2 256 224 256h16v96h-32C199.2 352 192 359.2 192 368C192 376.8 199.2 384 208 384h96c8.836 0 16-7.164 16-16C320 359.2 312.8 352 304 352z"
}, null, -1);
const _hoisted_3 = [
  _hoisted_2
];
function _sfc_render(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1, _hoisted_3);
}
var Info = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
icons.register("info", Info);
var index = (ctx) => {
  ctx.page({
    path: "/changelog",
    name: "\u66F4\u65B0\u65E5\u5FD7",
    icon: "info",
    position: "bottom",
    fields: ["releases"],
    component: _sfc_main$1
  });
};
export { index as default };
