// @koishijs/plugin-insight/lib/index.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// external/console/packages/insight/src/index.ts
import { camelize, capitalize, Context, Schema } from "koishi";
import { debounce } from "throttle-debounce";
import { DataService } from "@koishijs/plugin-console";
import { resolve } from "path";
function format(name) {
  return capitalize(camelize(name));
}
__name(format, "format");
function getName(plugin) {
  if (!plugin)
    return "App";
  if (!plugin.name || plugin.name === "apply")
    return "Anonymous";
  return format(plugin.name);
}
__name(getName, "getName");
function getSourceId(child) {
  const { state } = child.parent;
  if (state.runtime.isForkable) {
    return state.uid;
  } else {
    return state.runtime.uid;
  }
}
__name(getSourceId, "getSourceId");
var Insight = class extends DataService {
  constructor(ctx) {
    super(ctx, "insight");
    ctx.console.addEntry(process.env.KOISHI_BASE ? [
      process.env.KOISHI_BASE + "/dist/index.js",
      process.env.KOISHI_BASE + "/dist/style.css"
    ] : process.env.KOISHI_ENV === "browser" ? [
      import.meta.url.replace(/\/src\/[^/]+$/, "/client/index.ts")
    ] : {
      dev: resolve(__dirname, "../client/index.ts"),
      prod: resolve(__dirname, "../dist")
    });
    ctx.on("internal/fork", this.update);
    ctx.on("internal/runtime", this.update);
    ctx.on("internal/service", this.update);
  }
  stop() {
    this.update.cancel();
  }
  update = debounce(0, () => {
    this.refresh();
  });
  async get() {
    const nodes = [];
    const edges = [];
    for (const runtime of this.ctx.registry.values()) {
      let isActive = function(state) {
        return runtime.using.every((name2) => state.ctx[name2]);
      }, addNode = function(state) {
        const { uid, alias, disposables } = state;
        const weight = disposables.length;
        const node = { uid, name, weight };
        if (alias)
          node.name += ` <${format(alias)}>`;
        nodes.push(node);
      }, addEdge = function(type, source, target) {
        edges.push({ type, source, target });
      }, addDeps = function(state) {
        for (const name2 of runtime.using) {
          const ctx = state.ctx[name2][Context.source];
          const uid = ctx?.state.uid;
          if (!uid)
            continue;
          addEdge("dashed", state.uid, uid);
        }
      };
      __name(isActive, "isActive");
      __name(addNode, "addNode");
      __name(addEdge, "addEdge");
      __name(addDeps, "addDeps");
      const name = getName(runtime.plugin);
      const isReusable = runtime.plugin?.["reusable"];
      if (!isReusable) {
        if (!isActive(runtime))
          continue;
        addNode(runtime);
        addDeps(runtime);
      }
      for (const fork of runtime.children) {
        if (runtime.isForkable) {
          if (!isActive(fork))
            continue;
          addNode(fork);
          addEdge("solid", getSourceId(fork), fork.uid);
          if (!isReusable) {
            addEdge("solid", fork.uid, runtime.uid);
          } else {
            addDeps(fork);
          }
        } else {
          nodes[nodes.length - 1].weight += fork.disposables.length;
          addEdge("solid", getSourceId(fork), runtime.uid);
        }
      }
    }
    return { nodes, edges };
  }
};
__name(Insight, "Insight");
((Insight2) => {
  Insight2.using = ["console"];
  Insight2.Config = Schema.object({});
})(Insight || (Insight = {}));
var src_default = Insight;
export {
  src_default as default
};
