// @koishijs/plugin-database-mongo/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// koishi/plugins/database/mongo/src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");
var import_driver_mongo = __toESM(require("@minatojs/driver-mongo"));
var src_default = (0, import_koishi.defineDriver)(import_driver_mongo.default, import_koishi.Schema.object({
  protocol: import_koishi.Schema.string().description("要使用的协议名。").default("mongodb"),
  host: import_koishi.Schema.string().description("要连接到的主机名。").default("localhost"),
  port: import_koishi.Schema.natural().max(65535).description("要连接到的端口号。"),
  username: import_koishi.Schema.string().description("要使用的用户名。"),
  password: import_koishi.Schema.string().description("要使用的密码。").role("secret"),
  database: import_koishi.Schema.string().description("要访问的数据库名。").default("koishi")
}));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
