// koishi-plugin-dailygames/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apply = exports.Config = exports.name = void 0;
const koishi_1 = require("koishi");
exports.name = 'dailygames';
exports.Config = koishi_1.Schema.object({
    topics: koishi_1.Schema.boolean().description('是否加上近期热点？').default(true),
});
const axios = require('axios');

function apply(ctx,config) {
   ctx.command('发售', "近期即将发售的游戏")
  .action(async ({ options }) => {
    let success = false;
    let data;
    let error;
    for (let i = 0; i < 3; i++) {
      try {
        const response = await axios.get('http://app02.vgtime.com:8080/vgtime-app/api/v3/launch/wallpaper_today');
        data = response.data.data;
        success = true;
        break;
      } catch (err) {
        error = err;
      }
    }

    if (success) {
  const pictureUrl1 = decodeURIComponent(data.will_be_onsale_games.list[0].cover_vgtime);
  const pictureUrl2 = decodeURIComponent(data.will_be_onsale_games.list[1].cover_vgtime);
  const pictureUrl3 = decodeURIComponent(data.will_be_onsale_games.list[2].cover_vgtime);
  const imageData1 = await ctx.http.get(pictureUrl1, { responseType: 'arraybuffer' });
  const imageData2 = await ctx.http.get(pictureUrl2, { responseType: 'arraybuffer' });
  const imageData3 = await ctx.http.get(pictureUrl3, { responseType: 'arraybuffer' });
  const games = data.will_be_onsale_games.list.map((game, index) => {
    return {
      gameInfo: `\n\n游戏名：${game.title}\n发售日期：${game.publish_date}\n登陆平台：${game.platforms}\n简介：${game.recommend}`,
      imageData: [imageData1, imageData2, imageData3][index]
    };
   
  // 其他代码
      });
      return games.map(game => `${game.gameInfo}\n${koishi_1.segment.image(game.imageData)}\n\n`).join('');
    } else {
      return `请求失败：${error.message}`;
    }
  });




  ctx.command('daily', "查看游戏今昔史")
  .action(async ({ options }) => {
    let success = false;
    let data;
    let error;
    for (let i = 0; i < 3; i++) {
      try {
        const response = await axios.get('http://app02.vgtime.com:8080/vgtime-app/api/v3/launch/wallpaper_today');
        data = response.data.data;
        success = true;
        break;
      } catch (err) {
        error = err;
      }
    }

    if (success) {
      const date = new Date().toLocaleDateString('zh-CN', {
        month: 'long',
        day: 'numeric'
      });
      const history = data.user_sign.history_in_today.toString();

      // 处理 topic
      const topics = data.hot_topics.map(topic => {
        const objectId = topic.object_id;
        const topicUrl = `https://www.vgtime.com/topic/${objectId}.jhtml`;
        return `${topic.content}\n链接：${topicUrl}`;
      });
      const topicString = topics.join('\n\n');
      data.topic = topicString;

      // 获取图片
      const pictureUrl = decodeURIComponent(data.wallpaper.picture);
      const imageData = await ctx.http.get(pictureUrl, { responseType: 'arraybuffer' });

      if (config.topics) {
        return `历史上的 ${date}\n\n${history.replace(/<br\/>/g, '\n').replace(/^<br>/, '')}\n\n近期热点：\n\n${data.topic}\n\n${koishi_1.segment.image(imageData)}`;
      } else {
        return `历史上的 ${date}\n\n${history.replace(/<br\/>/g, '\n').replace(/^<br>/, '')}\n\n${koishi_1.segment.image(imageData)}`;
      }
    } else {
      return `请求失败：${error.message}`;
    }
  });

}

exports.apply = apply;
