// minato-decorators/dist/src/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelDecorators = void 0;
const meta_1 = require("./meta/meta");
const utils_1 = require("./utils");
class ModelDecorators {
    constructor() {
        this.DefineModel = (name) => meta_1.Metadata.set('ModelTableName', name);
        this.ModelField = (def) => (obj, key) => meta_1.Metadata.set('ModelField', def || (0, utils_1.inferType)(Reflect.getMetadata('design:type', obj, key)), 'ModelFieldKeys')(obj, key);
        this.Primary = () => meta_1.Metadata.set('ModelPrimaryKey', { autoIncrement: false });
        this.PrimaryGenerated = () => meta_1.Metadata.set('ModelPrimaryKey', { autoIncrement: true });
        this.Foreign = (referencingTable, referencingColumn) => meta_1.Metadata.set('ModelForeignKey', [referencingTable, referencingColumn]);
        this.Unique = (identifier = '_default') => meta_1.Metadata.append('ModelUnique', identifier);
        this.ChildModel = (cls) => (obj, key) => {
            if (!cls) {
                cls = Reflect.getMetadata('design:type', obj, key);
            }
            meta_1.Metadata.set('ChildModel', cls, 'ChildModelKeys')(obj, key);
        };
    }
}
exports.ModelDecorators = ModelDecorators;
