// @koishijs/plugin-auth/dist/index.js
import { createStorage as N, store as d, router as E, message as b, send as S, Schema as H, icons as B } from "../client.js";
import { ref as M, watch as R, defineComponent as L, resolveComponent as g, openBlock as n, createElementBlock as l, Fragment as V, unref as s, createElementVNode as a, toDisplayString as $, createVNode as r, withCtx as i, createTextVNode as F, withKeys as K, withModifiers as T, createCommentVNode as O, createBlock as U, computed as j, renderList as W, isRef as q } from "../vue.js";
import { useRouter as G } from "../vue-router.js";
const t = N("auth", 1, () => ({
  authType: 0
})), x = M(!1);
R(() => d.user, (o, e) => {
  if (x.value = !1, !o)
    return E.push("/login");
  if (e)
    return;
  b.success(`\u6B22\u8FCE\u56DE\u6765\uFF0C${o.name || "Koishi \u7528\u6237"}\uFF01`), Object.assign(t, o);
  const u = E.currentRoute.value.redirectedFrom;
  u && !u.path.startsWith("/login") ? E.push(u) : E.push("/profile");
});
async function P(o) {
  const e = new TextEncoder().encode(o), u = await crypto.subtle.digest("SHA-256", e), _ = new DataView(u);
  let p = "";
  for (let k = 0; k < _.byteLength; k += 4)
    p += ("00000000" + _.getUint32(k).toString(16)).slice(-8);
  return p;
}
const J = { class: "login-form" }, Q = { key: 0 }, X = /* @__PURE__ */ a("span", null, "\u5E73\u53F0\u8D26\u6237\u7ED1\u5B9A", -1), Y = [
  X
], Z = { key: 1 }, ee = /* @__PURE__ */ a("span", null, "\u5E73\u53F0\u8D26\u6237\u767B\u5F55", -1), te = [
  ee
], oe = { class: "hint" }, se = /* @__PURE__ */ a("p", { class: "hint" }, "\u8BF7\u7528\u4E0A\u8FF0\u8D26\u53F7\u5C06\u4E0B\u9762\u7684\u9A8C\u8BC1\u7801\u53D1\u9001\u7ED9\u4EFB\u610F\u6B63\u5728\u8FD0\u884C\u7684\u673A\u5668\u4EBA", -1), ne = { class: "token" }, ue = { class: "control" }, ce = { key: 0 }, re = /* @__PURE__ */ a("span", null, "\u5E73\u53F0\u8D26\u6237\u7ED1\u5B9A", -1), le = [
  re
], ae = { key: 1 }, ie = { key: 2 }, _e = /* @__PURE__ */ a("span", null, "\u5E73\u53F0\u8D26\u6237\u767B\u5F55", -1), de = [
  _e
], pe = {
  key: 0,
  class: "error"
}, he = { class: "control" }, me = {
  key: 0,
  class: "error"
}, fe = { class: "control" }, I = /* @__PURE__ */ L({
  __name: "login-form",
  setup(o) {
    const e = isSecureContext;
    e || (t.authType = 0);
    const u = M(), _ = M();
    let p = 0;
    async function k() {
      var C;
      const y = Date.now();
      if (y < p)
        return;
      const { platform: c, userId: m } = t;
      if (!(!c || !m)) {
        p = y + 1e3;
        try {
          _.value = await S("login/platform", c, m, (C = d.user) == null ? void 0 : C.id);
        } catch (v) {
          u.value = v.message;
        }
      }
    }
    async function h() {
      const { name: y, password: c } = t;
      try {
        await S("login/password", y, await P(c));
      } catch (m) {
        u.value = m.message;
      }
    }
    const w = G();
    function z() {
      d.user ? x.value = !1 : w.back();
    }
    return (y, c) => {
      const m = g("k-button"), C = g("k-tab"), v = g("k-icon"), A = g("el-input");
      return n(), l("div", J, [
        _.value ? (n(), l(V, { key: 0 }, [
          s(d).user ? (n(), l("h1", Q, Y)) : (n(), l("h1", Z, te)),
          a("p", oe, "\u6B22\u8FCE\u4F60\uFF0C" + $(_.value.name || "Koishi \u7528\u6237") + "\uFF01", 1),
          se,
          a("p", ne, $(_.value.token), 1),
          a("div", ue, [
            r(m, {
              onClick: c[0] || (c[0] = (f) => _.value.token = null)
            }, {
              default: i(() => [
                F("\u8FD4\u56DE\u4E0A\u4E00\u6B65")
              ]),
              _: 1
            })
          ])
        ], 64)) : (n(), l(V, { key: 1 }, [
          s(d).user ? (n(), l("h1", ce, le)) : s(e) ? (n(), l("h1", ae, [
            r(C, {
              data: ["\u5E73\u53F0\u8D26\u6237\u767B\u5F55", "\u7528\u6237\u540D\u5BC6\u7801\u767B\u5F55"],
              modelValue: s(t).authType,
              "onUpdate:modelValue": c[1] || (c[1] = (f) => s(t).authType = f)
            }, null, 8, ["modelValue"])
          ])) : (n(), l("h1", ie, de)),
          s(d).user || s(t).authType === 0 ? (n(), l(V, { key: 3 }, [
            r(A, {
              placeholder: "\u5E73\u53F0\u540D",
              modelValue: s(t).platform,
              "onUpdate:modelValue": c[2] || (c[2] = (f) => s(t).platform = f)
            }, {
              prefix: i(() => [
                r(v, { name: "at" })
              ]),
              _: 1
            }, 8, ["modelValue"]),
            r(A, {
              placeholder: "\u8D26\u53F7",
              modelValue: s(t).userId,
              "onUpdate:modelValue": c[3] || (c[3] = (f) => s(t).userId = f),
              onKeypress: K(T(k, ["stop"]), ["enter"])
            }, {
              prefix: i(() => [
                r(v, { name: "user" })
              ]),
              _: 1
            }, 8, ["modelValue", "onKeypress"]),
            u.value ? (n(), l("p", pe, $(u.value), 1)) : O("", !0),
            a("div", he, [
              r(m, { onClick: z }, {
                default: i(() => [
                  F("\u8FD4\u56DE")
                ]),
                _: 1
              }),
              r(m, { onClick: k }, {
                default: i(() => [
                  F("\u83B7\u53D6\u9A8C\u8BC1\u7801")
                ]),
                _: 1
              })
            ])
          ], 64)) : (n(), l(V, { key: 4 }, [
            r(A, {
              placeholder: "\u7528\u6237\u540D",
              modelValue: s(t).name,
              "onUpdate:modelValue": c[4] || (c[4] = (f) => s(t).name = f)
            }, {
              prefix: i(() => [
                r(v, { name: "user" })
              ]),
              _: 1
            }, 8, ["modelValue"]),
            r(A, {
              placeholder: "\u5BC6\u7801",
              modelValue: s(t).password,
              "onUpdate:modelValue": c[6] || (c[6] = (f) => s(t).password = f),
              onKeypress: K(T(h, ["stop"]), ["enter"]),
              type: s(t).showPass ? "text" : "password"
            }, {
              prefix: i(() => [
                r(v, { name: "lock" })
              ]),
              suffix: i(() => [
                r(v, {
                  name: s(t).showPass ? "eye" : "eye-slash",
                  onClick: c[5] || (c[5] = (f) => s(t).showPass = !s(t).showPass)
                }, null, 8, ["name"])
              ]),
              _: 1
            }, 8, ["modelValue", "onKeypress", "type"]),
            u.value ? (n(), l("p", me, $(u.value), 1)) : O("", !0),
            a("div", fe, [
              r(m, { onClick: z }, {
                default: i(() => [
                  F("\u8FD4\u56DE")
                ]),
                _: 1
              }),
              r(m, { onClick: h }, {
                default: i(() => [
                  F("\u767B\u5F55")
                ]),
                _: 1
              })
            ])
          ], 64))
        ], 64))
      ]);
    };
  }
});
const ge = /* @__PURE__ */ L({
  __name: "login",
  setup(o) {
    return (e, u) => {
      const _ = g("k-card"), p = g("k-layout");
      return n(), U(p, { main: "darker page-login" }, {
        default: i(() => [
          r(_, null, {
            default: i(() => [
              r(I)
            ]),
            _: 1
          })
        ]),
        _: 1
      });
    };
  }
});
const Ce = { class: "k-schema-header" }, ve = { class: "schema-item" }, ke = { class: "header" }, we = { class: "left" }, ye = /* @__PURE__ */ a("div", { class: "right" }, null, -1), Fe = /* @__PURE__ */ L({
  __name: "profile",
  setup(o) {
    const e = M({}), u = j(() => {
      const h = H.object({
        name: H.string().description("\u7528\u6237\u540D").default(t.name)
      }).description("\u57FA\u672C\u8D44\u6599");
      return isSecureContext && (h.dict.password = H.string().role("secret").description("\u5BC6\u7801").default(t.password)), h;
    });
    async function _() {
      return delete t.id, delete t.token, delete t.expire, S("user/logout");
    }
    async function p() {
      const h = { ...e.value };
      h.password && (h.password = await P(h.password));
      try {
        await S("user/update", h), b.success("\u4FEE\u6539\u6210\u529F\uFF01"), Object.assign(t, e), Object.assign(d.user, h), e.value = {};
      } catch (w) {
        b.error(w.message);
      }
    }
    const k = j(() => [{
      icon: "check",
      label: "\u5E94\u7528\u66F4\u6539",
      disabled: !e.value || !Object.keys(e.value).length,
      action: p
    }, {
      type: "error",
      icon: "sign-out",
      label: "\u9000\u51FA\u767B\u5F55",
      action: _
    }]);
    return (h, w) => {
      const z = g("el-button"), y = g("k-form"), c = g("k-content"), m = g("k-layout");
      return n(), U(m, {
        main: "page-profile",
        menu: s(k)
      }, {
        default: i(() => [
          r(c, null, {
            default: i(() => [
              r(y, {
                schema: s(u),
                modelValue: e.value,
                "onUpdate:modelValue": w[1] || (w[1] = (C) => e.value = C)
              }, {
                epilog: i(() => [
                  a("h2", Ce, [
                    F(" \u5E73\u53F0\u8D26\u6237\u7ED1\u5B9A "),
                    r(z, {
                      solid: "",
                      onClick: w[0] || (w[0] = (C) => x.value = !0)
                    }, {
                      default: i(() => [
                        F("\u6DFB\u52A0")
                      ]),
                      _: 1
                    })
                  ]),
                  (n(!0), l(V, null, W(s(d).user.accounts, ({ platform: C, id: v }) => (n(), l("div", ve, [
                    a("div", ke, [
                      a("div", we, $(C) + " (" + $(v) + ")", 1),
                      ye
                    ])
                  ]))), 256))
                ]),
                _: 1
              }, 8, ["schema", "modelValue"])
            ]),
            _: 1
          })
        ]),
        _: 1
      }, 8, ["menu"]);
    };
  }
});
const D = (o, e) => {
  const u = o.__vccOpts || o;
  for (const [_, p] of e)
    u[_] = p;
  return u;
}, $e = {}, xe = {
  class: "k-icon k-icon-at",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Be = /* @__PURE__ */ a("path", {
  fill: "currentColor",
  d: "M256 8C118.941 8 8 118.919 8 256c0 137.059 110.919 248 248 248 48.154 0 95.342-14.14 135.408-40.223 12.005-7.815 14.625-24.288 5.552-35.372l-10.177-12.433c-7.671-9.371-21.179-11.667-31.373-5.129C325.92 429.757 291.314 440 256 440c-101.458 0-184-82.542-184-184S154.542 72 256 72c100.139 0 184 57.619 184 160 0 38.786-21.093 79.742-58.17 83.693-17.349-.454-16.91-12.857-13.476-30.024l23.433-121.11C394.653 149.75 383.308 136 368.225 136h-44.981a13.518 13.518 0 0 0-13.432 11.993l-.01.092c-14.697-17.901-40.448-21.775-59.971-21.775-74.58 0-137.831 62.234-137.831 151.46 0 65.303 36.785 105.87 96 105.87 26.984 0 57.369-15.637 74.991-38.333 9.522 34.104 40.613 34.103 70.71 34.103C462.609 379.41 504 307.798 504 232 504 95.653 394.023 8 256 8zm-21.68 304.43c-22.249 0-36.07-15.623-36.07-40.771 0-44.993 30.779-72.729 58.63-72.729 22.292 0 35.601 15.241 35.601 40.77 0 45.061-33.875 72.73-58.161 72.73z"
}, null, -1), De = [
  Be
];
function Ee(o, e) {
  return n(), l("svg", xe, De);
}
const Ve = /* @__PURE__ */ D($e, [["render", Ee]]), be = {}, Se = {
  class: "k-icon check",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 448 512"
}, ze = /* @__PURE__ */ a("path", {
  fill: "currentColor",
  d: "M438.6 105.4C451.1 117.9 451.1 138.1 438.6 150.6L182.6 406.6C170.1 419.1 149.9 419.1 137.4 406.6L9.372 278.6C-3.124 266.1-3.124 245.9 9.372 233.4C21.87 220.9 42.13 220.9 54.63 233.4L159.1 338.7L393.4 105.4C405.9 92.88 426.1 92.88 438.6 105.4H438.6z"
}, null, -1), Ae = [
  ze
];
function Me(o, e) {
  return n(), l("svg", Se, Ae);
}
const Le = /* @__PURE__ */ D(be, [["render", Me]]), He = {}, Ue = {
  class: "k-icon k-icon-lock",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 448 512"
}, Ke = /* @__PURE__ */ a("path", {
  fill: "currentColor",
  d: "M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
}, null, -1), Te = [
  Ke
];
function Oe(o, e) {
  return n(), l("svg", Ue, Te);
}
const je = /* @__PURE__ */ D(He, [["render", Oe]]), Pe = {}, Ie = {
  class: "k-icon k-icon-sign-in",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Ne = /* @__PURE__ */ a("path", {
  fill: "currentColor",
  d: "M384 256c0-8.188-3.125-16.38-9.375-22.62l-128-128C237.5 96.22 223.7 93.47 211.8 98.44C199.8 103.4 192 115.1 192 128v64H48C21.49 192 0 213.5 0 240v32C0 298.5 21.49 320 48 320H192v64c0 12.94 7.797 24.62 19.75 29.56c11.97 4.969 25.72 2.219 34.88-6.938l128-128C380.9 272.4 384 264.2 384 256zM224 384V288H48C39.18 288 32 280.8 32 272v-32C32 231.2 39.18 224 48 224H224L223.1 128l128 128L224 384zM432 32h-96C327.2 32 320 39.16 320 48S327.2 64 336 64h96C458.5 64 480 85.53 480 112v288c0 26.47-21.53 48-48 48h-96c-8.844 0-16 7.156-16 16s7.156 16 16 16h96c44.13 0 80-35.88 80-80v-288C512 67.88 476.1 32 432 32z"
}, null, -1), Re = [
  Ne
];
function We(o, e) {
  return n(), l("svg", Ie, Re);
}
const qe = /* @__PURE__ */ D(Pe, [["render", We]]), Ge = {}, Je = {
  class: "k-icon k-icon-sign-out",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Qe = /* @__PURE__ */ a("path", {
  fill: "currentColor",
  d: "M176 448h-96C53.53 448 32 426.5 32 400v-288C32 85.53 53.53 64 80 64h96C184.8 64 192 56.84 192 48S184.8 32 176 32h-96C35.88 32 0 67.88 0 112v288C0 444.1 35.88 480 80 480h96C184.8 480 192 472.8 192 464S184.8 448 176 448zM502.6 233.4l-128-128c-9.156-9.156-22.91-11.91-34.88-6.938C327.8 103.4 320 115.1 320 128l.0918 63.1L176 192C149.5 192 128 213.5 128 240v32C128 298.5 149.5 320 176 320l144.1-.001L320 384c0 12.94 7.797 24.62 19.75 29.56c11.97 4.969 25.72 2.219 34.88-6.938l128-128C508.9 272.4 512 264.2 512 256S508.9 239.6 502.6 233.4zM352 384V288H176C167.2 288 160 280.8 160 272v-32C160 231.2 167.2 224 176 224H352l-.0039-96l128 128L352 384z"
}, null, -1), Xe = [
  Qe
];
function Ye(o, e) {
  return n(), l("svg", Je, Xe);
}
const Ze = /* @__PURE__ */ D(Ge, [["render", Ye]]), et = {}, tt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, ot = /* @__PURE__ */ a("path", {
  fill: "currentColor",
  d: "M256 0C114.6 0 0 114.6 0 256s114.6 256 256 256s256-114.6 256-256S397.4 0 256 0zM256 480c-47.24 0-91.04-14.78-127.2-39.84C132.9 390.9 173.8 352 224 352h64c50.25 0 91.14 38.94 95.21 88.16C347 465.2 303.2 480 256 480zM411.7 416.7C397.6 361.3 347.7 320 288 320H224c-59.73 0-109.6 41.3-123.7 96.72C58.27 375.1 32 319 32 256c0-123.5 100.5-224 224-224s224 100.5 224 224C480 319 453.7 375.1 411.7 416.7zM256 128C211.8 128 176 163.8 176 208C176 252.2 211.8 288 256 288s80-35.82 80-80C336 163.8 300.2 128 256 128zM256 256C229.5 256 208 234.5 208 208S229.5 160 256 160s48 21.53 48 48S282.5 256 256 256z"
}, null, -1), st = [
  ot
];
function nt(o, e) {
  return n(), l("svg", tt, st);
}
const ut = /* @__PURE__ */ D(et, [["render", nt]]), ct = /* @__PURE__ */ L({
  __name: "bind-dialog",
  setup(o) {
    return (e, u) => {
      const _ = g("el-dialog");
      return n(), U(_, {
        center: "",
        modelValue: s(x),
        "onUpdate:modelValue": u[0] || (u[0] = (p) => q(x) ? x.value = p : null),
        class: "bind-dialog"
      }, {
        default: i(() => [
          r(I)
        ]),
        _: 1
      }, 8, ["modelValue"]);
    };
  }
});
B.register("at", Ve);
B.register("check", Le);
B.register("lock", je);
B.register("sign-in", qe);
B.register("sign-out", Ze);
B.register("user-full", ut);
const it = (o) => {
  t.token && t.expire > Date.now() && S("login/token", t.id, t.token).catch((e) => b.error(e.message)), o.on("activity", (e) => e.authority > 0 && (!d.user || d.user.authority < e.authority)), o.state.disposables.push(E.beforeEach((e) => {
    if ((e.meta.authority || e.meta.fields.includes("user")) && !d.user)
      return history.state.forward === "/login" ? "/" : "/login";
    if (e.meta.authority && e.meta.authority > d.user.authority)
      return b.error("\u6743\u9650\u4E0D\u8DB3\u3002"), !1;
  })), o.page({
    path: "/login",
    name: "\u767B\u5F55",
    icon: "sign-in",
    position: () => d.user ? "hidden" : "bottom",
    component: ge
  }), o.page({
    path: "/profile",
    name: "\u7528\u6237\u8D44\u6599",
    icon: "user-full",
    fields: ["user"],
    position: () => d.user ? "bottom" : "hidden",
    component: Fe
  }), o.slot({
    type: "global",
    component: ct
  });
};
export {
  it as default
};
