// koishi-plugin-envinfo/dist/index.js
import { defineComponent as g, resolveComponent as h, openBlock as c, createBlock as x, withCtx as _, createElementBlock as a, Fragment as m, renderList as f, unref as u, createElementVNode as C, toDisplayString as i, createTextVNode as b } from "../vue.js";
import { store as p, capitalize as y, message as k } from "../client.js";
async function F(n) {
  try {
    return navigator.clipboard.writeText(n);
  } catch {
    const e = document.createElement("textarea"), o = document.activeElement;
    e.value = n, e.setAttribute("readonly", ""), e.style.contain = "strict", e.style.position = "absolute", e.style.left = "-9999px", e.style.fontSize = "12pt";
    const t = document.getSelection(), s = t ? t.rangeCount > 0 && t.getRangeAt(0) : null;
    document.body.appendChild(e), e.select(), e.selectionStart = 0, e.selectionEnd = n.length, document.execCommand("copy"), document.body.removeChild(e), s && (t.removeAllRanges(), t.addRange(s)), o && o.focus();
  }
}
const E = { class: "section" }, S = { class: "title" }, j = /* @__PURE__ */ g({
  __name: "status",
  setup(n) {
    async function e() {
      const o = Object.entries(p.envinfo).map(([t, s]) => `${y(t)}:
` + Object.entries(s).map(([r, l]) => `    ${r}: ${l}`).join(`
`)).join(`

`);
      await F(o), k.success("\u5DF2\u590D\u5236\u73AF\u5883\u4FE1\u606F\uFF01");
    }
    return (o, t) => {
      const s = h("k-status");
      return c(), x(s, {
        class: "version",
        onClick: e
      }, {
        tooltip: _(() => [
          (c(!0), a(m, null, f(u(p).envinfo, (r, l) => (c(), a("div", E, [
            C("p", S, i(u(y)(l)) + ":", 1),
            (c(!0), a(m, null, f(r, (v, d) => (c(), a("p", {
              class: "info",
              key: d
            }, i(d) + ": " + i(v), 1))), 128))
          ]))), 256))
        ]),
        default: _(() => [
          b(" Koishi v" + i(u(p).envinfo.koishi.Core) + " ", 1)
        ]),
        _: 1
      });
    };
  }
});
const A = (n, e) => {
  const o = n.__vccOpts || n;
  for (const [t, s] of e)
    o[t] = s;
  return o;
}, R = /* @__PURE__ */ A(j, [["__scopeId", "data-v-64987f47"]]), D = (n) => {
  n.slot({
    type: "status-left",
    component: R
  });
};
export {
  D as default
};
