// satori-decorators/dist/src/decorators/common.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnSelection = exports.OnPrivate = exports.OnPlatform = exports.OnChannel = exports.OnGuild = exports.OnSelf = exports.OnUser = exports.OnNowhere = exports.OnAnywhere = void 0;
const registrar_1 = require("../registrar");
_a = registrar_1.defaultRegistrarS.selectorDecorators(), exports.OnAnywhere = _a.OnAnywhere, exports.OnNowhere = _a.OnNowhere, exports.OnUser = _a.OnUser, exports.OnSelf = _a.OnSelf, exports.OnGuild = _a.OnGuild, exports.OnChannel = _a.OnChannel, exports.OnPlatform = _a.OnPlatform, exports.OnPrivate = _a.OnPrivate, exports.OnSelection = _a.OnSelection;
__exportStar(require("cordis-decorators/dist/src/decorators/common"), exports);
