// @koishijs/plugin-status/dist/index.js
import { Card as y, store as d, icons as C, defineExtension as O } from "../client.js";
import { defineAsyncComponent as Z, h as D, resolveComponent as g, openBlock as r, createElementBlock as u, defineComponent as V, createElementVNode as c, normalizeStyle as H, createVNode as _, withCtx as F, normalizeClass as S, createTextVNode as A, toDisplayString as p, computed as w, createBlock as b, unref as i, createCommentVNode as j, Fragment as k, renderList as z } from "../vue.js";
const P = Z(() => import("./echarts.99a7c6df.mjs"));
function B({ title: t, fields: e, options: n }) {
  return y.create(() => {
    const o = n(d);
    if (!!o)
      return D(g("k-card"), { class: "frameless", title: t }, () => D(P, { option: o, autoresize: !0 }));
  }, e);
}
var v;
((t) => {
  t.item = (e) => ({
    trigger: "item",
    formatter: e
  }), t.axis = (e) => ({
    trigger: "axis",
    axisPointer: {
      type: "cross"
    },
    formatter: e
  });
})(v || (v = {}));
const N = (t) => {
  t.slot({
    type: "chart",
    component: B({
      title: "\u6307\u4EE4\u8C03\u7528\u9891\u7387",
      fields: ["stats"],
      options({ stats: e }) {
        const n = Object.entries(e.commands).sort((o, s) => s[1] - o[1]).map(([o, s]) => ({ name: o, value: s }));
        if (!!n.length)
          return {
            tooltip: v.item(({ data: o }) => {
              const s = [o.name];
              return s.push(`\u65E5\u5747\u8C03\u7528\uFF1A${o.value}`), s.join("<br>");
            }),
            series: [{
              type: "pie",
              data: n,
              radius: ["35%", "65%"],
              minShowLabelAngle: 3
            }]
          };
      }
    })
  });
}, U = (t) => {
  t.slot({
    type: "chart",
    component: B({
      title: "\u5404\u7FA4\u53D1\u8A00\u6570\u91CF",
      fields: ["stats"],
      options({ stats: e }) {
        if (!!e.guilds.length)
          return {
            tooltip: v.item(({ data: n }) => {
              const o = [n.name];
              return o.push(`\u5E73\u53F0\uFF1A${n.platform}`), n.memberCount && o.push(`\u4EBA\u6570\uFF1A${n.memberCount}`), n.assignee && o.push(`\u63A5\u5165\uFF1A${n.assignee}`), o.push(`\u65E5\u5747\u53D1\u8A00\uFF1A${+n.value.toFixed(1)}`), o.push(`\u6628\u65E5\u53D1\u8A00\uFF1A${+n.last.toFixed(1)}`), o.join("<br>");
            }),
            series: [{
              type: "pie",
              data: e.guilds.sort((n, o) => o.value - n.value),
              radius: ["35%", "65%"],
              minShowLabelAngle: 3
            }]
          };
      }
    })
  });
}, R = "\u65E5\u4E00\u4E8C\u4E09\u56DB\u4E94\u516D", G = (t) => {
  t.slot({
    type: "chart",
    component: B({
      title: "\u5386\u53F2\u53D1\u8A00\u6570\u91CF",
      fields: ["stats"],
      options({ stats: e }) {
        if (!!Object.keys(e.history).length)
          return {
            tooltip: v.axis(([{ name: n, value: o }]) => {
              const s = new Date(n).getDay();
              return `${n} \u661F\u671F${R[s]}<br>\u53D1\u8A00\u6570\u91CF\uFF1A${o}`;
            }),
            xAxis: {
              type: "category",
              data: Object.keys(e.history).reverse()
            },
            yAxis: {
              type: "value"
            },
            series: {
              type: "line",
              smooth: !0,
              data: Object.values(e.history).reverse()
            }
          };
      }
    })
  });
}, M = (t) => `${(t - 0.5).toFixed()}:00-${(t + 0.5).toFixed()}:00`, q = (t) => {
  t.slot({
    type: "chart",
    component: B({
      title: "\u6BCF\u5C0F\u65F6\u53D1\u8A00\u6570\u91CF",
      fields: ["stats"],
      options({ stats: e }) {
        return {
          tooltip: v.axis((n) => {
            const [{ data: [o], dataIndex: s }] = n, a = e.hours[s], f = [
              `${M(o)}`,
              `\u6D88\u606F\u603B\u91CF\uFF1A${+(a.total || 0).toFixed(1)}`
            ];
            return n.reverse().forEach(({ seriesName: m, marker: h }, $) => {
              const x = $ === 0 ? a.command : $ === 1 ? a.dialogue : Math.max(0, a.total - a.command - a.dialogue);
              !x || f.push(`${h}${m}\uFF1A${+x.toFixed(1)}`);
            }), f.join("<br>");
          }),
          xAxis: {
            type: "value",
            min: 0,
            max: 24,
            minInterval: 1,
            maxInterval: 4,
            axisLabel: {
              formatter: (n) => n + ":00"
            },
            axisPointer: {
              label: {
                formatter: ({ value: n }) => M(n)
              }
            }
          },
          yAxis: {
            type: "value"
          },
          series: [{
            name: "\u5176\u4ED6",
            data: e.hours.map((n, o) => [o + 0.5, n.total || 0]),
            type: "bar",
            stack: "1",
            itemStyle: {
              color: "rgb(255,219,92)"
            }
          }, {
            name: "\u6559\u5B66",
            data: e.hours.map((n, o) => [o + 0.5, (n.command || 0) + (n.dialogue || 0)]),
            type: "bar",
            stack: "1",
            itemStyle: {
              color: "rgb(103,224,227)"
            }
          }, {
            name: "\u6307\u4EE4",
            data: e.hours.map((n, o) => [o + 0.5, n.command || 0]),
            type: "bar",
            stack: "1",
            itemStyle: {
              color: "rgb(55,162,218)"
            }
          }]
        };
      }
    })
  });
}, T = (t) => {
  t.plugin(G), t.plugin(q), t.plugin(U), t.plugin(N);
};
const l = (t, e) => {
  const n = t.__vccOpts || t;
  for (const [o, s] of e)
    n[o] = s;
  return n;
}, J = {}, K = { class: "status-light" };
function Q(t, e) {
  return r(), u("span", K);
}
const I = /* @__PURE__ */ l(J, [["render", Q], ["__scopeId", "data-v-cbad49e0"]]), W = { class: "bot" }, X = { class: "info" }, Y = { class: "cur-frequency" }, tt = { style: { "margin-right": "8px" } }, et = /* @__PURE__ */ V({
  __name: "preview",
  props: {
    data: null
  },
  setup(t) {
    const e = {
      online: "\u8FD0\u884C\u4E2D",
      offline: "\u79BB\u7EBF",
      connect: "\u6B63\u5728\u8FDE\u63A5",
      reconnect: "\u6B63\u5728\u91CD\u8FDE",
      disconnect: "\u6B63\u5728\u65AD\u5F00"
    };
    return (n, o) => {
      const s = g("el-tooltip"), a = g("k-icon");
      return r(), u("div", W, [
        c("div", {
          class: "avatar",
          style: H({ backgroundImage: `url(${t.data.avatar})` }),
          onClick: o[0] || (o[0] = (f) => n.$emit("avatar-click"))
        }, [
          _(s, {
            content: e[t.data.status],
            placement: "right"
          }, {
            default: F(() => [
              _(I, {
                class: S(t.data.status)
              }, null, 8, ["class"])
            ]),
            _: 1
          }, 8, ["content"])
        ], 4),
        c("div", X, [
          c("div", null, [
            _(a, { name: "robot" }),
            A(p(t.data.username), 1)
          ]),
          c("div", null, [
            _(a, { name: "platform" }),
            A(p(t.data.platform), 1)
          ]),
          c("div", Y, [
            c("span", tt, [
              _(a, { name: "arrow-up" }),
              c("span", null, p(t.data.messageSent) + "/min", 1)
            ]),
            c("span", null, [
              _(a, { name: "arrow-down" }),
              c("span", null, p(t.data.messageReceived) + "/min", 1)
            ])
          ])
        ])
      ]);
    };
  }
});
const nt = /* @__PURE__ */ l(et, [["__scopeId", "data-v-a12625c7"]]), ot = {
  key: 0,
  class: "el-popper__empty"
}, st = /* @__PURE__ */ V({
  __name: "index",
  setup(t) {
    const e = w(() => Object.values(d.profile.bots).reduce((o, s) => o + s.messageSent, 0)), n = w(() => Object.values(d.profile.bots).reduce((o, s) => o + s.messageReceived, 0));
    return (o, s) => {
      const a = g("k-icon"), f = g("k-status");
      return r(), b(f, null, {
        tooltip: F(() => [
          i(d).profile.bots.length ? j("", !0) : (r(), u("span", ot)),
          (r(!0), u(k, null, z(i(d).profile.bots, (m, h) => (r(), b(nt, {
            key: h,
            data: m
          }, null, 8, ["data"]))), 128))
        ]),
        default: F(() => [
          (r(!0), u(k, null, z(i(d).profile.bots, (m, h) => (r(), b(I, {
            key: h,
            class: S(m.status)
          }, null, 8, ["class"]))), 128)),
          _(a, { name: "arrow-up" }),
          c("span", null, p(i(e)) + "/min", 1),
          _(a, { name: "arrow-down" }),
          c("span", null, p(i(n)) + "/min", 1)
        ]),
        _: 1
      });
    };
  }
});
const ct = /* @__PURE__ */ l(st, [["__scopeId", "data-v-8f6c5b7b"]]), rt = (t) => {
  t.slot({
    type: "status-right",
    component: ct
  });
}, ut = { class: "load-bar" }, at = { class: "title" }, it = { class: "body" }, lt = /* @__PURE__ */ V({
  __name: "load-bar",
  props: {
    rate: null,
    title: null
  },
  setup(t) {
    const e = t;
    function n(m, h = 3) {
      return (m * 100).toFixed(h) + "%";
    }
    const o = ["used", "app", "free"], s = w(() => [
      e.rate[1] - e.rate[0],
      e.rate[0],
      1 - e.rate[1]
    ]), a = w(() => s.value.indexOf(Math.max(...s.value))), f = w(() => `${n(e.rate[0], 1)} / ${n(e.rate[1], 1)}`);
    return (m, h) => (r(), u("div", ut, [
      c("span", at, p(t.title), 1),
      c("span", it, [
        (r(!0), u(k, null, z(i(o), ($, x) => (r(), u("span", {
          key: $,
          class: S([$, "bar"]),
          style: H({ width: n(i(s)[x]) })
        }, [
          x === i(a) ? (r(), u(k, { key: 0 }, [
            A(p(i(f)), 1)
          ], 64)) : j("", !0)
        ], 6))), 128))
      ])
    ]));
  }
});
const L = /* @__PURE__ */ l(lt, [["__scopeId", "data-v-7fd230e1"]]);
function E(t, e = 3) {
  return (t * 100).toFixed(e) + "%";
}
const _t = { class: "load" }, dt = /* @__PURE__ */ V({
  __name: "index",
  setup(t) {
    return (e, n) => {
      const o = g("k-status");
      return r(), b(o, null, {
        tooltip: F(() => [
          c("div", _t, [
            _(L, {
              title: "CPU",
              rate: i(d).profile.cpu
            }, null, 8, ["rate"]),
            _(L, {
              title: "\u5185\u5B58",
              rate: i(d).profile.memory
            }, null, 8, ["rate"])
          ])
        ]),
        default: F(() => [
          A(" CPU: " + p(i(E)(i(d).profile.cpu[1], 0)) + " \u5185\u5B58: " + p(i(E)(i(d).profile.memory[1], 0)), 1)
        ]),
        _: 1
      });
    };
  }
});
const pt = /* @__PURE__ */ l(dt, [["__scopeId", "data-v-3923a153"]]), mt = (t) => {
  t.slot({
    type: "status-right",
    component: pt
  });
}, ht = {}, Ct = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, ft = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M6.625 215.5l168-176C179.2 34.7 185.4 32.02 192 32.02s12.84 2.688 17.38 7.438l168 176c9.125 9.594 8.781 24.78-.8125 33.94c-9.5 9.156-24.75 8.812-33.94-.8125L216 115.9V456c0 13.25-10.75 23.1-23.1 23.1S168 469.3 168 456V115.9l-126.6 132.7C32.22 258.2 16.97 258.5 7.438 249.4C-2.156 240.2-2.5 225 6.625 215.5z"
}, null, -1), gt = [
  ft
];
function vt(t, e) {
  return r(), u("svg", Ct, gt);
}
const $t = /* @__PURE__ */ l(ht, [["render", vt]]), xt = {}, yt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, wt = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M377.4 296.6l-168 176C204.8 477.3 198.6 480 192 480s-12.84-2.688-17.38-7.438l-168-176C-2.5 286.1-2.156 271.8 7.438 262.6c9.5-9.156 24.75-8.812 33.94 .8125L168 396.1V56.02c0-13.25 10.75-24.01 23.1-24.01S216 42.77 216 56.02v340.1l126.6-132.7c9.156-9.625 24.41-9.969 33.94-.8125C386.2 271.8 386.5 286.1 377.4 296.6z"
}, null, -1), Ft = [
  wt
];
function bt(t, e) {
  return r(), u("svg", yt, Ft);
}
const At = /* @__PURE__ */ l(xt, [["render", bt]]), kt = {}, Vt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Bt = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M255.1 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 0 232.4 0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84.02L255.1 96zM255.1 141.3L221.4 106.6C196.1 81.31 160 69.77 124.7 75.66C71.21 84.58 31.1 130.9 31.1 185.1V190.9C31.1 223.6 45.55 254.7 69.42 277L250.1 445.7C251.7 447.2 253.8 448 255.1 448C258.2 448 260.3 447.2 261.9 445.7L442.6 277C466.4 254.7 480 223.6 480 190.9V185.1C480 130.9 440.8 84.58 387.3 75.66C351.1 69.77 315.9 81.31 290.6 106.6L255.1 141.3z"
}, null, -1), zt = [
  Bt
];
function St(t, e) {
  return r(), u("svg", Vt, zt);
}
const Dt = /* @__PURE__ */ l(kt, [["render", St]]), Mt = {}, Lt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Et = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M256 0C397.4 0 512 114.6 512 256C512 397.4 397.4 512 256 512C172.2 512 97.87 471.8 51.19 409.6C45.88 402.5 47.31 392.5 54.37 387.2C61.44 381.9 71.47 383.3 76.78 390.4C117.7 444.8 182.7 480 256 480C379.7 480 480 379.7 480 256C480 132.3 379.7 32 256 32C166.7 32 89.51 84.3 53.55 160H144C152.8 160 160 167.2 160 176C160 184.8 152.8 192 144 192H16C7.164 192 0 184.8 0 176V48C0 39.16 7.164 32 16 32C24.84 32 32 39.16 32 48V131.1C75.66 53.29 159.6 0 256 0zM256 128C264.8 128 272 135.2 272 144V249.4L347.3 324.7C353.6 330.9 353.6 341.1 347.3 347.3C341.1 353.6 330.9 353.6 324.7 347.3L244.7 267.3C241.7 264.3 239.1 260.2 239.1 256V144C239.1 135.2 247.2 128 255.1 128H256z"
}, null, -1), Ht = [
  Et
];
function jt(t, e) {
  return r(), u("svg", Lt, Ht);
}
const It = /* @__PURE__ */ l(Mt, [["render", jt]]), Ot = {}, Zt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Pt = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M12.41 148.02l232.94 105.67c6.8 3.09 14.49 3.09 21.29 0l232.94-105.67c16.55-7.51 16.55-32.52 0-40.03L266.65 2.31a25.607 25.607 0 0 0-21.29 0L12.41 107.98c-16.55 7.51-16.55 32.53 0 40.04zm487.18 88.28l-58.09-26.33-161.64 73.27c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.51 209.97l-58.1 26.33c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 276.3c16.55-7.5 16.55-32.5 0-40zm0 127.8l-57.87-26.23-161.86 73.37c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.29 337.87 12.41 364.1c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 404.1c16.55-7.5 16.55-32.5 0-40z"
}, null, -1), Nt = [
  Pt
];
function Ut(t, e) {
  return r(), u("svg", Zt, Nt);
}
const Rt = /* @__PURE__ */ l(Ot, [["render", Ut]]), Gt = {}, qt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, Tt = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M32,224H64V416H32A31.96166,31.96166,0,0,1,0,384V256A31.96166,31.96166,0,0,1,32,224Zm512-48V448a64.06328,64.06328,0,0,1-64,64H160a64.06328,64.06328,0,0,1-64-64V176a79.974,79.974,0,0,1,80-80H288V32a32,32,0,0,1,64,0V96H464A79.974,79.974,0,0,1,544,176ZM264,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,264,256Zm-8,128H192v32h64Zm96,0H288v32h64ZM456,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,456,256Zm-8,128H384v32h64ZM640,256V384a31.96166,31.96166,0,0,1-32,32H576V224h32A31.96166,31.96166,0,0,1,640,256Z"
}, null, -1), Jt = [
  Tt
];
function Kt(t, e) {
  return r(), u("svg", qt, Jt);
}
const Qt = /* @__PURE__ */ l(Gt, [["render", Kt]]), Wt = {}, Xt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, Yt = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M448 32h-384c-35.35 0-64 28.65-64 64v320c0 35.35 28.65 64 64 64h384c35.35 0 64-28.65 64-64V96C512 60.65 483.3 32 448 32zM480 416c0 17.64-14.36 32-32 32H64c-17.64 0-32-14.36-32-32v-96c0-17.64 14.36-32 32-32h384C465.6 288 480 302.4 480 320V416zM480 264.9C470.6 259.4 459.7 256 448 256H64C52.28 256 41.45 259.4 32 264.9V96c0-17.64 14.36-32 32-32h384c17.64 0 32 14.36 32 32V264.9zM320 344c-13.25 0-24 10.74-24 24c0 13.25 10.75 24 24 24c13.26 0 24-10.75 24-24C344 354.7 333.3 344 320 344zM416 344c-13.25 0-24 10.74-24 24c0 13.25 10.75 24 24 24c13.26 0 24-10.75 24-24C440 354.7 429.3 344 416 344z"
}, null, -1), t1 = [
  Yt
];
function e1(t, e) {
  return r(), u("svg", Xt, t1);
}
const n1 = /* @__PURE__ */ l(Wt, [["render", e1]]), o1 = {}, s1 = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, c1 = /* @__PURE__ */ c("path", {
  fill: "currentColor",
  d: "M319.9 320c57.41 0 103.1-46.56 103.1-104c0-57.44-46.54-104-103.1-104c-57.41 0-103.1 46.56-103.1 104C215.9 273.4 262.5 320 319.9 320zM319.9 144c39.68 0 71.96 32.3 71.96 72S359.5 288 319.9 288S247.9 255.7 247.9 216S280.2 144 319.9 144zM369.9 352H270.1C191.6 352 128 411.7 128 485.3C128 500.1 140.7 512 156.4 512h327.2C499.3 512 512 500.1 512 485.3C512 411.7 448.4 352 369.9 352zM160.2 480c3.021-53.41 51.19-96 109.1-96H369.9c58.78 0 106.9 42.59 109.1 96H160.2zM512 160c44.18 0 80-35.82 80-80S556.2 0 512 0c-44.18 0-80 35.82-80 80S467.8 160 512 160zM512 32c26.47 0 48 21.53 48 48S538.5 128 512 128s-48-21.53-48-48S485.5 32 512 32zM128 160c44.18 0 80-35.82 80-80S172.2 0 128 0C83.82 0 48 35.82 48 80S83.82 160 128 160zM128 32c26.47 0 48 21.53 48 48S154.5 128 128 128S80 106.5 80 80S101.5 32 128 32zM561.1 192H496c-11.16 0-22.08 2.5-32.47 7.438c-7.984 3.797-11.39 13.34-7.594 21.31s13.38 11.39 21.31 7.594C483.3 225.5 489.6 224 496 224h65.08C586.1 224 608 246.7 608 274.7V288c0 8.844 7.156 16 16 16S640 296.8 640 288V274.7C640 229.1 604.6 192 561.1 192zM162.8 228.3c7.938 3.797 17.53 .375 21.31-7.594c3.797-7.969 .3906-17.52-7.594-21.31C166.1 194.5 155.2 192 144 192H78.92C35.41 192 0 229.1 0 274.7V288c0 8.844 7.156 16 16 16S32 296.8 32 288V274.7C32 246.7 53.05 224 78.92 224H144C150.4 224 156.7 225.5 162.8 228.3z"
}, null, -1), r1 = [
  c1
];
function u1(t, e) {
  return r(), u("svg", s1, r1);
}
const a1 = /* @__PURE__ */ l(o1, [["render", u1]]);
C.register("arrow-up", $t);
C.register("arrow-down", At);
C.register("heart", Dt);
C.register("history", It);
C.register("platform", Rt);
C.register("robot", Qt);
C.register("storage", n1);
C.register("users", a1);
const _1 = O((t) => {
  t.plugin(T), t.plugin(rt), t.plugin(mt), t.slot({
    type: "numeric",
    component: y.numeric({
      title: "\u8FD1\u671F\u6D88\u606F\u9891\u7387",
      icon: "history",
      fields: ["stats"],
      content({ stats: e }) {
        return Object.values(e.botSend).reduce((n, o) => n + o, 0).toFixed(1) + " / d";
      }
    })
  }), t.slot({
    type: "numeric",
    component: y.numeric({
      title: "\u8D44\u6E90\u670D\u52A1\u5668",
      icon: "storage",
      type: "size",
      fields: ["meta"],
      content: ({ meta: e }) => e.assetSize
    })
  }), t.slot({
    type: "numeric",
    component: y.numeric({
      title: "\u6D3B\u8DC3\u7528\u6237\u6570\u91CF",
      icon: "heart",
      fields: ["meta"],
      content: ({ meta: e }) => e.activeUsers
    })
  }), t.slot({
    type: "numeric",
    component: y.numeric({
      title: "\u6D3B\u8DC3\u7FA4\u7EC4\u6570\u91CF",
      icon: "users",
      fields: ["meta"],
      content: ({ meta: e }) => e.activeGuilds
    })
  });
});
export {
  _1 as default
};
