// @koishijs/plugin-status/dist/index.js
import { store as _, icons as b, defineExtension as Z } from "../client.js";
import { openBlock as o, createElementBlock as a, defineComponent as B, resolveComponent as h, createElementVNode as n, normalizeStyle as F, createVNode as i, withCtx as g, normalizeClass as V, createTextVNode as x, toDisplayString as u, computed as v, unref as s, createBlock as $, createCommentVNode as y, Fragment as k, renderList as C } from "../vue.js";
const p = (t, e) => {
  const l = t.__vccOpts || t;
  for (const [c, r] of e)
    l[c] = r;
  return l;
}, z = {}, E = { class: "status-light" };
function I(t, e) {
  return o(), a("span", E);
}
const M = /* @__PURE__ */ p(z, [["render", I], ["__scopeId", "data-v-cbad49e0"]]), S = { class: "bot" }, N = { class: "info" }, O = { class: "cur-frequency" }, P = { style: { "margin-right": "8px" } }, R = /* @__PURE__ */ B({
  __name: "preview",
  props: {
    data: null
  },
  setup(t) {
    const e = {
      online: "\u8FD0\u884C\u4E2D",
      offline: "\u79BB\u7EBF",
      connect: "\u6B63\u5728\u8FDE\u63A5",
      reconnect: "\u6B63\u5728\u91CD\u8FDE",
      disconnect: "\u6B63\u5728\u65AD\u5F00"
    };
    return (l, c) => {
      const r = h("el-tooltip"), d = h("k-icon");
      return o(), a("div", S, [
        n("div", {
          class: "avatar",
          style: F({ backgroundImage: `url(${t.data.avatar})` }),
          onClick: c[0] || (c[0] = (w) => l.$emit("avatar-click"))
        }, [
          i(r, {
            content: e[t.data.status],
            placement: "right"
          }, {
            default: g(() => [
              i(M, {
                class: V(t.data.status)
              }, null, 8, ["class"])
            ]),
            _: 1
          }, 8, ["content"])
        ], 4),
        n("div", N, [
          n("div", null, [
            i(d, { name: "robot" }),
            x(u(t.data.username), 1)
          ]),
          n("div", null, [
            i(d, { name: "platform" }),
            x(u(t.data.platform), 1)
          ]),
          n("div", O, [
            n("span", P, [
              i(d, { name: "arrow-up" }),
              n("span", null, u(t.data.messageSent) + "/min", 1)
            ]),
            n("span", null, [
              i(d, { name: "arrow-down" }),
              n("span", null, u(t.data.messageReceived) + "/min", 1)
            ])
          ])
        ])
      ]);
    };
  }
});
const U = /* @__PURE__ */ p(R, [["__scopeId", "data-v-a12625c7"]]), j = {
  key: 0,
  class: "el-popper__empty"
}, q = /* @__PURE__ */ B({
  __name: "index",
  setup(t) {
    const e = v(() => Object.values(_.profile.bots).reduce((c, r) => c + r.messageSent, 0)), l = v(() => Object.values(_.profile.bots).reduce((c, r) => c + r.messageReceived, 0));
    return (c, r) => {
      const d = h("k-icon"), w = h("k-status");
      return s(_).profile ? (o(), $(w, { key: 0 }, {
        tooltip: g(() => [
          s(_).profile.bots.length ? y("", !0) : (o(), a("span", j)),
          (o(!0), a(k, null, C(s(_).profile.bots, (m, f) => (o(), $(U, {
            key: f,
            data: m
          }, null, 8, ["data"]))), 128))
        ]),
        default: g(() => [
          (o(!0), a(k, null, C(s(_).profile.bots, (m, f) => (o(), $(M, {
            key: f,
            class: V(m.status)
          }, null, 8, ["class"]))), 128)),
          i(d, { name: "arrow-up" }),
          n("span", null, u(s(e)) + "/min", 1),
          i(d, { name: "arrow-down" }),
          n("span", null, u(s(l)) + "/min", 1)
        ]),
        _: 1
      })) : y("", !0);
    };
  }
});
const T = /* @__PURE__ */ p(q, [["__scopeId", "data-v-cbf8cb3a"]]), G = (t) => {
  t.slot({
    type: "status-right",
    component: T
  });
}, J = { class: "load-bar" }, K = { class: "title" }, Q = { class: "body" }, W = /* @__PURE__ */ B({
  __name: "load-bar",
  props: {
    rate: null,
    title: null
  },
  setup(t) {
    const e = t;
    function l(m, f = 3) {
      return (m * 100).toFixed(f) + "%";
    }
    const c = ["used", "app", "free"], r = v(() => [
      e.rate[1] - e.rate[0],
      e.rate[0],
      1 - e.rate[1]
    ]), d = v(() => r.value.indexOf(Math.max(...r.value))), w = v(() => `${l(e.rate[0], 1)} / ${l(e.rate[1], 1)}`);
    return (m, f) => (o(), a("div", J, [
      n("span", K, u(t.title), 1),
      n("span", Q, [
        (o(!0), a(k, null, C(s(c), (L, A) => (o(), a("span", {
          key: L,
          class: V([L, "bar"]),
          style: F({ width: l(s(r)[A]) })
        }, [
          A === s(d) ? (o(), a(k, { key: 0 }, [
            x(u(s(w)), 1)
          ], 64)) : y("", !0)
        ], 6))), 128))
      ])
    ]));
  }
});
const H = /* @__PURE__ */ p(W, [["__scopeId", "data-v-7fd230e1"]]);
function D(t, e = 3) {
  return (t * 100).toFixed(e) + "%";
}
const X = { class: "load" }, Y = /* @__PURE__ */ B({
  __name: "index",
  setup(t) {
    return (e, l) => {
      const c = h("k-status");
      return s(_).profile ? (o(), $(c, { key: 0 }, {
        tooltip: g(() => [
          n("div", X, [
            i(H, {
              title: "CPU",
              rate: s(_).profile.cpu
            }, null, 8, ["rate"]),
            i(H, {
              title: "\u5185\u5B58",
              rate: s(_).profile.memory
            }, null, 8, ["rate"])
          ])
        ]),
        default: g(() => [
          x(" CPU: " + u(s(D)(s(_).profile.cpu[1], 0)) + " \u5185\u5B58: " + u(s(D)(s(_).profile.memory[1], 0)), 1)
        ]),
        _: 1
      })) : y("", !0);
    };
  }
});
const tt = /* @__PURE__ */ p(Y, [["__scopeId", "data-v-35a62917"]]), et = (t) => {
  t.slot({
    type: "status-right",
    component: tt
  });
}, nt = {}, st = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, ot = /* @__PURE__ */ n("path", {
  fill: "currentColor",
  d: "M6.625 215.5l168-176C179.2 34.7 185.4 32.02 192 32.02s12.84 2.688 17.38 7.438l168 176c9.125 9.594 8.781 24.78-.8125 33.94c-9.5 9.156-24.75 8.812-33.94-.8125L216 115.9V456c0 13.25-10.75 23.1-23.1 23.1S168 469.3 168 456V115.9l-126.6 132.7C32.22 258.2 16.97 258.5 7.438 249.4C-2.156 240.2-2.5 225 6.625 215.5z"
}, null, -1), ct = [
  ot
];
function at(t, e) {
  return o(), a("svg", st, ct);
}
const rt = /* @__PURE__ */ p(nt, [["render", at]]), lt = {}, _t = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 384 512"
}, it = /* @__PURE__ */ n("path", {
  fill: "currentColor",
  d: "M377.4 296.6l-168 176C204.8 477.3 198.6 480 192 480s-12.84-2.688-17.38-7.438l-168-176C-2.5 286.1-2.156 271.8 7.438 262.6c9.5-9.156 24.75-8.812 33.94 .8125L168 396.1V56.02c0-13.25 10.75-24.01 23.1-24.01S216 42.77 216 56.02v340.1l126.6-132.7c9.156-9.625 24.41-9.969 33.94-.8125C386.2 271.8 386.5 286.1 377.4 296.6z"
}, null, -1), ut = [
  it
];
function dt(t, e) {
  return o(), a("svg", _t, ut);
}
const pt = /* @__PURE__ */ p(lt, [["render", dt]]), mt = {}, ft = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
}, vt = /* @__PURE__ */ n("path", {
  fill: "currentColor",
  d: "M12.41 148.02l232.94 105.67c6.8 3.09 14.49 3.09 21.29 0l232.94-105.67c16.55-7.51 16.55-32.52 0-40.03L266.65 2.31a25.607 25.607 0 0 0-21.29 0L12.41 107.98c-16.55 7.51-16.55 32.53 0 40.04zm487.18 88.28l-58.09-26.33-161.64 73.27c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.51 209.97l-58.1 26.33c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 276.3c16.55-7.5 16.55-32.5 0-40zm0 127.8l-57.87-26.23-161.86 73.37c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.29 337.87 12.41 364.1c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 404.1c16.55-7.5 16.55-32.5 0-40z"
}, null, -1), ht = [
  vt
];
function gt(t, e) {
  return o(), a("svg", ft, ht);
}
const wt = /* @__PURE__ */ p(mt, [["render", gt]]), $t = {}, xt = {
  class: "k-icon",
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 640 512"
}, yt = /* @__PURE__ */ n("path", {
  fill: "currentColor",
  d: "M32,224H64V416H32A31.96166,31.96166,0,0,1,0,384V256A31.96166,31.96166,0,0,1,32,224Zm512-48V448a64.06328,64.06328,0,0,1-64,64H160a64.06328,64.06328,0,0,1-64-64V176a79.974,79.974,0,0,1,80-80H288V32a32,32,0,0,1,64,0V96H464A79.974,79.974,0,0,1,544,176ZM264,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,264,256Zm-8,128H192v32h64Zm96,0H288v32h64ZM456,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,456,256Zm-8,128H384v32h64ZM640,256V384a31.96166,31.96166,0,0,1-32,32H576V224h32A31.96166,31.96166,0,0,1,640,256Z"
}, null, -1), kt = [
  yt
];
function bt(t, e) {
  return o(), a("svg", xt, kt);
}
const Bt = /* @__PURE__ */ p($t, [["render", bt]]);
b.register("arrow-up", rt);
b.register("arrow-down", pt);
b.register("platform", wt);
b.register("robot", Bt);
const Lt = Z((t) => {
  t.plugin(G), t.plugin(et);
});
export {
  Lt as default
};
