// better-lock/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BetterLock = exports.BetterLockWaitTimeoutError = exports.BetterLockQueueOverflowError = exports.BetterLockJobAbortedError = exports.BetterLockInvalidArgumentError = exports.BetterLockInternalError = exports.BetterLockExecutionTimeoutError = exports.BetterLockError = void 0;
var better_lock_1 = require("./better_lock");
var errors_1 = require("./errors");
Object.defineProperty(exports, "BetterLockError", { enumerable: true, get: function () { return errors_1.BetterLockError; } });
Object.defineProperty(exports, "BetterLockExecutionTimeoutError", { enumerable: true, get: function () { return errors_1.BetterLockExecutionTimeoutError; } });
Object.defineProperty(exports, "BetterLockInternalError", { enumerable: true, get: function () { return errors_1.BetterLockInternalError; } });
Object.defineProperty(exports, "BetterLockInvalidArgumentError", { enumerable: true, get: function () { return errors_1.BetterLockInvalidArgumentError; } });
Object.defineProperty(exports, "BetterLockJobAbortedError", { enumerable: true, get: function () { return errors_1.BetterLockJobAbortedError; } });
Object.defineProperty(exports, "BetterLockQueueOverflowError", { enumerable: true, get: function () { return errors_1.BetterLockQueueOverflowError; } });
Object.defineProperty(exports, "BetterLockWaitTimeoutError", { enumerable: true, get: function () { return errors_1.BetterLockWaitTimeoutError; } });
function attachExports(target, exports) {
    Object.assign(target, exports);
    return target;
}
exports.BetterLock = attachExports(better_lock_1.BetterLock, {
    BetterLock: better_lock_1.BetterLock,
    Error: errors_1.BetterLockError,
    InternalError: errors_1.BetterLockInternalError,
    InvalidArgumentError: errors_1.BetterLockInvalidArgumentError,
    WaitTimeoutError: errors_1.BetterLockWaitTimeoutError,
    ExecutionTimeoutError: errors_1.BetterLockExecutionTimeoutError,
    QueueOverflowError: errors_1.BetterLockQueueOverflowError,
    JobAbortedError: errors_1.BetterLockJobAbortedError,
});
exports.default = exports.BetterLock;
